---
output: html_document
editor_options: 
  chunk_output_type: console
---


# QC and cell filtering {#QC}
## load packages

```{r}

source("./00a_dependencies.R")

```


## load data

```{r}

gex <- readRDS("../results/GEX1_RAW.rds")

```


## gather gene annotations for GRCh38.98

```{r}

#read GRCh38.98 .gtf file used during alignment
#gtf.GRCh38.98 <- readGFF("../annot/refdata-gex-GRCh38-2020-A.genes.gtf.gz")

#annot <- unique(gtf.GRCh38.98[gtf.GRCh38.98$type == "gene", c(1, 9, 11, 12)])
#colnames(annot) <- c("chromosome_name", "gene_id", "gene_biotype", "external_gene_name")
#annot$chromosome_name <- as.character(annot$chromosome_name)
#annot[!grepl("^chr[123456789XYMT]", annot[, "chromosome_name"]), "chromosome_name"] <- "other"
#saveRDS(annot, "../annot/refdata-gex-GRCh38-2020-A.genes.gtf.annot.rds")

annot <- readRDS("../annot/refdata-gex-GRCh38-2020-A.genes.gtf.annot.rds")


```


## mark doublets

```{r}

# scDblFinder requires around 3 min for this data set using 16 cores
gex <-
  as.Seurat(
    scDblFinder(
      as.SingleCellExperiment(gex),
      samples = "orig.ident",
      BPPARAM = MulticoreParam(16)
    )
  )

gex@meta.data$ident <- NULL


```


## cualculate QC and mark cells per sample

```{r}

gex <- SplitObject(gex, split.by = "orig.ident")

counts_features_quantile <- 0.98
mito_cutoff <- 20
ribo_cutoff <- 5
hb_cutoff <- 1

gex <- lapply(gex, annot = annot, function(x, annot) {
  total <- colSums(x) #Total number of cells
  
  x$percent_mito <- colSums(x[grepl("^MT-", rownames(x)), ]) / total * 100
  x$percent_ribo <- colSums(x[grepl("^RP[LS]", rownames(x)), ]) / total * 100
  x$percent_hb <-   colSums(x[grepl("^HB[AB]", rownames(x)), ]) / total * 100
  
  gene_biotype <- annot[match(rownames(x), annot$external_gene_name), "gene_biotype"]
  x$percent_protein_coding <- colSums(x[(gene_biotype == "protein_coding") &
                                          (!is.na(gene_biotype)), ]) / total * 100
  
  
  x$count_th <- x$nCount_RNA > quantile(x$nCount_RNA, 
                                        counts_features_quantile)
  x$feature_th <- (x$nFeature_RNA > quantile(x$nFeature_RNA, 
                                             counts_features_quantile)) |
    (x$nFeature_RNA < 200)
  
  x$mito_th <- x$percent_mito > mito_cutoff
  x$ribo_th <- x$percent_ribo < ribo_cutoff
  x$hb_th <- x$percent_hb > hb_cutoff
  
  return(x)
  
})

gex <- merge(gex[[1]] , gex[-1])

#mark cells to be removed
gex$cells_discard <- rowSums(gex@meta.data[, grepl("_th", colnames(gex@meta.data))]) > 0


```


# summarize QC

```{r}

discarded_stats <- as_tibble(gex@meta.data) %>%
  group_by(orig.ident) %>%
  summarise(
    cells = length(orig.ident),
    count_discarded = sum(count_th),
    feature_discarded = sum(feature_th),
    mito_discarded = sum(mito_th),
    ribo_discarded = sum(ribo_th),
    hb_discarded = sum(hb_th),
    pred_doublets = sum(scDblFinder.class == "doublet"),
    total_discarded = sum(cells_discard)
  )

write.csv(discarded_stats, 
          paste0("../results/GEX_stats_discarded.csv"), 
          row.names = FALSE, 
          quote = FALSE)

p_discarded_stats <- discarded_stats %>% 
  gather(discarded, cellcount, cells:total_discarded) %>% 
  ggplot(aes(x = orig.ident, 
             y = cellcount, 
             fill = discarded)) + 
      geom_bar(stat = "identity", 
               color = "black", 
               position = position_dodge()) + 
      theme(legend.title = element_blank(),
            axis.text.x = element_text(face = "bold", angle = 90, vjust = 0.5, hjust = 1),
            axis.text.y = element_text(face = "bold")) +
      ggtitle(paste0("Number of cells total/number of discarded cells (", 
                     sum(discarded_stats$cells), "/", 
                     sum(discarded_stats$total_discarded), ")"))

ggsave2(paste0("../results/GEX_discarded_cellcounts.png"), 
          p_discarded_stats,
          width = 25, height = 20, unit = "cm")


```


## plot SFigXX, QC

```{r}

#violins
my_pars <- c("nCount_RNA", "nFeature_RNA", "percent_mito", "percent_ribo", "percent_hb")
names(my_pars) <- c("count_th", "feature_th", "mito_th", "ribo_th", "hb_th")

p_out <- list()
for (i in my_pars){
  
  print(i)
  
  p_out[[i]] <- plotColData(as.SingleCellExperiment(gex), 
                    x = "orig.ident", 
                    y = i, 
                    colour_by = names(my_pars)[which(my_pars == i)]) + 
    ggtitle(paste0(i, 
                   ", cells marked to be discarded: ", 
                   sum(gex@meta.data[,names(my_pars)[which(my_pars == i)]]))) +
    xlab("") + 
    ylab(i) + 
    theme(axis.text.x = element_text(angle = 90, 
                                     hjust = 1, 
                                     vjust = 0.5,
                                     size = 8),
          plot.title = element_text(size = 8),
          legend.position = "none")

}

#get legend
p_legend <- plotColData(as.SingleCellExperiment(gex), 
                                 x = "orig.ident", 
                                 y = my_pars[1], 
                                 colour_by = "cells_discard") +
  scale_fill_continuous(name = "Discarded cells")

p_out[["legend"]] <- get_legend(p_legend) %>% ggpubr::as_ggplot()

ggsave2(paste0("../results/SFig23.png"), 
        plot_grid(plotlist = p_out, ncol = 2, labels = NULL, align = "hv") + 
          bgcolor("white") + 
          border("white"),
        width = 21, height = 21, unit = "cm")


```


## filter genes and filter cells

```{r}

#filter annotation file, keeping proteing coding and VDJ genes
annot_coding <- unique(annot[!(annot$chromosome_name %in% c("chrY", "other", "chrM")) & 
                       (annot$gene_biotype %in% c("protein_coding", 
                                    "IG_V_gene", "IG_D_gene", "IG_J_gene", "IG_C_gene",
                                    "TR_V_gene", "TR_D_gene", "TR_J_gene", "TR_C_gene")), 
                       c("external_gene_name", "gene_biotype", "chromosome_name")])

genes_keep <- rownames(gex) %in% annot_coding$external_gene_name

gex <- gex[genes_keep, !gex$cells_discard] 

Idents(object = gex) <- "orig.ident"


```


## save rds GEX2_FILTERED

```{r}

saveRDS(gex, file = "../results/GEX2_FILTERED.rds")
#gex <- readRDS("../results/GEX2_FILTERED.rds")

```


## print sessionInfo()

```{r}

sessionInfo()

```

