---
output: html_document
editor_options: 
  chunk_output_type: console
---


# VDJ analysis {#VDJ}
## load packages

```{r}

source("./00a_dependencies.R")

```


## load Seurat object

```{r}

gex <- readRDS("../results/GEX6_BANNOT.rds")

```


## load pheno file

```{r}

source("./00b_pheno.R")

#Four non-overlapping groups
#1. SSAB: SSA+/SSB+
#2. SSA+: SSA+/SSB-
#3. SSA-: SSA-/SSB-
#4. CTRL: CTRL


```

## cellType colors

```{r}

source("./00c_colors.R")

```


## immcantation
### load output from Immcantation and perform mutation analysis

```{r}

files_B <- list.files("../data/clono_B_VDJ_tmp", 
                      pattern = "*germ-pass.tsv", 
                      recursive = TRUE, 
                      full.names = TRUE)
ncpu <- cpuCount()
imm <- NULL
for (i in files_B) {
  
  cat("reading: ", print(i), "\n")
  tmp <- read_rearrangement(i)
  
  ####---- get sample name
  j <- gsub("_heavy_germ-pass.tsv", "", gsub(".*out/", "", i))
  cat("sample name: ", print(j), "\n")
  
  ####---- number of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = FALSE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_v_total'
  
  ####---- frequency of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = TRUE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_v_total'
  
  ####---- number of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")

  ####---- number of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
   
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_total'
  
  ####---- frequency of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_total'
 
  ####---- properties for CDR3 aa's
  tmp <- aminoAcidProperties(tmp, 
                             seq = "junction", 
                             nt = TRUE, 
                             trim = TRUE, 
                             label = "cdr3")
  
  tmp$cell_id <- paste0(j, "_", gsub("-1", "", tmp$cell_id))
  tmp$orig.ident <- gsub("a|b", "", gsub("_.*", "", tmp$cell_id))
  tmp$sequence_id <- paste0(j, "_", gsub("-1", "", tmp$sequence_id))
  rownames(tmp) <- tmp$cell_id
  
  if (is.null(imm)) {
    imm <- tmp
  } else {
    imm <- rbind(imm, tmp)
  }
  
}


```


### integrate immcantation VDJ data with seurat object

```{r}

#select immcantation columns NOT to be added to gex
cols <- c("sequence", "sequence_alignment", "germline_alignment", 
          "v_cigar", "d_cigar", "j_cigar",
          "fwr1", "fwr2", "fwr3", "fwr4", "cdr1", "cdr2", 
          "germline_alignment_d_mask")

imm_add <- imm[, !(colnames(imm) %in% cols)]
rownames(imm_add) <- imm_add$cell_id

gex <- AddMetaData(gex, imm_add)


```


### add pheno data to immcantation object

```{r}

#imm$orig.ident <- gsub("a|b", "", gsub("_.*", "", imm$cell_id))
imm <- merge(imm, pheno[, c(2,8)], 
             by.x = "orig.ident", 
             by.y = "orig.ident", 
             all.x = TRUE); head(imm)
imm <- imm[, c(2:ncol(imm), 1)]


```


### save immcantation VDJ data

```{r}

colnames(imm)[colnames(imm) == 'orig.ident'] <- 'repertoire_id'
imm$duplicate_count <- imm$umi_count
colnames(imm)[colnames(imm) == 'clone_id'] <- 'clone_id_perSample'

#write_rearrangement(imm, "../results/VDJ_immcantation_out.tsv")
#imm <- read_rearrangement("../results/VDJ_immcantation_out.tsv")

#imm_out <- imm[, c("sequence_id", "v_call", "j_call", "junction")]
#write_rearrangement(imm_out, "../results/VDJ_immcantation_out_forDefineClones.tsv")


```


## read compairr out

```{r}

cairr.aa <- read.table("../results/VDJ_compairr_out_aa.tsv", header = TRUE)
cairr.aa$barcode <- gsub("_contig_1", "", gsub("_contig_2", "", gsub("_contig_3", "", cairr.aa$sequence_id)))
rownames(cairr.aa) <- cairr.aa$barcode

colnames(cairr.aa)[1:2] <- paste0(colnames(cairr.aa)[1:2], "_airr")

#add to gex
gex <- AddMetaData(gex, cairr.aa[, 1:2])

gex$cloneType_airr_short <- gex$cloneType_airr
gex$cloneType_airr_short <- as.factor(gsub(" \\((.*)\\).*", "", gex$cloneType_airr_short))


```


## VDJ mutation
### add categorical variable for total mutation burden

```{r}

# unmutated
# low < 1%
# med < 2%
# high > 2%

gex$mu_load <- ifelse(gex$mu_freq_total == 0, "unmutated", 
                 ifelse(gex$mu_freq_total > 0 & gex$mu_freq_total <= 0.01, "low",
                        ifelse(gex$mu_freq_total > 0.01 & gex$mu_freq_total <= 0.02, "med",
                               ifelse(gex$mu_freq_total > 0.02, "high", NA))))

gex$mu_load <- factor(gex$mu_load, levels = c("high", "med", "low", "unmutated"))


```


### add categorical variable for clonotype expansion

```{r}

# "Hyperexpanded (100 < X <= 500)"
# "Large (20 < X <= 100)"
# "Medium (5 < X <= 20)"
# "Small (1 < X <= 5)"
# "Single (0 < X <= 1)"

table(gex$cluster_size_airr, useNA ="always")
factor(slot(seurat, "meta.data")$cloneType, 
                levels = c("Hyperexpanded (100 < X <= 500)", 
                           "Large (20 < X <= 100)", 
                           "Medium (5 < X <= 20)", 
                           "Small (1 < X <= 5)", 
                           "Single (0 < X <= 1)", 
                           NA))

gex$cloneType_airr <- factor(ifelse(is.na(gex$cluster_size_airr), NA, 
                             ifelse(gex$cluster_size_airr == 1, "Single (0 < X <= 1)", 
                                    ifelse((gex$cluster_size_airr > 1) & (gex$cluster_size_airr <= 5), "Small (1 < X <= 5)", ifelse((gex$cluster_size_airr > 5) & (gex$cluster_size_airr <= 20), "Medium (5 < X <= 20)", ifelse((gex$cluster_size_airr > 20) & (gex$cluster_size_airr <= 100), "Large (20 < X <= 100)", "Hyperexpanded (100 < X)"))))), 
                             levels = c("Hyperexpanded (100 < X <= 500)",
                                        "Large (20 < X <= 100)", 
                                        "Medium (5 < X <= 20)", 
                                        "Small (1 < X <= 5)", 
                                        "Single (0 < X <= 1)", NA))

```


## scRepertiore BCR
### read 10x BCR files

```{r}

# read 10x VDJ BCR files
samples_B <- list.files("../data/clono_B_VDJ")

clono_B <- lapply(samples_B, function(x){
  
  tmp <- read.csv(paste0("../data/clono_B_VDJ/", x, "/", x, "_B_filtered_contig_annotations.csv"), 
                  stringsAsFactors = FALSE)
  return(tmp)
  
})

names(clono_B) <- samples_B

#remove trailing -1
clono_B <- lapply(seq_along(clono_B), function(i){
  clono_B[[i]]$barcode <- gsub("-1", "", clono_B[[i]]$barcode)
  return(clono_B[[i]])
})

names(clono_B) <- samples_B

clono_B <- clono_B[order(names(clono_B))]

saveRDS(clono_B, paste0("../results/clono_B_list.rds"))
#clono_B <- readRDS("../results/clono_B_list.rds")


```


### generate a combined data frame for BCR sequences using scRepertoire

```{r}

#split combineBCR and do it per sample to avoid memory issues
#takes on average 20min/sample to run
combined_B <- lapply(1:length(clono_B), function(x) {
  
  print(names(clono_B[x]))
  print(nrow(clono_B[[x]]))
  print(Sys.time())
  
  tmp <- combineBCR(clono_B[[x]],
                    samples = names(clono_B[x]),
                    ID = NULL,
                    removeNA = TRUE,
                    removeMulti = TRUE)

  return(tmp)
  
})

#reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
combined_B <- lapply(1:length(combined_B), function(x){
  tmp <- combined_B[[x]][[1]]
  return(tmp)
})

#---- handle samples where libraries were created and sequenced twice
combined_B[["P007"]] <- rbind(combined_B[["P007a"]], combined_B[["P007b"]])
combined_B[["P007a"]] <- NULL
combined_B[["P007b"]] <- NULL

combined_B[["P008"]] <- rbind(combined_B[["P008a"]], combined_B[["P008b"]])
combined_B[["P008a"]] <- NULL
combined_B[["P008b"]] <- NULL

combined_B <- combined_B[order(names(combined_B))]

names_tmp <- names(combined_B)
combined_B <- lapply(seq_along(combined_B), function(i) {
  combined_B[[i]]$sample <- gsub("a|b", "", combined_B[[i]]$sample)
  return(combined_B[[i]])
})

names(combined_B) <- names_tmp

saveRDS(combined_B, paste0("../results/combined_BCR_scRepertoire.rds"))
#combined_B <- readRDS("../results/combined_BCR_scRepertoire.rds")


```


### Integrate scRepertoir out with Seurat object

```{r}

gex <- combineExpression(combined_B, gex)

```


## add IgH call and VDJ genes without allele info to Seurat object

```{r}

igh_10x <- lapply(1:length(clono_B), function(x) {
  
  tmp <- clono_B[[x]][clono_B[[x]]$chain == "IGH" & 
                        clono_B[[x]]$productive == "true" & 
                        !is.na(clono_B[[x]]$c_gene) & 
                        clono_B[[x]]$c_gene != "" &
                        clono_B[[x]]$c_gene != "IGLC1" &
                        clono_B[[x]]$c_gene != "IGLC2" , c(1, 10)]
  tmp$barcode <- paste0(names(clono_B)[x], "_", tmp$barcode)
  names(tmp)[2] <- "c_gene_10x"

  tmp <- tmp[!duplicated(tmp$barcode), ]
  
  rownames(tmp) <- tmp$barcode
  tmp$barcode <- NULL
  
  return(tmp)
  }
)

igh_10x <- do.call(rbind.data.frame, igh_10x)
table(igh_10x$c_gene_10x, useNA = "always")
gex <- AddMetaData(gex, igh_10x)

gex$c_gene_10x <- factor(gex$c_gene_10x, levels = sort(unique(gex$c_gene_10x)))
gex$c_gene_10x <- as.character(gex$c_gene_10x)


#get  IGH _call genes without allele info
gex$IGHV_gene <- unlist(lapply(strsplit(gex$v_call,
                                   split = "\\*"), function(x) {return(x[1])}))

gex$IGHD_gene <- unlist(lapply(strsplit(gex$d_call,
                                   split = "\\*"), function(x) {return(x[1])}))

gex$IGHJ_gene <- unlist(lapply(strsplit(gex$j_call,
                                   split = "\\*"), function(x) {return(x[1])}))

gex$VDJ_genes <- paste(gex$IGHV_gene, gex$IGHD_gene, gex$IGHJ_gene, sep = "_")
gex$VDJ_genes[gex$VDJ_genes == "NA_NA_NA"] <- NA


```


## subset gex@meta.data for plotting

```{r}

gex.meta <- gex@meta.data[(gex$cluster_cellType_manual_fine == "Memory_DN1_ID3" |
                          gex$cluster_cellType_manual_fine == "Memory_DN2" |
                          gex$cluster_cellType_manual_fine == "Memory_DN2_CXCR3" |
                          gex$cluster_cellType_manual_fine == "Memory_DN2_ITGAX" |
                          gex$cluster_cellType_manual_fine == "Memory_DN4" |
                          gex$cluster_cellType_manual_fine == "Memory_Classical" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_CD1C" |
                          gex$cluster_cellType_manual_fine == "Naive" |
                          gex$cluster_cellType_manual_fine == "Naive_IFN" |
                          gex$cluster_cellType_manual_fine == "Naive_Transitional"), ]


```


## immcantation and scRepertoir output visualization
### VDJ gene usage
####  plot Fig3A, IgH class per cell type main

```{r}

# subset gex to enable cal plotting
# to_remove <- c("pca_1", "pca_2", "scpred", "scpred_projection", 
#                 "harmony_1", "harmony_2", "umap_1", "umap_2_3d")
# gex@reductions <- gex@reductions[!(names(gex@reductions) %in% to_remove)]
# gex[["RNA"]]@counts <- matrix()
# gex[["RNA"]]@data <- matrix()


####---- IgH per cluster
p1 <- DimPlot(gex %>% filter(!is.na(c_gene_10x)), 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              #order = TRUE,
              shuffle = TRUE,
              #order = c("IGHM", "IGHD", "IGHG1",  "IGHG2",  "IGHG3", "IGHG4", "IGHA1", "IGHA2", "IGHE"),
              cols = c(brewer.pal(9, "Set3")),
              na.value = "white",
              raster = TRUE,
              raster.dpi = c(1024, 1024)) + 
  ggtitle(expression(~italic(IGH)~'constant gene')) + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        title = element_text(face = "bold"),
        legend.position = "none")

p2 <- ggplot(gex.meta %>% filter(!is.na(c_gene_10x)), 
             aes(x = cluster_cellType_manual_main,
                 fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") +
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"),
            legend.position = "bottom") +
      scale_fill_brewer(palette="Set3", na.value="grey", name = "",  drop = FALSE) 

p_legend <- get_legend(p2) %>% ggpubr::as_ggplot()

p2 <- p2 + theme(legend.position = "none")

p3 <- plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h", rel_widths = c(1.5, 1))


cell_summary <-  as_tibble(gex.meta) %>% 
  filter(!is.na(c_gene_10x)) %>%
  select(orig.ident, cluster_cellType_manual_main, patient_group, c_gene_10x) %>%
  add_count(orig.ident, cluster_cellType_manual_main, c_gene_10x, .drop = FALSE) %>% 
  distinct() %>%
  group_by(orig.ident, cluster_cellType_manual_main) %>%
  mutate(per =  prop.table(n)) %>%
  drop_na()

p_list <- list()
for (i in as.character(sort(unique(cell_summary$c_gene_10x))) ){
  p_list[[i]] <- list()
  for (j in sort(unique(cell_summary$cluster_cellType_manual_main))) {
    p_list[[i]][[j]] <- ggplot(cell_summary[cell_summary$cluster_cellType_manual_main == j & 
                                             cell_summary$c_gene_10x == i, ], 
                              aes(x = patient_group, y = per,  fill = patient_group)) + 
                     theme_classic() +
                     geom_violin(scale = "width",
                                 draw_quantiles = c(0.25, 0.5, 0.75),
                                 trim = TRUE,
                                 alpha = 0.8) +
                     geom_jitter(size = 0.5) +
                     ylab(paste0("fraction ", i)) + 
                     xlab("") +
                     ggtitle(paste0(j)) +
                     theme(legend.position = "none") +
                     scale_fill_manual(values = scales::hue_pal()(4)) +
                     stat_compare_means(method = "wilcox.test", 
                                        comparisons = list(c("CTRL", "SSA-"), 
                                                           c("CTRL", "SSA+"), 
                                                           c("CTRL", "SSAB")), 
                                        label = "p",
                                        size = 3.5) +
                     scale_y_continuous(labels = scales::percent_format(accuracy = 1L), 
                                        limits = c(0, 1.4))
      }
    }


plot_grid(plotlist = p_list[["IGHM"]])

ggsave2("../results/Fig3A.pdf", 
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "h", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"), 
        width = 20, height = 14, unit = "cm", dpi = 300)

p1 <- DimPlot(gex %>% filter(!is.na(c_gene_10x)), 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.005, ncol = 2, 
              #order = c("IGHM", "IGHD", "IGHG1",  "IGHG2",  "IGHG3", "IGHG4", "IGHA1", "IGHA2", "IGHE"),
              #order = TRUE,
              shuffle = TRUE,
              cols = c(brewer.pal(9, "Set3")),
              na.value = "white",
              raster = FALSE) + 
  ggtitle(expression(~italic(IGH)~'constant gene')) + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        title = element_text(face = "bold"),
        legend.position = "none")
p3 <- plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h", rel_widths = c(1.5, 1))

ggsave2("../results/Fig3A.png", 
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "h", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"), 
        width = 40, height = 28, unit = "cm", dpi = 300)


```


#### plot Fig3B VDJ correlation main 
#### plot SFig8A-C main
 
```{r}

colors_cellTypeMain2 <- colors_cellType[c("memory5", "dn3", "memory3", "naive4", "naive6")]
names(colors_cellTypeMain2) <- sort(unique(gex.meta$cluster_cellType_manual_main))

#VDJ gene usage correlation
VDJ_cor <- function(i, #seurat object meta data containing data for VDJ gene usage
                    correl, #which correlation to use i.e. spearman or pearson...
                    VDJgene, #VDJ gene to plot
                    figNo #output figure name
                    ){
  
  cat(VDJgene, "\n") 
  cat(correl, "\n")
  
  #remove allele information
  i[, VDJgene] <- unlist(lapply(strsplit(as.character(i[, VDJgene]), 
                                         split = "\\*"),
                                function(x) {return(x[1])}))
  
  i$celltype_patient_group <- paste0(i$cluster_cellType_manual_main, 
                                     "__", 
                                     i$patient_group)
  
  tmp <- as.data.frame.matrix(table(list(i[, VDJgene], i$celltype_patient_group)))
  
  tmp <- t(t(tmp)/colSums(tmp))
  
  cor_tmp <- cor(tmp, method = correl)
  print(head(cor_tmp))
  cor_tmp[is.na(cor_tmp)] <- 0

  ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                     `SSA-` = "#7CAE00", 
                                     `SSA+` = "#00BFC4", 
                                     SSAB = "#C77CFF"),
                   #cluster = scales::hue_pal()(15),
                   cell_type = colors_cellTypeMain2)
  
  titles <- data.frame(gexnames = c("v_call", "d_call", "j_call", "c_gene_10x"),
                       outnames = c("IGHV", "IGHD", "IGHJ", "IGH constant"))
  
  for (j in c("png", "pdf")){
    p <- pheatmap::pheatmap(cor_tmp, 
                       annotation_col = data.frame(cell_type = as.character(sub("__.*", 
                                                                          "", 
                                                                          colnames(cor_tmp))),
                                                   patient_group = as.character(sub(".*__", 
                                                                              "",
                                                                              colnames(cor_tmp))),
                                                   row.names = colnames(cor_tmp)),
                       annotation_colors = ann_cols[1:2],
                       #breaks = seq(0, 1, length.out = 100),
                       filename = paste0("../results/", figNo, "_main",".", j),
                       fontsize = 10, 
                       fontsize_row = 4, 
                       fontsize_col = 4,
                       border_color = NA,
                       show_rownames = FALSE,
                       show_colnames = FALSE,
                       clustering_method = "ward.D2",
                       cellwidth = 8,
                       cellheight = 10,
                       main = paste(titles$outnames[titles$gexnames == VDJgene], 
                                     "genes,", correl, "correlation"),
                       color = c(colorRampPalette(c("lightblue1", 
                                                    "lightblue2", 
                                                    "firebrick2"))(99)))
      }
    
      return(p)
    
  }

cor.p.v_call <- VDJ_cor(gex.meta, "pearson", "v_call", "Fig3B")
cor.p.c_call <- VDJ_cor(gex.meta, "pearson", "c_gene_10x", "SFig8A")
cor.p.d_call <- VDJ_cor(gex.meta, "pearson", "d_call", "SFig8B")
cor.p.j_call <- VDJ_cor(gex.meta, "pearson", "j_call", "SFig8C")


```


#### plot Fig3C, IGHV usage per patient group

```{r}

vdj_gene <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = "patient_group", 
                       mode = "gene",
                       fill = TRUE)

vdj_gene$fc <- NA

#calculate FC relative to CTRL
for (i in 1:nrow(vdj_gene)){
  ifelse(vdj_gene$patient_group[i] == "CTRL", 
         vdj_gene$fc[i] <- NA,
         vdj_gene$fc[i] <- log2((vdj_gene$seq_freq[i] * 100 + 1)/ (vdj_gene$seq_freq[vdj_gene$gene == vdj_gene$gene[i] & vdj_gene$patient_group == "CTRL"] * 100  + 1)))
}

# get 45 most used IGHV genes
vdj_gene <- vdj_gene[order(vdj_gene$seq_count, decreasing = TRUE), ]
tmp <- vdj_gene[which(vdj_gene$gene %in% unique(vdj_gene$gene)[1:45]), ]

# plot fc relative to CTRL
p1 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                      y = fc, 
                      colour = patient_group)) +
  theme_classic() +
      ggtitle(expression(~italic(IGHV)~'gene usage per patient group')) +
      theme(axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2, 0, 0, 0), "cm")) +
          ylab("log2FC to CTRL") +
          xlab("") +
          geom_point() +
      geom_hline(yintercept = 0)

# plot gene counts
p2 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0, 0, 0, 0.2), "cm")) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

p_legend <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
  theme_classic() +
  scale_fill_discrete(name = "") +
  geom_col(position = "dodge")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p3 <- plot_grid(p1, p2, nrow = 2, labels = NULL,
                align = "v", axis = "l", rel_heights = c(1, 2.8)) + 
           bgcolor("white") + 
           border("white")


 ggsave2(paste0("../results/Fig3C.png"),
          plot_grid(p3, p_legend, 
                    ncol = 2, 
                    labels = NULL, 
                    align = "v", 
                    rel_widths = c(12, 2)) + 
           bgcolor("white") + 
           border("white"),
          width = 20, height = 12, unit = "cm")
 
  ggsave2(paste0("../results/Fig3C.pdf"),
          plot_grid(p3, p_legend, 
                    ncol = 2, 
                    labels = NULL, 
                    align = "v", 
                    rel_widths = c(12, 2)) + 
           bgcolor("white") + 
           border("white"),
          width = 20, height = 12, unit = "cm")


```


#### plot SFig9A, IgH class per cellTypeMain per patient

```{r}

#gex.meta$c_gene_10x <- factor(gex.meta$c_gene_10x, levels = sort(unique(gex.meta$c_gene_10x)))

tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                   DNEG = "#7CAE00", 
                                   SSA = "#00BFC4", 
                                   SSAB = "#C77CFF"))

colors_samples <- c(rep(ann_cols$patient_group[1], times = 4),
                    rep(ann_cols$patient_group[2], times = 6),
                    rep(ann_cols$patient_group[3], times = 8),
                    rep(ann_cols$patient_group[4], times = 9))


p <- ggplot(gex.meta %>% filter(!is.na(c_gene_10x)),
            aes(x = orig.ident,
                fill = c_gene_10x)) +
      theme_classic() +
      ggtitle("IGH constant") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5,
                                       color = colors_samples), 
            legend.position = "none") +
      ylab("proportion of cells") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      #scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_bar(position = "fill") +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "", drop = FALSE)

## get legend
p_legend <- ggplot(gex.meta %>% filter(!is.na(c_gene_10x)), 
             aes(x = orig.ident,
                 fill = c_gene_10x)) +
      geom_bar(position = "fill") +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "", drop = FALSE)
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()


ggsave2(paste0("../results/SFig9A.png"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.2)) + 
          bgcolor("white") + 
          border("white"),
          width = 34, height = 10, unit = "cm")

ggsave2(paste0("../results/SFig9A.pdf"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.2)) + 
          bgcolor("white") + 
          border("white"),
          width = 34, height = 10, unit = "cm")


```


#### plot SFig9B, IgH class per cellTypeFine per patient

```{r}

#gex.meta$c_gene_10x <- factor(gex.meta$c_gene_10x, levels = sort(unique(gex.meta$c_gene_10x)))

tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
p[[i]] <-  ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ] %>% 
                    filter(!is.na(c_gene_10x)), 
                  aes(x = orig.ident,
                      fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ggtitle(i) +
      scale_x_discrete(drop = FALSE) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = colors_samples),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            legend.position = "none") +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "", drop = FALSE) 

print(table(gex.meta$c_gene_10x[gex.meta$cluster_cellType_manual_fine == i]))

}

p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == unique(gex$cluster_cellType_manual_fine)[1], ] %>% filter(!is.na(c_gene_10x)), 
             aes(x = orig.ident,
                 fill = c_gene_10x)) +
      geom_bar(position = "fill") +
      theme_classic() + 
      theme(legend.position = "bottom", 
            legend.box = "horizontal", 
            panel.border = NULL) +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "", drop = FALSE)

p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot() 

y_label <- ggdraw() + 
  draw_label("proportion of cells", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

p1 <- plot_grid(plotlist = p, ncol = 3, labels = NULL, align = "hv")  + 
  bgcolor("white") + 
  border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/SFig9B.png"), 
          plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) +
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

ggsave2(paste0("../results/SFig9B.pdf"), 
          plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) +
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")


```


#### plot SFig10, allele usage for differentially used IGHV genes IGHV1-69

```{r}

vdj_allele <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = "patient_group", 
                       mode = "allele",
                       fill = TRUE)

#IGHV1-69
vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene), ] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]

p1 <- ggplot(tmp1, aes(x = reorder(gene, -seq_freq), 
                       y = seq_freq, 
                       fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 

            legend.position = "none"
            ) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-69 alleles") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L), name = "Patient Group") +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

ggsave2(paste0("../results/SFig10.png"), 
          plot_grid(p1, ncol = 1, labels = NULL, align = "h")+ 
          bgcolor("white") + 
          border("white"),
          width = 15, height = 10, unit = "cm")

ggsave2(paste0("../results/SFig10.pdf"), 
          plot_grid(p1, ncol = 1, labels = NULL, align = "h")+ 
          bgcolor("white") + 
          border("white"),
          width = 15, height = 10, unit = "cm")


```


#### plot SFig11A-B, top IGHD and IGHJ genes per patient subgroup

```{r}

get_topVDJ <- function(i, #seurat object
                       VDJgene = "tmp" #VDJ gene to plot
                       ) {
  
  #remove allele information
  vh_col <- unlist(lapply(strsplit(as.character(i@meta.data[, VDJgene]), 
                                                   split = "\\*"),
                                          function(x) {return(x[1])})) 
  
  vh <- as.data.frame.matrix(table(vh_col, i$patient_group))
  vh <- vh[order(rowSums(vh), decreasing = TRUE), ]
  
  #export counts per patient group
  #write.csv(vh, paste0("../results/VDJ_geneUsage_", VDJgene, "_genesPerPatientGroup.csv")) 
  
  #get percentages of repertoir
  vh <-  as.data.frame(t(t(vh) / colSums(vh))) 
  vh$gene <- rownames(vh)
  
  nvh <- ifelse(nrow(vh) <= 50, nrow(vh), 50)
  
  titles <- data.frame(gexnames = c("v_call", "d_call", "j_call", "c_gene_10x"),
                       outnames = c("IGHV", "IGHD", "IGHJ", "IGH constant"))
  
  p <- as_tibble(vh[1:nvh, ]) %>% pivot_longer(
       cols = 1:4,
       names_to = "patient_group",
       names_prefix = "",
       values_to = "count",
       values_drop_na = TRUE) %>%  
          ggplot(aes(x = reorder(gene, -count), y = count, fill = patient_group)) +
              geom_col(position = "dodge") +
              theme_classic() +
              xlab("") +
              ylab("Percent of repertoire") +
              scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
              scale_fill_discrete(name = "Patient group") +
              ggtitle(paste0(titles$outnames[titles$gexnames == VDJgene], " genes")) +
              theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                    axis.text.y = element_text(size = 6),
                    strip.text.y = element_text(size = 8)) +
              facet_grid(rows = vars(patient_group))

    ggsave2(paste0("../results/VDJ_geneUsage_", VDJgene, "_geneUsage_perPatientGroup_2.png"), 
          p,
          width = 30, height = 14, unit = "cm")
    
    return(p)
  
   }

pv <- get_topVDJ(gex, "v_call")
pd <- get_topVDJ(gex, "d_call")
pj <- get_topVDJ(gex, "j_call")
pc <- get_topVDJ(gex, "c_gene_10x")

ggsave2(paste0("../results/SFig11A-B.png"), 
          plot_grid(pd, pj, align = "hv", labels = NULL, ncol = 1) +
          bgcolor("white") + 
          border("white"),
          width = 15, height = 18, unit = "cm", dpi = 300)


ggsave2(paste0("../results/SFig11A-B.pdf"), 
          plot_grid(pd, pj, align = "hv", labels = NULL, ncol = 1) +
          bgcolor("white") + 
          border("white"),
          width = 15, height = 18, unit = "cm", dpi = 300)


```


## clonotype analysis
### diversity
#### clonotype diversity

```{r}
####---------------------------------------------------------------------
####---- clonal diversity
####---------------------------------------------------------------------

####---- abundance estimation
patient_group_estimateAbundance <-
  alakazam::estimateAbundance(
    gex.meta,
    group = "patient_group",
    clone = "cluster_no_airr",
    ci = 0.95,
    nboot = 50
  )

saveRDS(patient_group_estimateAbundance, "../results/VDJ_clono_patient_group_estimateAbundance.rds")
#patient_group_estimateAbundance <- readRDS("../results/VDJ_clono_patient_group_estimateAbundance.rds")

#Error: vector memory exhausted (limit reached?) 
#nboot = 100 works locally, nboot = 200 does not

####---- per patient group
patient_group_alphaDiversity <- alakazam::alphaDiversity(
  gex.meta,
  group = "patient_group",
  clone = "cluster_no_airr",
  min_q = 0,
  max_q = 4,
  step_q = 0.1,
  ci = 0.95,
  nboot = 50
)

saveRDS(patient_group_alphaDiversity, "../results/VDJ_clono_patient_group_alphaDiversity.rds")
#patient_group_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_alphaDiversity.rds")


```


#### plot SFig17

```{r}

p1 <- plotAbundanceCurve(patient_group_estimateAbundance, 
                         colors = c(CTRL = "#F8766D", 
                                    `SSA-` = "#7CAE00", 
                                    `SSA+` = "#00BFC4", 
                                     SSAB = "#C77CFF")) + 
  theme_classic() + 
  theme(legend.position = "none",
        legend.title = element_blank()) 

p2.1 <- plotDiversityCurve(patient_group_alphaDiversity,
                           colors = c(CTRL = "#F8766D", 
                                    `SSA-` = "#7CAE00", 
                                    `SSA+` = "#00BFC4", 
                                     SSAB = "#C77CFF")) + 
  theme_classic() + 
  theme(legend.title = element_blank())

p_legend <- get_legend(p2.1) %>% ggpubr::as_ggplot()

p2 <- p2.1 + theme(legend.position = "none")


ggsave2(paste0("../results/SFig17.png"),
          plot_grid(p1, p2, p_legend,
                    nrow = 1, labels = NULL, 
                    align = "vh", axis = "l", rel_widths = c(1, 1, 0.4)) + 
          bgcolor("white") + 
          border("white"),
          width = 22, height = 12, unit = "cm")

ggsave2(paste0("../results/SFig17.pdf"),
          plot_grid(p1, p2, p_legend,
                    nrow = 1, labels = NULL, 
                    align = "vh", axis = "l", rel_widths = c(1, 1, 0.4)) + 
          bgcolor("white") + 
          border("white"),
          width = 22, height = 12, unit = "cm")

#rm(patient_group_estimateAbundance); gc()
#rm(patient_group_alphaDiversity); gc()


```


### clonal expansion
#### plot Fig4A, cloneType per cell type fine based on compairr data (cellTypeMain)

```{r}

gex$cloneType_airr <- gsub("<=", "≤", gex$cloneType_airr)

p1 <- DimPlot(gex %>% filter(!is.na(cloneType_airr)), 
              group.by = "cloneType_airr", 
              split.by = "patient_group",
              reduction = "umap_2", pt.size = 0.6, ncol = 2,
              order = c("Single (0 < X ≤ 1)",
                        "Small (1 < X ≤ 5)",
                        "Medium (5 < X ≤ 20)",
                        "Large (20 < X ≤ 100)"), 
              shuffle = TRUE,
              cols = c(brewer.pal(4, "Dark2")),
              na.value = "white",
              raster = TRUE,
              raster.dpi = c(1024, 1024)) +
  ggtitle("clonotype") +  
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none")

gex.meta$cloneType_airr <- gsub("<=", "≤", gex.meta$cloneType_airr)
gex.meta$cloneType_airr <- factor(gex.meta$cloneType_airr, levels = c("Large (20 < X ≤ 100)",
                                                                      "Medium (5 < X ≤ 20)",
                                                                      "Small (1 < X ≤ 5)",
                                                                      "Single (0 < X ≤ 1)"))

p2 <- ggplot(gex.meta %>% filter(!is.na(cloneType_airr)), 
             aes(x = cluster_cellType_manual_main, 
                 fill = cloneType_airr)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") + 
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"),
            legend.position = "bottom") +
      scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "")


p_legend <- get_legend(p2) %>% ggpubr::as_ggplot()

p2 <- p2 + theme(legend.position = "none")

p3 <- plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h", rel_widths = c(1.5, 1))

ggsave2("../results/Fig4A.pdf", 
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "h", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"), 
        width = 20, height = 14, unit = "cm", dpi = 300)

p1 <- DimPlot(gex %>% filter(!is.na(cloneType_airr)), 
              group.by = "cloneType_airr", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.02, ncol = 2,
              order = c("Single (0 < X ≤ 1)",
                        "Small (1 < X ≤ 5)",
                        "Medium (5 < X ≤ 20)",
                        "Large (20 < X ≤ 100)"),
              shuffle = TRUE,
              cols = c(brewer.pal(4, "Dark2")),
              na.value = "white",
              raster = FALSE) +
  ggtitle("clonotype") +  
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none")

p3 <- plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h", rel_widths = c(1.5, 1))

ggsave2("../results/Fig4A.png", 
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "h", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"), 
        width = 40, height = 28, unit = "cm", dpi = 300)


```


### most prevelent clonotypes across samples
#### plot Fig4B, common clonotypes across samples

```{r}

####---------------------------------------------------------------------
####---- make pie charts
####---------------------------------------------------------------------

imm2 <- gex.meta[, c("cluster_no_airr", "orig.ident")]
imm3 <- table(imm2)
imm4 <- data.frame(table(rowSums(imm3 > 0)))
colnames(imm4) <- c("clonocount", "cellcount")
imm4$perc <- round(imm4$cellcount / sum(imm4$cellcount) * 100, digits = 5)


####---- pie1 - all clonotypes
p1 <- ggplot(imm4, aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1) +
  geom_text(aes(label = c(paste0(round(imm4$perc[1], digits = 2), "%"), 
                rep("", times = nrow(imm4) - 1))), 
            position = position_stack(vjust = 0.5, reverse = TRUE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(format(sum(imm4$cellcount), big.mark = " "), " clonotypes")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5))


####---- get legend
p_legend <- ggplot(imm4, aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0)+
  theme_void() +
  theme(legend.position = "bottom", legend.box = "horizontal") +
  scale_fill_discrete(name = "Samples per \nclonotype")

p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

####---- get colors
g <- ggplot_build(p1)
c.p1 <- unique(g$data[[1]]["fill"])


####---- pie2  for all but unique clonotypes
p2 <- ggplot(imm4[2:nrow(imm4), ], 
             aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1)+
  geom_text(aes(label = c(paste0(round(imm4$perc[2:5], digits = 3), "%"), 
                rep("", times = nrow(imm4) - 5))), 
            position = position_stack(vjust = 0.75, reverse = FALSE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(format(sum(imm4$cellcount[2:nrow(imm4)]), big.mark = " "), " clonotypes\npresent in ≥ 2 samples")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5)
        ) +
  scale_fill_manual(values = c.p1$fill[2:nrow(c.p1)])


####---- pie3  clonotypes present in ≥ 4 samples
p3 <- ggplot(imm4[4:nrow(imm4), ], 
             aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1) +
  geom_text(aes(label = c(paste0(round(imm4$perc[4:6], digits = 3)
                                 , "%"
                                 ),
                rep("", times = nrow(imm4) - 6))),
            position = position_stack(vjust = 0.6,
                                      reverse = FALSE),
            size = 3.2) +
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(sum(imm4$cellcount[4:nrow(imm4)]), 
                 " clonotypes\npresent in ≥ 4 samples")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5)) +
  scale_fill_manual(values = c.p1$fill[4:nrow(c.p1)])

p4 <- plot_grid(p1, p2, p3, nrow = 1, labels = NULL,
                align = "vh", axis = "l") + 
  bgcolor("white") + 
  border("white")

####---- save pies
ggsave2(paste0("../results/Fig4B.png"),
          plot_grid(p4, p_legend, 
                    ncol = 1, 
                    labels = NULL, 
                    align = "v", 
                    rel_heights = c(5, 2)) + 
          bgcolor("white") + 
          border("white"),
          width = 20, height = 12, unit = "cm")

ggsave2(paste0("../results/Fig4B.pdf"),
          plot_grid(p4, p_legend, 
                    ncol = 1, 
                    labels = NULL, 
                    align = "v", 
                    rel_heights = c(5, 2)) + 
          bgcolor("white") + 
          border("white"),
          width = 20, height = 12, unit = "cm")


```


#### get clonotypes occuring in ≥ 2 samples

```{R}

pheno_2 <- unique(gex.meta[, c("orig.ident", "patient_group")])
tmp <- as.data.frame.matrix(table(gex.meta$cluster_no_airr,
                                  gex.meta$orig.ident))

#get clonotypes present in ??? 2 samples
tmp <- tmp[rowSums(tmp > 0) > 1, ]

#export info for 1640 clonotypes
 # outfile <- gex.meta[gex.meta$cluster_no_airr %in% rownames(tmp), c("barcode", "orig.ident", "cluster_cellType_manual_fine", "cluster_cellType_manual_main", "cluster_no_airr", "v_call", "d_call", "j_call", "c_call", "IGHV", "c_gene_10x", "junction", "junction_aa", "cdr3", "patient_group", "mu_load", "mu_freq_total")]
 # write.csv(outfile, "../results/1799_clonotypes.csv", sep = ",", col.names = TRUE, row.names = FALSE, quote = TRUE)

#tmp <- read.csv("../results/1640_clonotypes.csv")

#make the clonotype data frame per sample binary
tmp <- (tmp > 0) * 1

#get number of samples for each clonotype within each subgroup
pheno3 <- as.data.frame.matrix(table(pheno_2$orig.ident, pheno_2$patient_group))
tmp3 <- as.data.frame(as.matrix(tmp) %*% as.matrix(pheno3))


#normalize to number of samples in subgroup
tmp3 <- t(t(tmp3)/c(table(pheno_2$patient_group)))
h.3 <- hclust(dist(tmp3, method = "binary"), method = "complete")
#saveRDS(h.3, "../results/h.3.rds")


#plot(as.dendrogram(h.3)); abline(h = 0.2, col = "red")
clr <- cutree(h.3, h = 0.2) # number of row clusters #15

tmp4 <- rowsum(tmp, clr) #number of clonotypes per sample in each row clade
h.4 <- hclust(dist(t(tmp4), method = "binary"), method = "complete")
#plot(as.dendrogram(h.4)); abline(h = 0.2, col ="red") 
clc <- cutree(h.4, h = 0.2) #number of colum clusters #4

ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                   DNEG = "#7CAE00", 
                                   SSA = "#00BFC4", 
                                   SSAB = "#C77CFF"),
                 cluster = scales::hue_pal()(15))

p1 <- pheatmap(tmp,
               cluster_cols = h.4,
               cluster_rows = h.3,
               #cluster_cols = FALSE,
               #border_color = NA,
               cutree_rows = length(unique(clr)),
               cutree_cols = length(unique(clc)),
               annotation_col = data.frame(patient_group = 
                         factor(pheno_2$patient_group[match(colnames(tmp),
                                                  pheno_2$orig.ident)], 
                                levels = c("CTRL", "DNEG", "SSA", "SSAB")), 
                                           row.names = colnames(tmp)),
               annotation_row = data.frame(cluster = factor(clr), row.names = names(clr)),
               annotation_names_row = TRUE,
               annotation_colors = ann_cols[1],
               color = c("gray95", colorRampPalette(c("gray70", 
                                                      "orange3", 
                                                      "firebrick"))(90)),
               legend = FALSE,
               fontsize_col = 6,
               treeheight_row = 40,
               treeheight_col = 30,
               main = paste0(nrow(tmp), " clonotypes occuring in ??? 2 samples"),
               filename = "../results/Fig5D_II.png",
               show_rownames = FALSE)


```


#### plot Fig4C, UpSetR of clonotypes (1640)

```{R}

pheno_2 <- unique(gex.meta[, c("orig.ident", "patient_group")])
tmp <- as.data.frame.matrix(table(gex.meta$cluster_no_airr,
                                  gex.meta$orig.ident))

#get clonotypes present in at least 2 samples
tmp <- tmp[rowSums(tmp > 0) > 1, ]

tmp2 <- as.data.frame(t(rowsum(t(tmp), group = pheno$patient_group[1:27])))
tmp2 <- as.data.frame((tmp2 > 0) * 1)
tmp2$ident <- as.character(rownames(tmp2))

venn.list <- list()[1:4]
names(venn.list) <- colnames(tmp2)[1:4]
for (i in colnames(tmp2)[1:4]){
  for (j in 1:nrow(tmp2))
  venn.list[[i]][j] <- ifelse(tmp2[j, i] == 1, tmp2$ident[j], NA)
}
venn.list <- lapply(venn.list, function(x){x <- na.omit(x)})

png("../results/Fig4C.png", width = 12, height = 11, units = "cm", res = 200)
UpSetR::upset(fromList(venn.list), order.by = "freq")
dev.off()

pdf("../results/Fig4C.pdf", width = 5, height = 4.5)
UpSetR::upset(fromList(venn.list), order.by = "freq")
dev.off()

```


### CDR3 sequence similarities with published data

```{r}

####---------
#1)
#Corsiero_et_al_2014
#Accumulation of self-reactive naive and memory B cell reveals sequential defects in B cell tolerance checkpoints in Sjögren's syndrome
#https://pubmed.ncbi.nlm.nih.gov/25535746/

corsiero_2 <- read_excel("../suppl/Corsiero_et_al_2014/STable2.xlsx")
corsiero_4 <- read_excel("../suppl/Corsiero_et_al_2014/STable4.xlsx")
corsiero_5 <- read_excel("../suppl/Corsiero_et_al_2014/STable5.xlsx")

corsiero <- as.data.frame(rbind(corsiero_2, corsiero_4, corsiero_5))

corsiero_p <- data.frame(Patient = str_sort(unique(corsiero$Patient), numeric = T))
corsiero_p$age_onset <- c(59, 54, 39, 59, 64, 41, 56)
corsiero_p$ANA <- c(rep(1, times = nrow(corsiero_p)))
corsiero_p$RF <- NA
corsiero_p$SSA <- c(1,1,1,1,0,1,1)
corsiero_p$SSB <- c(0,1,0,0,0,1,1)

corsiero <- left_join(corsiero, corsiero_p, by = "Patient")

corsiero$dataset <- "Corsiero"
corsiero$gender <- "F"

#saveRDS(corsiero, "../suppl/Corsiero_et_al_2014/corsiero.RDS")
corsiero <- readRDS("../suppl/Corsiero_et_al_2014/corsiero.RDS")


####---------
#2)
#Visser et al 2020
#Repertoire Analysis of B-Cells Located in Striated Ducts of Salivary Glands of Patients With Sj??gren's Syndrome
#https://pubmed.ncbi.nlm.nih.gov/32760405/
visser <- as.data.frame(read_excel("../suppl/Visser_et_al_2020/Table2.xlsx"))

#remove allele info
visser$`VH-GENE` <- unlist(lapply(strsplit(visser$`VH-GENE`, split = "\\*"), function(x) {return(x[1])}))
visser$`JH-GENE` <- unlist(lapply(strsplit(visser$`JH-GENE`, split = "\\*"), function(x) {return(x[1])}))
visser$`DH-GENE` <- unlist(lapply(strsplit(visser$`DH-GENE`, split = "\\*"), function(x) {return(x[1])}))

visser_p <- data.frame(PATIENT = str_sort(unique(visser$PATIENT), numeric = T))
visser_p$age_onset <- NA
visser_p$ANA <- c(1,0,1,0,1)
visser_p$RF <-  c(1,1,1,0,1)
visser_p$SSA <- c(1,1,1,0,1)
visser_p$SSB <- c(1,0,1,0,NA)

visser <- left_join(visser, visser_p, by = "PATIENT")

visser$dataset <- "Visser"
visser$gender <- "F"

#saveRDS(visser, "../suppl/Visser_et_al_2020/visser.RDS")
visser <- readRDS("../suppl/Visser_et_al_2020/visser.RDS")


####--------- 
#3)
#Wang_et_al_2018
#Molecular Profiling and Clonal Tracking of Secreted Rheumatoid Factors in Primary Sj??gren's Syndrome
#https://pubmed-ncbi-nlm-nih-gov.ezproxy.its.uu.se/29697211/

wang <- as.data.frame(read_excel("../suppl/Wang_et_al_2018/Table3.xlsx"))

wang_p <- data.frame(Patient = str_sort(unique(wang$Patient), numeric = T))
wang_p$age_onset <- c(19, 54, 48)
wang_p$ANA <- NA
wang_p$RF <- 1
wang_p$SSA <- 1
wang_p$SSB <- 1
wang_p$cell_type <- NA

wang <- left_join(wang, wang_p, by = "Patient")

wang$dataset <- "Wang"
wang$gender <- "F"

#saveRDS(wang, "../suppl/Wang_et_al_2018/wang.RDS")
wang <- readRDS("../suppl/Wang_et_al_2018/wang.RDS")


####--------- concatenate external data sets
corsiero <- corsiero[, c(1, 5:15)] %>% distinct()
visser <- visser[, c(1, 4, 7:16)] %>% distinct()
wang <- wang[, c(1,2:10,12:13)] %>% distinct()

colnames(visser) <- colnames(wang)
colnames(corsiero) <- colnames(wang)

length(corsiero$CDR3) #[1] 354
length(unique(corsiero$CDR3)) #[1] 347
length(visser$CDR3) #[1] 105
length(unique(visser$CDR3)) #[1] 98
length(wang$CDR3) #[1] 13
length(unique(wang$CDR3)) #[1] 12

CDR3_data <- rbind(corsiero, visser, wang); head(CDR3_data); nrow(CDR3_data) #[1] 472

#saveRDS(CDR3_data, "../results/CDR3_data_literature.rds")
CDR3_data <- readRDS("../results/CDR3_data_literature.rds")


```


#### plot Fig4D

```{r}

CDR3_data <- readRDS("../results/CDR3_data_literature.rds")
CDR3_data <- CDR3_data %>% distinct(CDR3, dataset, Patient)
CDR3_data <- CDR3_data[, c("CDR3", "dataset", "Patient")]
colnames(CDR3_data) <- c("CDR3", "group", "patient")

tmp <- as.data.frame.matrix(table(gex.meta$cluster_no_airr,
                                  gex.meta$orig.ident))

#get clonotypes present in ≥ 2 samples
tmp <- tmp[rowSums(tmp > 0) > 1, ]
tmp2 <- as.data.frame(t(rowsum(t(tmp), group = pheno$patient_group[1:27])))
tmp2 <- as.data.frame((tmp2 > 0) * 1)
tmp2$ident <- as.character(rownames(tmp2))

m <- gex.meta[(gex.meta$cluster_no_airr %in% rownames(tmp2)) & 
                       !gex.meta$patient_group == "CTRL", 
                     c("junction_aa", "patient_group", "orig.ident")] %>% distinct()
colnames(m) <- c("CDR3", "group", "patient")
nrow(m) #[1] 4169

m <- rbind(m, CDR3_data)
m$rowname <- paste(m$CDR3, m$group, paste(m$patient, sep = "_"), sep = "_")

res <- stringdist::stringdistmatrix(m$CDR3, m$CDR3,  method = "lv")
res <- as.data.frame(Matrix(res, sparse = T, dimnames = list(m$rowname, m$rowname)))

ann_cols <- list(group = c(Corsiero = brewer.pal(9, "YlOrRd")[2],
                           Visser = brewer.pal(9, "YlOrRd")[5],  
                           Wang = brewer.pal(9, "YlOrRd")[7], 
                           `SSA-`= "#7CAE00",
                           `SSA+` = "#00BFC4", 
                           SSAB = "#C77CFF"))

#annotation <- data.frame(group = factor(m$group), row.names = m$rowname)

for (i in c("png")){
  
  print(i)
  
  h1 <- pheatmap(res, 
           show_rownames = FALSE, 
           show_colnames = FALSE, 
           filename = paste0("../results/Fig4D_tmp.", i),
           annotation_col = data.frame(group = factor(m$group), row.names = m$rowname),
           annotation_row = data.frame(group = factor(m$group), row.names = m$rowname),
           annotation_colors = ann_cols[1], 
           legend = TRUE,
           use_raster = TRUE,
           clustering_method = "ward.D2")
  
  saveRDS(h1, paste0("../results/CDR3_lv_distance_heatmap_all.rds"))
  
  p_list <- list()
  for (j in unique(m$group)[c(2, 3, 1, 4:6)]){
    
    print(j)
    print(dim(res[(m$group %in% j), ]))
    
    p_list[[j]] <- pheatmap(res[, which(m$group %in% j)],
            cluster_rows = h1$tree_row,
            cluster_cols = TRUE,
            annotation_row = data.frame(group = factor(m$group), row.names = m$rowname),
            annotation_col = data.frame(group = factor(m$group), row.names = m$rowname),
            annotation_colors = ann_cols[1],
            filename = paste0("../results/Fig4D", "_tmp_", j, ".", i),
            clustering_method = "ward.D2",
            legend = TRUE,
            show_rownames = FALSE, 
            show_colnames = FALSE, 
            main = j )
    
  }
  
  saveRDS(p_list, paste0("../results/CDR3_lv_distance_heatmap_byGroup.rds"))
  
}

#p_list <- readRDS("../results/CDR3_lv_distance_heatmap_byGroup.rds")

corder <- c(p_list$`SSA-`$tree_col$labels[p_list$`SSA-`$tree_col$order],
  p_list$`SSA+`$tree_col$labels[p_list$`SSA+`$tree_col$order],
  p_list$SSAB$tree_col$labels[p_list$SSAB$tree_col$order],
  p_list$Corsiero$tree_col$labels[p_list$Corsiero$tree_col$order],
  p_list$Visser$tree_col$labels[p_list$Visser$tree_col$order],
  p_list$Wang$tree_col$labels[p_list$Wang$tree_col$order])
  

for (i in c("png", "pdf")) {

  print(i)
  
  h1 <- pheatmap(res[corder, corder], 
           show_rownames = FALSE, 
           show_colnames = FALSE, 
           filename = paste0("../results/Fig4D.", i),
           annotation_col = data.frame(group = factor(m$group), row.names = m$rowname),
           annotation_row = data.frame(group = factor(m$group), row.names = m$rowname),
           annotation_colors = ann_cols[1], 
           legend = TRUE,
           use_raster = TRUE,
           cluster_rows = FALSE,
           cluster_cols = FALSE
           )
  
}

```


#### plot Fig4E-F

```{r}

cl1 <- paste0("SSA-", cutree(p_list$`SSA-`$tree_col, k = 1))
cl2 <- paste0("SSA+", cutree(p_list$`SSA+`$tree_col, k = 1 ))
cl3 <- paste0("SSAB", cutree(p_list$SSAB$tree_col, k = 1 ))
cl4 <- paste0("Corsiero", cutree(p_list$Corsiero$tree_col, k = 1 ))
cl5 <- paste0("Visser",cutree(p_list$Visser$tree_col, k = 1 ))
cl6 <- paste0("Wang", cutree(p_list$Wang$tree_col, k = 1 ))

cl_all <- factor(c(cl1, cl2, cl3, cl4, cl5, cl6))

o <- c(p_list$`SSA-`$tree_col$labels,
       p_list$`SSA+`$tree_col$labels,
       p_list$`SSAB`$tree_col$labels,
       p_list$Corsiero$tree_col$labels,
       p_list$Visser$tree_col$labels,
       p_list$Wang$tree_col$labels)

mean_mx <- rowsum(res[o, o], cl_all)/c(table(cl_all))
mean_mx <- rowsum(t(mean_mx), cl_all)/c(table(cl_all))
mean_mx <- mean_mx[ c(2:4, 1, 5:6), c(2:4, 1, 5:6)]
colnames(mean_mx) <- sort(unique(m$group))
round(mean_mx, digits =1)

res_tmp <- as.matrix((res > 6) * 1)
perc_mx <- rowsum(res_tmp[o, o], cl_all)/c(table(cl_all))
perc_mx <- rowsum(t(perc_mx), cl_all)/c(table(cl_all))*100
perc_mx <- perc_mx[ c(2:4, 1, 5:6), c(2:4, 1, 5:6)]
colnames(perc_mx) <- sort(unique(m$group))


# minimum lv distances between groups
dist_mx <- data.frame(matrix(ncol = length(sort(unique(m$group))), 
                             nrow = length(sort(unique(m$group)))), 
                      row.names = sort(unique(m$group)))
colnames(dist_mx) <- sort(unique(m$group))

count_mx <- data.frame(matrix(ncol = length(sort(unique(m$group))),
                              nrow = length(sort(unique(m$group)))), 
                       row.names = sort(unique(m$group)))
colnames(count_mx) <- sort(unique(m$group))

for (i in sort(unique(m$group))) {
  
  print(i)
  
  for (j in sort(unique(m$group))) {

  print(j)
    
  res_tmp <- res[m$group %in% j, m$group %in% i]
  
  if(ncol(res_tmp) == nrow(res_tmp)){ 
    
    res_tmp <- as.matrix(res_tmp)
    diag(res_tmp) <- NA
    res_tmp <- as.data.frame(res_tmp)
    
    res_min <-  which(res_tmp == min(res_tmp, na.rm = TRUE), arr.ind = TRUE)
    
    dist_mx[j, i] <- min(res_tmp[res_min])
    count_mx[j, i] <- length(res_tmp[res_min])
    
    } else {res_min <-  which(res_tmp == min(res_tmp), arr.ind = TRUE)
            dist_mx[j, i] <- min(res_tmp[res_min])
            count_mx[j, i] <- length(res_tmp[res_min])
    }
  
  }
  
}

  
p_df <- dist_mx %>% rstatix::cor_gather() 
p_df_count_mx  <- count_mx %>% rstatix::cor_gather() 
p_df_perc_mx  <- perc_mx %>% rstatix::cor_gather() 
p_df_mean_mx  <- mean_mx %>% rstatix::cor_gather() 
  
p_df <- cbind(p_df, p_df_count_mx[,3], p_df_perc_mx[,3], p_df_mean_mx[,3])

colnames(p_df)[3:6] <- c("min_dist","count", "perc", "mean_dist")
p_df$var1 <- factor(p_df$var1, levels = c("SSA-", "SSA+", "SSAB", "Corsiero", "Visser", "Wang"))
p_df$var2 <- factor(p_df$var2, levels = c("SSA-", "SSA+", "SSAB", "Corsiero", "Visser", "Wang"))

p1 <- ggplot(p_df) + 
      geom_point(aes(x = var1, 
                     y = var2, 
                     color = min_dist, 
                     size = perc)) + 
      scale_colour_gradient(low = "blue", high = "red", name = "dist") +
      scale_size(range = c(1, 8)) +
      theme_classic() + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            legend.key.size = unit(0.4, "cm"),
            legend.spacing.y = unit(0, 'cm')) +
      guides(color = guide_colourbar(order = 1),
             size =  guide_legend(order = 2)) +
  ggtitle("min edit distance")

p2 <- ggplot(p_df) + 
      geom_point(aes(x = var1, 
                     y = var2, 
                     color = mean_dist, 
                     size = perc)) + 
      scale_colour_gradient(low = "blue", high = "red", name = "dist") +
      scale_size(range = c(1, 8)) +
      theme_classic() + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            legend.key.size = unit(0.4, "cm"),
            legend.spacing.y = unit(0, 'cm')) +
      guides(color = guide_colourbar(order = 1),
             size =  guide_legend(order = 2)) +
  ggtitle("mean edit distance")


ggsave2(paste0("../results/Fig4E.png"), 
          plot_grid(p1, 
                    ncol = 1, 
                    labels = NULL) + 
          bgcolor("white") + 
          border("white"),
          width = 8, height = 6.9, unit = "cm", dpi = 300)

ggsave2(paste0("../results/Fig4E.pdf"), 
          plot_grid(p1, 
                    ncol = 1, 
                    labels = NULL) + 
          bgcolor("white") + 
          border("white"),
          width = 8, height = 6.9, unit = "cm", dpi = 300)

ggsave2(paste0("../results/Fig4F.png"), 
          plot_grid(p2, 
                    ncol = 1, 
                    labels = NULL) + 
          bgcolor("white") + 
          border("white"),
          width = 8, height = 6.9, unit = "cm", dpi = 200)

ggsave2(paste0("../results/Fig4F.pdf"), 
          plot_grid(p2, 
                    ncol = 1, 
                    labels = NULL) + 
          bgcolor("white") + 
          border("white"),
          width = 8, height = 6.9, unit = "cm", dpi = 200)


```




#### plot SFig18A, clonotypes per patient per cellTypeMain

```{r}
tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- ggplot(gex.meta %>% filter(!is.na(cloneType_airr)),
            aes(x = orig.ident,
                fill = cloneType_airr)) +
      theme_classic() +
      ggtitle("clonotypes") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5,
                                       color = colors_samples), 
            legend.position = "none") +
      ylab("proportion of cells") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      #scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_bar(position = "fill") +
      scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "c gene", drop = TRUE)

## get legend
p_legend <- ggplot(gex.meta %>% filter(!is.na(cloneType_airr)), 
             aes(x = orig.ident,
                 fill = cloneType_airr)) +
      geom_bar(position = "fill") +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "", drop = TRUE)
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

ggsave2(paste0("../results/SFig18A.png"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.2)) + 
          bgcolor("white") + 
          border("white"),
          width = 34, height = 10, unit = "cm")

ggsave2(paste0("../results/SFig18A.pdf"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.2)) + 
          bgcolor("white") + 
          border("white"),
          width = 34, height = 10, unit = "cm")


```


#### plot SFig18B, clonotypes per patient per cellTypeFine

```{r}
tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
  p[[i]] <-  ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ] %>% 
                      filter(!is.na(cloneType_airr)), 
                    aes(x = orig.ident,
                        fill = cloneType_airr)) +
        theme_classic() +
        geom_bar(position = "fill") + 
        ggtitle(i) +
        scale_x_discrete(drop = FALSE) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = colors_samples),
              axis.title.x = element_blank(),
              axis.title.y = element_blank(),
              legend.position = "none") +
        scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "", drop = TRUE) 
  
  print(table(gex.meta$c_gene_10x[gex.meta$cluster_cellType_manual_fine == i]))

}

p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == unique(gex$cluster_cellType_manual_fine)[1], ] %>% filter(!is.na(cloneType_airr)), 
             aes(x = orig.ident,
                 fill = cloneType_airr)) +
      geom_bar(position = "fill") +
      theme_classic() + 
      theme(legend.position = "bottom", 
            legend.box = "horizontal", 
            panel.border = NULL) +
      scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "", drop = TRUE)

p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot() 

y_label <- ggdraw() + 
  draw_label("proportion of cells", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

p1 <- plot_grid(plotlist = p, ncol = 3, labels = NULL, align = "hv")  + 
  bgcolor("white") + 
  border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/SFig18B.png"), 
          plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) +
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

ggsave2(paste0("../results/SFig18B.pdf"), 
          plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) +
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")


```


## mutation status
####  plot Fig3D, mutation load per cell type main

```{r}

p1 <- DimPlot(gex %>% filter(!is.na(mu_load)), 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              cols = c(brewer.pal(4, "Dark2")),
              #order = c(NA, "unmutated", "low", "med", "high"),
              shuffle = TRUE,
              na.value = "white",
              raster = TRUE,
              raster.dpi = c(1024, 1024)) +
  ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none")


p2 <- ggplot(gex.meta %>% filter(!is.na(mu_load)), 
             aes(x = cluster_cellType_manual_main, 
                 fill = mu_load)) +
          theme_classic() +
          geom_bar(position = "fill") + 
          ylab("proportion of cells") +
          xlab("") +
          facet_grid(rows = vars(patient_group)) + 
          theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = "bold"),
                legend.position = "bottom") +
          scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "Mutation load")

p_legend <- get_legend(p2) %>% ggpubr::as_ggplot()

p2 <- p2 + theme(legend.position = "none")

p3 <- plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h", rel_widths = c(1.5, 1))


# !!! Plot mutation load, fraction unmutated per celltypeMain

ggsave2("../results/Fig3D.pdf", 
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "h", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"), 
        width = 20, height = 14, unit = "cm")

p1 <- DimPlot(gex %>% filter(!is.na(mu_load)), 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.02, ncol = 2, 
              cols = c(brewer.pal(4, "Dark2")),
              #order = c(NA, "unmutated", "low", "med", "high"),
              shuffle = TRUE,
              na.value = "white",
              raster = FALSE) +
  ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none")

p3 <- plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h", rel_widths = c(1.5, 1))

ggsave2("../results/Fig3D.png", 
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "h", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"), 
        width = 40, height = 28, unit = "cm", dpi = 300)


```


#### plot Fig3E, total VDJ mutation frequency - celltypeMain - patient_group

```{r}

meanlist <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_main))) {
  gex.meta[gex.meta$cluster_cellType_manual_main == i, ] %>% 
    
  }


p_list1 <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_main))) {
p_list1[[i]] <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_main == i, ], 
            aes(x = patient_group, 
                y = mu_freq_total, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle(i) +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5),
            plot.title = element_text(size = 12),
            legend.position = "bottom",
            plot.margin = unit(c(0, 0, 0, 0), "lines"),
            title = element_text(face = "bold")) +
      ylab("") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L), limits = c(0, 0.3)) +
      scale_fill_discrete(name = "") +
      geom_violin(alpha = 0.8,
                  scale = "width",
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      geom_jitter(size = 0.05, alpha = 0.05) +
      stat_compare_means(method = "t.test", 
                         comparisons = list(#c("CTRL", "SSA-"), 
                                            #c("CTRL", "SSA+"), 
                                            c("CTRL", "SSAB")), 
                         label = "p",
                         size = 3.5,
                         method.args = list(exact = TRUE)
                         )
  }


#draw y label
y_label <- ggdraw() + 
  draw_label("VDJ, mutated bp (%)", 
             fontface = 'bold',
             hjust = 0.5, 
             angle = 90, 
             vjust = 0)

p_legend <- p_list1[[1]] %>% ggpubr::as_ggplot()

p_list1[[1]] <- p_list1[[1]] + theme(legend.position = "none")

for (i in 2:length(p_list1)){
p_list1[[i]] <- p_list1[[i]] + theme(axis.text.y = element_blank(), 
                                   axis.ticks.y = element_blank(), 
                                   axis.title.y = element_blank()
                                   ) + 
  theme(legend.position = "none")
}

p4 <- plot_grid(plotlist = p_list1, nrow = 1, rel_widths = c(1.25, 1, 1, 1, 1), labels = NULL, align = "vh") + 
          bgcolor("white") + 
          border("white")


ggsave2(paste0("../results/Fig3E.png"), 
          plot_grid(y_label, p4, nrow = 1, rel_widths = c(0.05, 1), align = "h") + 
          bgcolor("white") + 
          border("white"),
          width = 22, height = 10, unit = "cm")

ggsave2(paste0("../results/Fig3E.pdf"), 
          plot_grid(y_label, p4, nrow = 1, rel_widths = c(0.05, 1), align = "h") + 
          bgcolor("white") + 
          border("white"),
          width = 22, height = 10, unit = "cm")


```


#### plot Fig3F and SFig15, mu freq by for IGHV4-34 gene, celltype and pat gr.

```{r}
#select top10 used IGHV genes
vdj_gene <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       
                       mode = "gene",
                       fill = TRUE)
vdj_gene <- vdj_gene[order(vdj_gene$seq_count, decreasing = TRUE), ]
vdj_gene <- vdj_gene$gene[1:10]

gex.meta %>%
  group_by(cluster_cellType_manual_main, IGHV_gene, patient_group) %>%
  summarise(mean = mean(mu_freq_total, na.rm = T), n = n()) %>% filter(IGHV_gene == "IGHV4-34")
gex.meta %>%
  group_by(cluster_cellType_manual_fine, IGHV_gene, patient_group) %>%
  summarise(mean = mean(mu_freq_total, na.rm = T), n = n()) %>% filter(IGHV_gene == "IGHV4-34") %>% as.data.frame()
gex.meta[gex.meta$IGHV_gene %in% vdj_gene, ] %>%
  group_by(cluster_cellType_manual_fine, IGHV_gene, patient_group) %>%
  summarise(mean = mean(mu_freq_total, na.rm = T), n = n()) %>% filter(cluster_cellType_manual_fine == "Memory_Cl") %>% as.data.frame()

p_list <- list()
p_list2 <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))) {
p_list[[i]] <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i & gex.meta$IGHV_gene %in% vdj_gene, ], 
            aes(x = IGHV_gene, 
                y = mu_freq_total, 
                fill = patient_group)) +
      theme_classic() + 
      ggtitle(paste0(i)) +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5),
            plot.title = element_text(size = 12),
            plot.margin = unit(c(0, 0, 0, 0), "lines"),
            legend.position = "none",
            title = element_text(face = "bold")) +
      ylab("VDJ mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      scale_fill_discrete(name = "") +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) 
  p_list2[[i]] <- p_list[[i]] + theme(legend.position = "none") + ylab("")

  }

#get label
y_label <- ggdraw() + 
  draw_label("VDJ, mutated bp (%)", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

## get legend
p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i & gex.meta$IGHV_gene %in% vdj_gene, ], 
            aes(x = IGHV_gene, 
                y = mu_freq_total, 
                fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p1 <- plot_grid(plotlist = p_list2, ncol = 3, labels = NULL, align = "hv") + 
          bgcolor("white") + 
          border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/Fig3F.png"),
        p_list[["Memory_Classical"]],
          width = 18, height = 10, unit = "cm")

ggsave2(paste0("../results/Fig3F.pdf"),
        p_list[["Memory_Classical"]],
          width = 18, height = 10, unit = "cm")

gap <- 0
p3 <- plot_grid(plotlist = list(y_label, 
                                p_list1[[1]],
                                NULL,
                                p_list1[[2]],
                                NULL,
                                p_list1[[3]],
                                NULL,
                                p_list1[[4]],
                                NULL,
                                p_list1[[5]],
                                NULL,
                                p_list[["Memory_Classical"]] + ylab("")), 
                ncol = 12, labels = NULL, align = "hv", 
                rel_widths = c(0.4, 1, gap, 1, gap, 1, gap, 1, gap, 1, 0.6, 3.5)) + 
  bgcolor("white") + 
  border("white")


ggsave2(paste0("../results/Fig3E-F.png"),
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
  bgcolor("white") + 
  border("white"),
          width = 47, height = 10, unit = "cm")

ggsave2(paste0("../results/Fig3E-F.pdf"),
        plot_grid(p3, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
  bgcolor("white") + 
  border("white"),
          width = 47, height = 10, unit = "cm")


ggsave2(paste0("../results/SFig15.png"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

ggsave2(paste0("../results/SFig15.pdf"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")


```


#### plot SFig16 mu freq for IGHV4-34, per celltype and sample.

```{r}

#select top10 used IGHV genes
vdj_gene <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       
                       mode = "gene",
                       fill = TRUE)
vdj_gene <- vdj_gene[order(vdj_gene$seq_count, decreasing = TRUE), ]
vdj_gene <- vdj_gene$gene[1:10]

# fix sample order
tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)


ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                   DNEG = "#7CAE00", 
                                   SSA = "#00BFC4", 
                                   SSAB = "#C77CFF"))

colors_samples <- c(rep(ann_cols$patient_group[1], times = 4),
                    rep(ann_cols$patient_group[2], times = 6),
                    rep(ann_cols$patient_group[3], times = 8),
                    rep(ann_cols$patient_group[4], times = 9))

p_list <- list()
p_list2 <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))) {
p_list[[i]] <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i & 
                                 gex.meta$IGHV_gene == "IGHV4-34" & 
                                 !is.na(gex.meta$mu_freq_total), ], 
            aes(x = orig.ident, 
                y = mu_freq_total,
                fill = patient_group)) +
      theme_classic() + 
      ggtitle(paste0("IGHV4-34 cells in: ", i)) +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       color = colors_samples),
            plot.title = element_text(size = 10),
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      scale_x_discrete(drop=FALSE) +
      scale_fill_discrete(name = "", drop = FALSE) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) 
  p_list2[[i]] <- p_list[[i]] + theme(legend.position = "none") + ylab("")

  }

#get label
y_label <- ggdraw() + 
  draw_label("mutated bp (%)", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

## get legend
p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i & 
                                 gex.meta$IGHV_gene == "IGHV4-34" & 
                                 !is.na(gex.meta$mu_freq_total), ], 
            aes(x = IGHV_gene, 
                y = mu_freq_total, 
                fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p1 <- plot_grid(plotlist = p_list2, ncol = 3, labels = NULL, align = "hv") + 
          bgcolor("white") + 
          border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/SFig16.png"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

ggsave2(paste0("../results/SFig16.pdf"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

```


#### plot SFig12A, total VDJ mutation frequency - celltypeMain - orig.ident 

```{r}

tmp <- data.frame(orig.ident = gex.meta$orig.ident, patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, levels = tmp$orig.ident)

p <- ggplot(gex.meta,
            aes(x = orig.ident, 
                y = mu_freq_total,
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5,
                                       size = 8), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent, limits = c(0, 0.2))  +
      scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)

## get legend
p_legend <- ggplot(gex.meta, 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

ggsave2(paste0("../results/SFig12A.png"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 34, height = 12, unit = "cm")

ggsave2(paste0("../results/SFig12A.pdf"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 34, height = 12, unit = "cm")


```


####  plot SFig12B, total VDJ mutation frequency - celltypeFine - orig.ident

```{r}

# relevel orig.ident to plot based on patient group order
tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
p[[i]] <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ], 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            legend.position = "none") +
      ggtitle(i) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L),
                         breaks = seq(0, 0.2, 0.05), 
                         limits = c(0, 0.2)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)  
}


y_label <- ggdraw() + 
  draw_label("mutated bp (%)", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

## get legend
p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == unique(gex$cluster_cellType_manual_fine)[1], ], 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p1 <- plot_grid(plotlist = p, ncol = 3, labels = NULL, align = "hv") + 
          bgcolor("white") + 
          border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/SFig12B.png"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

ggsave2(paste0("../results/SFig12B.pdf"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")


```


#### plot SFig13A, CDR3 mutation count ratio R/S - cell type fine - patient_group, celltypeMain

```{r}
gex.meta$mu_count_cdr3_RSratio <- log2((gex.meta$mu_count_cdr3_r + 1) / 
                                         (gex.meta$mu_count_cdr3_s + 1))

#remove cells with no mutations
#Removed 181168 rows containing non-finite values (stat_ydensity). 
gex.meta$mu_count_cdr3_RSratio[(gex.meta$mu_count_cdr3_r + 
                                  gex.meta$mu_count_cdr3_s) == 0] <- NA

nrow(gex.meta[!is.na(gex.meta$mu_count_cdr3_RSratio), ])

p2 <- ggplot(gex.meta, 
             aes(x = cluster_cellType_manual_main, 
                 y = mu_count_cdr3_RSratio, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0("CDR3 mutation ratio Replacement/Silent for ", 
                     nrow(gex.meta[!is.na(gex.meta$mu_count_cdr3_RSratio), ]),
                     " mutated cells")) +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5),
            plot.title = element_text(size = 12),
            legend.title = element_blank()) +
      ylab("mutation count ratio R/S (log2)") +
      xlab("") +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)

ggsave2(paste0("../results/SFig13A.png"), 
          p2,
          width = 20, height = 10, unit = "cm")

ggsave2(paste0("../results/SFig13A.pdf"), 
          p2,
          width = 20, height = 10, unit = "cm")


```


#### plot SFig13B, fraction of CDR3 mutated cells per sample and celltypeFine

```{r}

p3 <- as_tibble(gex.meta) %>% 
  select(cluster_cellType_manual_fine, 
         patient_group,
         orig.ident,
         mu_count_cdr3_r, 
         mu_count_cdr3_s) %>% 
  replace_na(list(mu_count_cdr3_r = 0, mu_count_cdr3_s = 0)) %>%
  mutate(mu_cdr3_binary = mu_count_cdr3_r #+ mu_count_cdr3_s
         ,
         mu_cdr3_binary = (mu_cdr3_binary > 0) * 1,
         mu_cdr3_binary = recode(mu_cdr3_binary, `0` = "Unmutated", `1` = "Mutated")) %>%
  group_by(cluster_cellType_manual_fine,
           patient_group,
           orig.ident, 
           mu_cdr3_binary) %>%
  summarise(cdr3_n = n()) %>%
  mutate(cdr3_freq = cdr3_n / sum(cdr3_n)) %>% 
  filter(mu_cdr3_binary == "Mutated") %>%
  ggplot(aes(x = cluster_cellType_manual_fine, 
             y = cdr3_freq, 
             fill = patient_group)) +
  geom_violin(alpha = 0.8, 
              scale = "width", 
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE) +
  ylab("Proportion of cells with \nreplacement mutations within CDR3") +
  xlab("") +
  ggtitle("Replacement mutations in CDR3") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5),
        plot.title = element_text(size = 12),
        legend.title = element_blank()) +
  stat_compare_means(method = "anova", hide.ns = TRUE, label = "p.signif") +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1L), name = "") 

ggsave2(paste0("../results/SFig13B.png"), 
          plot_grid(p3, labels = NULL, align = "h")+ 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 10, unit = "cm")

ggsave2(paste0("../results/SFig13B.pdf"), 
          plot_grid(p3, labels = NULL, align = "h")+ 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 10, unit = "cm")


```


#### plot SFig13B, fraction of CDR3 mutated cells per sample and celltypeMain

```{r}

p3 <- as_tibble(gex.meta) %>% 
  select(cluster_cellType_manual_main, 
         patient_group,
         orig.ident,
         mu_count_cdr3_r, 
         mu_count_cdr3_s) %>% 
  replace_na(list(mu_count_cdr3_r = 0, mu_count_cdr3_s = 0)) %>%
  mutate(mu_cdr3_binary = mu_count_cdr3_r,
         mu_cdr3_binary = (mu_cdr3_binary > 0) * 1,
         mu_cdr3_binary = recode(mu_cdr3_binary, `0` = "Unmutated", `1` = "Mutated")) %>%
  group_by(cluster_cellType_manual_main,
           patient_group,
           orig.ident, 
           mu_cdr3_binary) %>%
  summarise(cdr3_n = n()) %>%
  mutate(cdr3_freq = cdr3_n / sum(cdr3_n)) %>% 
  filter(mu_cdr3_binary == "Mutated") %>%
  ggplot(aes(x = cluster_cellType_manual_main, 
             y = cdr3_freq, 
             fill = patient_group)) +
  geom_violin(alpha = 0.8, 
              scale = "width", 
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE) +
  geom_jitter(size = 0.05, alpha = 0.05) +
  ylab("Proportion of cells with \nreplacement mutations within CDR3") +
  xlab("") +
  ggtitle("Replacement mutations in CDR3") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5),
        plot.title = element_text(size = 12),
        legend.title = element_blank()) +
  stat_compare_means(method = "anova", hide.ns = TRUE, label = "p.signif") +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1L), name = "") 

ggsave2(paste0("../results/SFig13B.png"), 
          plot_grid(p3, labels = NULL, align = "h")+ 
          bgcolor("white") + 
          border("white"),
          width = 20, height = 10, unit = "cm")

ggsave2(paste0("../results/SFig13B.pdf"), 
          plot_grid(p3, labels = NULL, align = "h")+ 
          bgcolor("white") + 
          border("white"),
          width = 20, height = 10, unit = "cm")


```


## CDR3 attributes
#### plot SFig14, CDR3 aa attributes

```{r}

cdr3_attr <- c("cdr3_aa_length", "cdr3_aa_gravy","cdr3_aa_bulk",
               "cdr3_aa_aliphatic", "cdr3_aa_polarity", "cdr3_aa_charge", 
               "cdr3_aa_basic", "cdr3_aa_acidic", "cdr3_aa_aromatic")

#CDR3 aa length
p1 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_length,
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa length") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("length (bp)") +
      xlab("") +
      scale_y_continuous(limits = c(4, 31)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#grand average of hydrophobicity
p2 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_gravy,
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa grand average of hydrophobicity") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("hydrophobicity score") +
      xlab("") +
      scale_y_continuous(limits = c(-3, 3)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#average bulkiness of amino acids
p3 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_bulk, #average bulkiness of amino acids
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa bulkiness") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("average bulkiness of amino acids") +
      xlab("") +
      scale_y_continuous(limits = c(5, 19)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#cdr3_aa_aliphatic
p4 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aliphatic, #aliphatic index
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aliphatic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("aliphatic index") +
      xlab("") +
      scale_y_continuous(limits = c(0, 2.2)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#cdr3_aa_polarity
p5 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_polarity, #average polarity of amino acids
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa polarity") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("average polarity") +
      xlab("") +
      scale_y_continuous(limits = c(6, 11)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#cdr3_aa_charge
p6 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_charge, #net charge
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa charge") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("net charge") +
      xlab("") +
      scale_y_continuous(limits = c(-7, 5)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)

#cdr3_aa_basic
p7 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_basic, #fract. of informative positions that are Arg, His or Lys
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa basic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("% of pos that are Arg, His or Lys") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L), limits = c(0, 0.5)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#cdr3_aa_acidic
p8 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_acidic, #fraction of informative positions that are Asp or Glu
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa acidic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("% of pos. that are Asp or Glu") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L), limits = c(0, 0.5)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)


#cdr3_aa_aromatic
p9 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic, #fract. of informative pos. that are His, Phe, Trp or Tyr
                  fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aromatic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8), 
            legend.position = "none") +
      ylab("% of pos. that are His, Phe, Trp or Tyr") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L), limits = c(0, 0.5)) +
      geom_violin(alpha = 0.8,
              scale = "width",
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE,
              na.rm = TRUE)

#get legend
p_legend <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic,
                  fill = patient_group)) +
      theme_classic() +
      geom_boxplot() +
      theme(legend.position = "bottom",
            legend.title = element_blank())
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()


p_grid <- plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, p9, 
                    nrow = 3, ncol = 3, labels = "AUTO", align = "hv")

ggsave2(paste0("../results/SFig14.png"), 
          plot_grid(p_grid, p_legend, ncol = 1, align = "none", rel_heights = c(18, 1)) +
          bgcolor("white") + 
          border("white") ,
          width = 35, height = 30, unit = "cm")

ggsave2(paste0("../results/SFig14.pdf"), 
          plot_grid(p_grid, p_legend, ncol = 1, align = "none", rel_heights = c(18, 1)) +
          bgcolor("white") + 
          border("white") ,
          width = 35, height = 30, unit = "cm")


```


## save rds GEX7_VDJ

```{r}

#saveRDS(gex, file = "../results/GEX7_BVDJ.rds")
#gex <- readRDS("../results/GEX7_BVDJ.rds")

```


## print sessionInfo()

```{r}

sessionInfo()

```



