---
output: html_document
editor_options: 
  chunk_output_type: console
---

# VDJ analysis {#VDJ}
## Load packages

```{r}

source("./00_dependencies.R")

```


## Load Seurat object

```{r}

gex <- readRDS("../results/GEX6_BANNOT.rds"); gex
#colnames(gex@meta.data)

```


## Load pheno file

```{r}

pheno <- read.table("../suppl/pheno/pSS_pheno_210623.csv", sep = ";", header = TRUE)
colnames(pheno)[c(1, 3)] <- c("Sample_id", "orig.ident")
pheno$orig.ident <- gsub("P007[ab]", "P007", pheno$orig.ident)
pheno$orig.ident <- gsub("P008[ab]", "P008", pheno$orig.ident)
pheno <- unique(pheno[, c(1, 3, 5, 7:8, 11:12)])

pheno$patient_group <- paste0(ifelse(pheno$SSA == "YES", "SSA", ""), 
                            "_", 
                            ifelse(pheno$SSB == "YES", "SSB", ""))
pat_annot <- c(SSAB = "SSA_SSB", SSA = "SSA_", DNEG="NA_NA", DNEG="_")
pheno$patient_group <- names(pat_annot)[match(pheno$patient_group, pat_annot)]
pheno$patient_group[grep("^C00", pheno$orig.ident) ] <- "CTRL"

head(pheno); dim(pheno)

#Four non-overlapping groups
#1. SSAB - SSA+/SSB+
#2. SSA  - SSA+/SSB-
#3. DNEG - SSA-/SSB-
#4. CTRL - CTRL


```


## immcantation
### Load output from Immcantation and perform mutation analysis

```{r}

files_B <- list.files("../data/clono_B_VDJ_tmp", 
                      pattern = "*germ-pass.tsv", 
                      recursive = TRUE, 
                      full.names = TRUE)
ncpu <- cpuCount()
imm <- NULL
for (i in files_B) {
  
  cat("reading: ", print(i), "\n")
  tmp <- read_rearrangement(i)
  
  ####---- get sample name
  j <- gsub("_heavy_germ-pass.tsv", "", gsub(".*out/", "", i))
  cat("sample name: ", print(j), "\n")
  
  ####---- number of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = FALSE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_v_total'
  
  ####---- frequency of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = TRUE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_v_total'
  
  ####---- number of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")

  ####---- number of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
   
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_total'
  
  ####---- frequency of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_total'
 
  ####---- properties for CDR3 aa's
  tmp <- aminoAcidProperties(tmp, 
                             seq = "junction", 
                             nt = TRUE, 
                             trim = TRUE, 
                             label = "cdr3")
  
  tmp$cell_id <- paste0(j, "_", gsub("-1", "", tmp$cell_id))
  tmp$orig.ident <- gsub("a|b", "", gsub("_.*", "", tmp$cell_id))
  tmp$sequence_id <- paste0(j, "_", gsub("-1", "", tmp$sequence_id))
  rownames(tmp) <- tmp$cell_id
  
  if (is.null(imm)) {
    imm <- tmp
  } else {
    imm <- rbind(imm, tmp)
  }
  
}

head(imm)


```


### integrate immcantation VDJ data with seurat object

```{r}

#select immcantation columns
# cols <- c("sequence_id", "productive", "vj_in_frame", "stop_codon", "locus", 
#           "v_call", "d_call", "j_call", "c_call", 
#           "cell_id", "clone_id", "umi_count", 
#           "v_call_10x", "d_call_10x", "j_call_10x", 
#           "germline_v_call","germline_d_call", "germline_j_call",
#           "cdr3", "consensus_count",
#           "mu_count_total", "mu_freq_total", 
#           "mu_count_v_r_total", "mu_count_v_s_total", "mu_freq_v_r_total", "mu_freq_v_s_total",
#           "mu_freq_cdr1_r", "mu_freq_cdr1_s", "mu_freq_cdr2_r", "mu_freq_cdr2_s", "mu_freq_cdr3_r", "mu_freq_cdr3_s", "mu_freq_fwr1_r", "mu_freq_fwr1_s", "mu_freq_fwr2_r", "mu_freq_fwr2_s", "mu_freq_fwr3_r", "mu_freq_fwr3_s", "mu_freq_fwr4_r", "mu_freq_fwr4_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "cdr3_aa_length", "cdr3_aa_gravy", "cdr3_aa_bulk", "cdr3_aa_aliphatic", "cdr3_aa_polarity", "cdr3_aa_charge", "cdr3_aa_basic", "cdr3_aa_acidic", "cdr3_aa_aromatic")

#colnames(imm)
#select immcantation columns NOT to be added to gex
cols <- c("sequence", "sequence_alignment", "germline_alignment", 
          "v_cigar", "d_cigar", "j_cigar",
          "fwr1", "fwr2", "fwr3", "fwr4", "cdr1", "cdr2", 
          "germline_alignment_d_mask")

imm_add <- imm[, !(colnames(imm) %in% cols)]
rownames(imm_add) <- imm_add$cell_id

gex <- AddMetaData(gex, imm_add)
colnames(gex@meta.data)
table(gex$orig.ident, useNA = "always")


```


### add pheno data to immcantation object

```{r}

#imm$orig.ident <- gsub("a|b", "", gsub("_.*", "", imm$cell_id))
imm <- merge(imm, 
                 pheno[, c(2,8)], 
                 by.x = "orig.ident", 
                 by.y = "orig.ident", 
                 all.x = TRUE); head(imm)
imm <- imm[, c(2:ncol(imm), 1)]
colnames(imm)

```


### save immcantation VDJ data

```{r}

colnames(imm)[colnames(imm) == 'orig.ident'] <- 'repertoire_id'
imm$duplicate_count <- imm$umi_count
colnames(imm)[colnames(imm) == 'clone_id'] <- 'clone_id_perSample'

write_rearrangement(imm, "../results/VDJ_immcantation_out.tsv")
#imm <- read_rearrangement("../results/VDJ_immcantation_out.tsv")

#imm_out <- imm[, c("sequence_id", "v_call", "j_call", "junction")]
#write_rearrangement(imm_out, "../results/VDJ_immcantation_out_forDefineClones.tsv")
table(imm$repertoire_id, useNA = "always")

```


### read compairr out

```{r}

cairr.aa <- read.table("../results/VDJ_compairr_out_aa.tsv", header = TRUE)
cairr.aa$barcode <- gsub("_contig_1", "", gsub("_contig_2", "", gsub("_contig_3", "", cairr.aa$sequence_id)))
rownames(cairr.aa) <- cairr.aa$barcode

colnames(cairr.aa)[1:2] <- paste0(colnames(cairr.aa)[1:2], "_airr")


#add to gex
gex <- AddMetaData(gex, cairr.aa[, 1:2])
colnames(gex@meta.data)


```


### VDJ mutation
#### add categorical variable for total mutation burden

```{r}

# unmutated
# low < 1%
# med < 2%
# high > 2%

gex$mu_load <- ifelse(gex$mu_freq_total == 0, "unmutated", 
                 ifelse(gex$mu_freq_total > 0 & gex$mu_freq_total <= 0.01, "low",
                        ifelse(gex$mu_freq_total > 0.01 & gex$mu_freq_total <= 0.02, "med",
                               ifelse(gex$mu_freq_total > 0.02, "high", NA))))


```


#### add categorical variable for clonotype expansion

```{r}

# "Hyperexpanded (100 < X <= 500)"
# "Large (20 < X <= 100)"
# "Medium (5 < X <= 20)"
# "Small (1 < X <= 5)"
# "Single (0 < X <= 1)"

table(gex$cluster_size_airr, useNA ="always")
factor(slot(seurat, "meta.data")$cloneType, 
                levels = c("Hyperexpanded (100 < X <= 500)", 
                           "Large (20 < X <= 100)", 
                            "Medium (5 < X <= 20)", 
                            "Small (1 < X <= 5)", 
                            "Single (0 < X <= 1)", NA))

gex$cloneType_airr <- factor(ifelse(is.na(gex$cluster_size_airr), NA, 
                             ifelse(gex$cluster_size_airr == 1, "Single (0 < X <= 1)", 
                                    ifelse((gex$cluster_size_airr > 1) & (gex$cluster_size_airr <= 5), "Small (1 < X <= 5)", ifelse((gex$cluster_size_airr > 5) & (gex$cluster_size_airr <= 20), "Medium (5 < X <= 20)", ifelse((gex$cluster_size_airr > 20) & (gex$cluster_size_airr <= 100), "Large (20 < X <= 100)", "Hyperexpanded (100 < X)"))))), 
                             levels = c("Hyperexpanded (100 < X <= 500)",
                                        "Large (20 < X <= 100)", 
                                        "Medium (5 < X <= 20)", 
                                        "Small (1 < X <= 5)", 
                                        "Single (0 < X <= 1)", NA))


```


#### rank clonotypes based on mutation status

```{r}

as_tibble(as.data.frame.matrix(table(gex$cluster_no_airr, gex$mu_load))) %>% 
  
colnames(gex@meta.data)

```


## scRepertiore BCR
### Read 10x BCR files

```{r}

# read 10x VDJ BCR files
samples_B <- list.files("../data/clono_B_VDJ")

clono_B <- lapply(samples_B, function(x){
  
  tmp <- read.csv(paste0("../data/clono_B_VDJ/", x, "/", x, "_B_filtered_contig_annotations.csv"), 
                  stringsAsFactors = FALSE)
  return(tmp)
  
})

names(clono_B) <- samples_B

#remove trailing -1
clono_B <- lapply(seq_along(clono_B), function(i){
  clono_B[[i]]$barcode <- gsub("-1", "", clono_B[[i]]$barcode)
  return(clono_B[[i]])
})

names(clono_B) <- samples_B

clono_B <- clono_B[order(names(clono_B))]

print(object.size(clono_B), units = "MB") #227.6 Mb
saveRDS(clono_B, paste0("../results/clono_B_list.rds"))
#clono_B <- readRDS("../results/clono_B_list.rds")


```


### generate a combined data frame for BCR sequences using scRepertoire

```{r}

#split combineBCR and do it per sample to avoid memory issues
#takes on average 20min/sample to run
combined_B <- lapply(1:length(clono_B), function(x) {
  
  print(names(clono_B[x]))
  print(nrow(clono_B[[x]]))
  print(Sys.time())
  
  tmp <- combineBCR(clono_B[[x]],
                    samples = names(clono_B[x]),
                    ID = NULL,
                    removeNA = TRUE,
                    removeMulti = TRUE)

  return(tmp)
  
})

#reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
combined_B <- lapply(1:length(combined_B), function(x){
  tmp <- combined_B[[x]][[1]]
  return(tmp)
})

#---- handle samples where libraries were created and sequenced twice
combined_B[["P007"]] <- rbind(combined_B[["P007a"]], combined_B[["P007b"]])
combined_B[["P007a"]] <- NULL
combined_B[["P007b"]] <- NULL

combined_B[["P008"]] <- rbind(combined_B[["P008a"]], combined_B[["P008b"]])
combined_B[["P008a"]] <- NULL
combined_B[["P008b"]] <- NULL

combined_B <- combined_B[order(names(combined_B))]

names_tmp <- names(combined_B)
combined_B <- lapply(seq_along(combined_B), function(i) {
  combined_B[[i]]$sample <- gsub("a|b", "", combined_B[[i]]$sample)
  return(combined_B[[i]])
})

names(combined_B) <- names_tmp
#----

print(object.size(combined_B), units = "MB") #181.3 Mb
saveRDS(combined_B, paste0("../results/combined_BCR_scRepertoire.rds"))
#combined_B <- readRDS("../results/combined_BCR_scRepertoire.rds")


```


### Integrate scRepertoir out with Seurat object

```{r}

gex <- combineExpression(combined_B, gex)

```


## add 10x IgH call to Seurat object

```{r}

igh_10x <- lapply(1:length(clono_B), function(x) {
  
  tmp <- clono_B[[x]][clono_B[[x]]$chain == "IGH" & 
                        clono_B[[x]]$productive == "true" & 
                        !is.na(clono_B[[x]]$c_gene) & 
                        clono_B[[x]]$c_gene != "" &
                        clono_B[[x]]$c_gene != "IGLC1" &
                        clono_B[[x]]$c_gene != "IGLC2" , c(1, 10)]
  tmp$barcode <- paste0(names(clono_B)[x], "_", tmp$barcode)
  names(tmp)[2] <- "c_gene_10x"

  tmp <- tmp[!duplicated(tmp$barcode), ]
  
  rownames(tmp) <- tmp$barcode
  tmp$barcode <- NULL
  
  return(tmp)
  }
)

igh_10x <- do.call(rbind.data.frame, igh_10x)
table(igh_10x$c_gene_10x, useNA = "always")
gex <- AddMetaData(gex, igh_10x)
table(gex$orig.ident, useNA = "always")

```


## Analyse and visualize immcantation and scRepertoir output
### VDJ gene usage
#### VDJ correlation

```{r}

#VDJ gene usage correlation
VDJ_cor <- function(i, #seurat object meta data containing data for VDJ gene usage
                    correl, #which correlation to use i.e. spearman or pearson...
                    VDJgene #VDJ gene to plot
                    ){
  
  cat(VDJgene, "\n") 
  cat(correl, "\n")
  
  #remove allele information
  i[, VDJgene] <- unlist(lapply(strsplit(i[, VDJgene],
                                                   split = "\\*"),
                                          function(x) {return(x[1])}))
  
  i$celltype_patient_group <- paste0(i$cluster_cellType_manual_fine, 
                                     "-", 
                                     i$patient_group)
  
  # tmp <- table(list(i@meta.data[, VDJgene], 
  #                   i$celltype_patient_group))
  
  tmp <- as.data.frame.matrix(table(list(i[, VDJgene],
                    i$celltype_patient_group)))
  
  cor_tmp <- cor(tmp, method = correl)
  cor_tmp[is.na(cor_tmp)] <- 0

  p <- pheatmap::pheatmap(cor_tmp, 
                     annotation_col = data.frame(cell_type = factor(sub("-.*", "", colnames(cor_tmp))),
                                                 patient_group = factor(sub(".*-", "", colnames(cor_tmp))),
                                                 row.names = colnames(cor_tmp)),
                      breaks = seq(0, 1, length.out = 100),
                      filename = paste0("../results/VDJ_Cor_", VDJgene, "_", 
                                        correl, "_per_cluster_PatientGroup.png"),
                      fontsize = 8, 
                      fontsize_row = 4, 
                      fontsize_col = 4,
                      border_color = NA,
                      main = paste(VDJgene, correl, "correlation"),
                      color = c(colorRampPalette(c("lightblue1", "lightblue2", "firebrick2"))(99)))
  
}

#for all cell types
VDJ_cor(gex@meta.data, "spearman", "v_call")
VDJ_cor(gex@meta.data, "pearson", "v_call")

VDJ_cor(gex@meta.data, "spearman", "d_call")
VDJ_cor(gex@meta.data, "pearson", "d_call")

VDJ_cor(gex@meta.data, "spearman", "j_call")
VDJ_cor(gex@meta.data, "pearson", "j_call")

VDJ_cor(gex@meta.data, "spearman", "c_call")
VDJ_cor(gex@meta.data, "pearson", "c_call")


#for a subset of cell types
gex.meta <- gex@meta.data[gex$cluster_cellType_manual_fine == "DN1_ID3" | 
                          gex$cluster_cellType_manual_fine == "DN2" |
                          gex$cluster_cellType_manual_fine == "DN2_CXCR3" |
                          gex$cluster_cellType_manual_fine == "DN2_ITGAX" |
                          gex$cluster_cellType_manual_fine == "DN4_IGHE" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                          gex$cluster_cellType_manual_fine == "Naive" |
                          gex$cluster_cellType_manual_fine == "Naive_IFN" |
                          gex$cluster_cellType_manual_fine == "Naive_Transitional", ]


VDJ_cor(gex.meta, "spearman", "v_call")
VDJ_cor(gex.meta, "pearson", "v_call")

VDJ_cor(gex.meta, "spearman", "d_call")
VDJ_cor(gex.meta, "pearson", "d_call")

VDJ_cor(gex.meta, "spearman", "j_call")
VDJ_cor(gex.meta, "pearson", "j_call")

VDJ_cor(gex.meta, "spearman", "c_call")
VDJ_cor(gex.meta, "pearson", "c_call")


```


#### top VDJ genes

```{r}

# get most used VDJ genes per patient subgroup

get_topVDJ <- function(i, #seurat object
                       VDJgene = "tmp" #VDJ gene to plot
  ) {
  
  #remove allele information
  vh_col <- unlist(lapply(strsplit(i@meta.data[, VDJgene], 
                                                   split = "\\*"),
                                          function(x) {return(x[1])})) 
  
  vh <- as.data.frame.matrix(table(vh_col, i$patient_group))
  vh <- vh[order(rowSums(vh), decreasing = TRUE), ]
  
  #export counts per patient group
  #write.csv(vh, paste0("../results/VDJ_geneUsage_", VDJgene, "_genesPerPatientGroup.csv")) 
  
  #get percentages of repertoir
  vh <-  as.data.frame(t(t(vh) / colSums(vh))) 
  vh$gene <- rownames(vh)
  
  nvh <- ifelse(nrow(vh) <= 50, nrow(vh), 50)
  
  p <- as_tibble(vh[1:nvh, ]) %>% pivot_longer(
       cols = 1:4,
       names_to = "patient_group",
       names_prefix = "",
       values_to = "count",
       values_drop_na = TRUE) %>%  
          ggplot(aes(x = reorder(gene, -count), y = count, fill = patient_group)) +
              geom_col(position = "dodge") +
              theme_classic() +
              xlab("") +
              ylab("Percent of repertoire") +
              scale_y_continuous(labels = percent) +
              ggtitle(paste0(VDJgene, " genes")) +
              theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
              facet_grid(rows = vars(patient_group))

    ggsave2(paste0("../results/VDJ_geneUsage_", VDJgene, "_geneUsage_perPatientGroup.png"), 
          p,
          width = 30, height = 14, unit = "cm")
  
   }

get_topVDJ(gex, "v_call")
get_topVDJ(gex, "d_call")
get_topVDJ(gex, "j_call")
get_topVDJ(gex, "c_call")


```


#### Fig5B IGHV usage per patient group

```{r}

####---- make Fig5B
#VDJgene <- "v_call"
vdj_gene <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group"), 
                       mode = "gene",
                       fill = TRUE)

vdj_gene$fc <- NA

#calculate FC relative to CTRL
table(vdj_gene$gene)
for (i in 1:nrow(vdj_gene)){
  ifelse(vdj_gene$patient_group[i] == "CTRL", 
         vdj_gene$fc[i] <- NA,
         vdj_gene$fc[i] <- log2((vdj_gene$seq_freq[i] * 100 + 1)/ (vdj_gene$seq_freq[vdj_gene$gene == vdj_gene$gene[i] & vdj_gene$patient_group == "CTRL"] *100  + 1)))
}

vdj_gene <- vdj_gene[order(vdj_gene$seq_count, decreasing = TRUE), ]
tmp <- vdj_gene[which(vdj_gene$gene %in% unique(vdj_gene$gene)[1:45]), ]

p1 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = fc, 
                          colour = patient_group)) +
  theme_classic() +
      ggtitle(paste0("IGHV gene usage per patient group")) +
      theme(axis.text.x = element_blank(),
            axis.ticks.x=element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0,0,0), "cm")) +
          ylab("log2FC to CTRL") +
          xlab("") +
          geom_point() +
      geom_hline(yintercept = 0)

p2 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0,0,0,0.2), "cm")) +
      ylab("Percent of repertoire") +
      xlab("") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

p_legend <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
  scale_fill_discrete(name = "Patient group") +
      geom_col(position = "dodge")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p3 <- plot_grid(p1, p2, nrow = 2, labels = NULL,
                align = "v", axis = "l", rel_heights = c(1, 2.8))


 ggsave2(paste0("../results/VDJ_countGenes_", VDJgene 
                 ,".png"),
          plot_grid(p3, p_legend, 
                    ncol = 2, 
                    labels = NULL, 
                    align = "v", 
                    rel_widths = c(12, 2)) + bgcolor("white"),
          width = 20, height = 12, unit = "cm")


```


#### allele usage for differentially used IGHV genes

```{r}

vdj_allele <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group"), 
                       mode = "allele",
                       fill = TRUE)

#IGHV1-69
vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]

p1 <- ggplot(tmp1, aes(x = reorder(gene, -seq_freq), 
                       y = seq_freq, 
                       fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2, 0, 0, 0.2), "cm")) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-69 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup.png"), 
          plot_grid(p1, ncol = 1, labels = NULL, align = "h"),
          width = 15, height = 10, unit = "cm")

vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]


#
vdj_allele <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group", "cluster_cellType_manual_main"),
                       mode = "allele",
                       fill = TRUE)

vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]


p1 <- ggplot(tmp1, aes(x = reorder(gene, -seq_freq), 
                       y = seq_freq, 
                       fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            axis.text.y = element_text(hjust = 1, 
                                       vjust = 0.5,
                                       size = 6),
            strip.background = element_blank(), 
            #strip.text.y = element_blank(),
            #legend.position = "none",
            plot.margin = unit(c(0.2, 0, 0, 0.2), "cm"),
            strip.text.y = element_text(angle = 0, vjust = 0)) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-69 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(cluster_cellType_manual_main))

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perCelltypeMain.png"), 
          plot_grid(p1, ncol = 1, labels = NULL, align = "h"),
          width = 15, height = 10, unit = "cm")


#get v_call genes
gex@meta.data$IGHV <- unlist(lapply(strsplit(gex@meta.data$v_call, 
                                             split = "\\*"), function(x) {return(x[1])}))

gex@meta.data$IGHV1_69_umi <- 0
gex@meta.data$IGHV1_69_umi[which(gex@meta.data$IGHV == "IGHV1-69")] <- gex@meta.data$umi_count[which(gex@meta.data$IGHV == "IGHV1-69")]

# featureplots for IGHV1_69 UMIs
p <- list()
for (i in sort(unique(gex$patient_group))){
p[[i]] <- FeaturePlot(gex %>% tidyseurat::filter(patient_group == i), 
            features = "IGHV1_69_umi",
            reduction = "umap_2",  
            raster = FALSE, 
            pt.size = 0.1, 
            ncol = 1, 
            order = TRUE,
            by.col = TRUE) +
  xlab("UMAP 1") +
  ylab("UMAP 2") +
  ggtitle(paste0(i, " IGHV1-69 UMI ")) +
  theme(axis.text = element_blank(),
          axis.ticks = element_blank())
  
}

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup_featureplot.png"), 
          plot_grid(plotlist = p, ncol = 2, labels = NULL, align = "h"),
          width = 20, height = 20, unit = "cm")


#percent of IGHV-69 cells per patient group per celltype fine
tmp <- as.data.frame.matrix(table(gex@meta.data$IGHV, gex@meta.data$cluster_cellType_manual_fine))
gex@meta.data$cluster_cellType_manual_fine <- as.factor(gex@meta.data$cluster_cellType_manual_fine)

p <- as_tibble(gex@meta.data) %>% 
  select(orig.ident, cluster_cellType_manual_fine, IGHV, patient_group) %>% 
  filter(IGHV == "IGHV1-69")  %>%  
  add_count(orig.ident, cluster_cellType_manual_fine, .drop = FALSE) %>% 
  distinct() %>%
  group_by(orig.ident) %>%
  mutate(per =  prop.table(n)) %>% 
  ggplot(aes(x = cluster_cellType_manual_fine,
             y = per,
             fill = patient_group)) +
  scale_y_continuous(labels = percent) +
  theme_classic() +
  xlab("") +
  ylab("percent of IGHV1-69 cells per patient group") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  geom_boxplot()

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup_perCelltypeFine.png"), 
          p,
          width = 20, height = 12, unit = "cm")

 
#IGHV1-18
vdj_allele2 <- vdj_allele[grepl("^IGHV1-18*", vdj_allele$gene), ] %>% arrange(desc(seq_count))
tmp2 <- vdj_allele2[which(vdj_allele2$gene %in% unique(vdj_allele2$gene)[1:10]), ]

p2 <- ggplot(tmp2, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0,0,0.2), "cm")) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-18 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

#IGHV3-30-3
vdj_allele3 <- vdj_allele[grepl("^IGHV3-30-3*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp3 <- vdj_allele3[which(vdj_allele3$gene %in% unique(vdj_allele3$gene)[1:10]), ]

p3 <- ggplot(tmp3, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0,0,0.2), "cm")) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV3-30-3 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))


```


### clonotype analysis

```{r}

####---------------------------------------------------------------------
####---- clonal diversity
####---------------------------------------------------------------------

####---- abundance estimation
patient_group_estimateAbundance <- estimateAbundance(gex@meta.data, 
                                              group = "patient_group", 
                                              clone = "cluster_no_airr",
                                              ci = 0.95, 
                                              nboot = 100)
saveRDS(patient_group_estimateAbundance, "../results/VDJ_clono_patient_group_estimateAbundance.rds")
patient_group_estimateAbundance <- readRDS("../results/VDJ_clono_patient_group_estimateAbundance.rds")

#Error: vector memory exhausted (limit reached?) 
#nboot = 100 works locally, nboot = 200 does not

#plotAbundanceCurve(patient_group_estimateAbundance) + theme_classic()

png("../results/VDJ_clono_Abundances_patient_group.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_abundances, 
     #colors = sample_colors, 
     main_title = "Clonal abundances", 
     legend_title = "patient group")
dev.off()

rm(patient_group_estimateAbundance); gc()

####---- diversity curve
####---- per orig.ident
orig.ident_alphaDiversity <- alphaDiversity(gex@meta.data, 
                                   group = "orig.ident", 
                                   clone = "cluster_no_airr",
                                   min_q = 0, 
                                   max_q = 4, 
                                   step_q = 0.1,
                                   ci = 0.95, 
                                   nboot = 50)

saveRDS(orig.ident_alphaDiversity, "../results/VDJ_clono_orig.ident_alphaDiversity.rds")
#orig.ident_alphaDiversity <- readRDS("../results/VDJ_clono_orig.ident_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_orig.ident.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(orig.ident_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "Clonal diversity", 
     legend_title = "Sample")
dev.off()

plotDiversityCurve(orig.ident_alphaDiversity,main_title = "Clonal diversity", 
     legend_title = "Sample") + theme_classic()


rm(orig.ident_alphaDiversity); gc()

####---- per patient group
patient_group_alphaDiversity <- alphaDiversity(gex@meta.data, 
                                   group = "patient_group", 
                                   clone = "cluster_no_airr",
                                   min_q = 0, 
                                   max_q = 4, 
                                   step_q = 0.1,
                                   ci = 0.95, 
                                   nboot = 100)

#saveRDS(patient_group_alphaDiversity, "../results/VDJ_clono_patient_group_alphaDiversity.rds")
patient_group_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_alphaDiversity.rds")
head(patient_group_alphaDiversity@tests)

patient_group_alphaDiversity@tests[patient_group_alphaDiversity@tests$pvalue < 0.1,]

png("../results/VDJ_clono_alphaDiversity_patient_group.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "Clonal diversity", 
     legend_title = "Patient group")
dev.off()

rm(patient_group_alphaDiversity); gc()

####---- per c_call
# remove c_call classes with too few cells
tmp <- gex@meta.data
tmp <- tmp[!is.na(tmp$c_call),]
tmp <- tmp[!(tmp$c_call == "IGHE" | 
             tmp$c_call == "IGLC1" |
             tmp$c_call == "IGLC2"), ]

c_call_alphaDiversity <- alphaDiversity(tmp, 
                                        group = "c_call", 
                                        clone = "cluster_no_airr",
                                        min_q = 0, 
                                        max_q = 4, 
                                        step_q = 0.1,
                                        ci = 0.95, 
                                        nboot = 100)

saveRDS(c_call_alphaDiversity, 
        "../results/VDJ_clono_c_call_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_c_call.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(c_call_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "c call")
dev.off()

#diversity per patient group and celltype
tmp$patient_group_c_call <- paste0(tmp$patient_group, "_", tmp$c_call)
patient_group_c_call_alphaDiversity <- alphaDiversity(tmp, 
                                        group = "patient_group_c_call", 
                                        clone = "cluster_no_airr",
                                        min_q = 0, 
                                        max_q = 4, 
                                        step_q = 0.1,
                                        ci = 0.95, 
                                        nboot = 100)

saveRDS(patient_group_c_call_alphaDiversity, 
        "../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")
patient_group_c_call_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_patient_group_c_call.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_c_call_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "patient group\nc call")
dev.off()

plotDiversityCurve(patient_group_c_call_alphaDiversity) + theme_classic()


#diversity per patient group and celltype
tmp$patient_group_c_call <- paste0(tmp$patient_group, "_", tmp$c_call)
patient_group_c_call_alphaDiversity <- alphaDiversity(tmp, 
                                        group = "patient_group_c_call", 
                                        clone = "cluster_no_airr",
                                        min_q = 0, 
                                        max_q = 4, 
                                        step_q = 0.1,
                                        ci = 0.95, 
                                        nboot = 100)

saveRDS(patient_group_c_call_alphaDiversity, 
        "../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")
patient_group_c_call_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_patient_group_c_call.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_c_call_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "patient group\nc call")
dev.off()

plotDiversityCurve(patient_group_c_call_alphaDiversity) 


####---- Calculate clonal diversity indices
#four samples, 2 DNEG, 2 SSA and 1 SSAB have less repertoir diversity!
#in which celltype?
#in cells expressing which IGHC?

tmp <- gex
tmp$patient_group_c_call <- paste0(tmp$patient_group, "_", tmp$c_call)

combined2 <- expression2List(tmp, split.by = "patient_group")
p <- clonalDiversity(combined2, cloneCall = "cluster_no_airr", group.by = "orig.ident", x.axis = "patient_group") 

p1 <- p + 
  theme_classic() + 
  xlab("") + 
  geom_text(check_overlap = TRUE,
            size = 2.5,
            aes(label = orig.ident)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

ggsave2(paste0("../results/VDJ_clono_diversityIndex_plots.png"),
        p1,
        width = 25, height = 12, unit = "cm")


####---------------------------------------------------------------------
####---- make pie charts
####---------------------------------------------------------------------

imm2 <- gex@meta.data[,c("cluster_no_airr", "orig.ident")]
imm3 <- table(imm2)
imm4 <- data.frame(table(rowSums(imm3 > 0)))
colnames(imm4) <- c("clonocount", "cellcount")
imm4$perc <- round(imm4$cellcount / sum(imm4$cellcount) * 100, digits = 5)


####---- pie1 - all clonotypes
p1 <- ggplot(imm4, aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1)+
  geom_text(aes(label = c(paste0(round(imm4$perc[1], digits = 2), "%"), 
                rep("", times = nrow(imm4) - 1))), 
            position = position_stack(vjust = 0.5, reverse = TRUE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(format(sum(imm4$cellcount), big.mark = " "), " clonotypes")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5))

####---- get legend
p_legend <- ggplot(imm4, aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity", 
           #color = "white"
           ) +
  coord_polar(theta = "y", start = 0)+
  theme_void() +
  theme(legend.position = "bottom", legend.box = "horizontal") +
  scale_fill_discrete(name = "Samples per \nclonotype")

p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

####---- get colors
g <- ggplot_build(p1)
c.p1 <- unique(g$data[[1]]["fill"])

####---- pie2  for all but unique clonotypes
p2 <- ggplot(imm4[2:nrow(imm4), ], 
             aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1)+
  geom_text(aes(label = c(paste0(round(imm4$perc[2:5], digits = 2), "%"), 
                rep("", times = nrow(imm4) - 5))), 
            position = position_stack(vjust = 0.75, reverse = FALSE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(format(sum(imm4$cellcount[2:nrow(imm4)]), big.mark = " "), " clonotypes")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5)
        ) +
  scale_fill_manual(values = c.p1$fill[2:nrow(c.p1)])


####---- pie3  clonotypes present in ≥ 6 samples
p3 <- ggplot(imm4[6:nrow(imm4), ], 
             aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1)+
  geom_text(aes(label = c(paste0(round(imm4$cellcount[6:9], digits = 3)
                                 #, "%"
                                 ), 
                rep("", times = nrow(imm4) - 9))), 
            position = position_stack(vjust = 0.6, reverse = FALSE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(sum(imm4$cellcount[6:nrow(imm4)]), " clonotypes\npresent in ≥ 6 samples")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5)) +
  scale_fill_manual(values = c.p1$fill[6:nrow(c.p1)])

p4 <- plot_grid(p1, p2, p3, nrow = 1, labels = NULL,
                align = "vh", axis = "l")

####---- save pies
ggsave2(paste0("../results/VDJ_clono_clonotypeDistribution_perSample_2.png"),
          plot_grid(p4, p_legend, 
                    ncol = 1, 
                    labels = NULL, 
                    align = "v", 
                    rel_heights = c(5, 2)) + bgcolor("white"),
          width = 20, height = 12, unit = "cm")


####--------------------------------------------------------------------- 
####---- get clonotypes present in ≥ 6 samples using output from compairr
####---------------------------------------------------------------------
pheno_2 <- unique(gex@meta.data[, c("orig.ident", "patient_group")])
annot_junc <- unique(gex@meta.data[, c("cluster_no_airr", "junction_aa")])

tmp <- as.data.frame.matrix(table(gex@meta.data$cluster_no_airr,
                                  gex@meta.data$orig.ident))

tmp <- tmp[rowSums(tmp > 0) > 5, ]
tmp <- tmp[, order(pheno_2$patient_group[match(colnames(tmp), 
                                               pheno_2$orig.ident)]) ]

pheatmap(tmp,
         cluster_cols = FALSE,
         cluster_rows = TRUE,
         #border_color = NA,
         annotation_col = data.frame(patient_group = factor(pheno_2$patient_group[match(colnames(tmp), pheno_2$orig.ident)]), row.names = colnames(tmp)),
         clustering_method = "ward.D2",
         color = c("gray95", colorRampPalette(c("gray70", 
                                                "orange3", 
                                                "firebrick"))(90)),
         fontsize_row = 6,
         fontsize_col = 8,
         main = paste0(nrow(tmp), " clonotypes occuring in ≥ 6 samples"),
         labels_row = annot_junc$junction_aa[match(rownames(tmp), 
                                                   annot_junc$cluster_no_airr)],
         filename = "../results/VDJ_clono_clonotypeDistribution_perSample_topClonotypes.png"
         )  


####---------------------------------------------------------------------
####---- clonotype correlation using cluster_no_airr
####---------------------------------------------------------------------
tmp <- gex@meta.data
tmp$celltype_patient_group <- paste0(tmp$cluster_cellType_manual_main, 
                                     "-", 
                                     tmp$patient_group)

tmp2 <- table(list(tmp$cluster_no_airr, 
                    tmp$celltype_patient_group))
  
  cor_tmp <- cor(tmp2, method = "spearman")
  cor_tmp[is.na(cor_tmp)] <- 0
  
p <- pheatmap::pheatmap(cor_tmp, 
                     annotation_col = data.frame(cell_type = factor(sub("-.*", "", colnames(cor_tmp))),
                                                 patient_group = factor(sub(".*-", "", colnames(cor_tmp))),
                                                  row.names = colnames(cor_tmp)),
                      breaks = seq(0, 1, length.out = 100),
                      filename = paste0("../results/VDJ_clono_spearman_per_cluster_PatientGroup.png"),
                      fontsize = 8, 
                      fontsize_row = 4, 
                      fontsize_col = 4,
                      border_color = NA,
                      main = paste("clonotype spearman correlation"),
                      color = colorRampPalette(c("lightblue", "gray30","firebrick2", "red"))(99))























StartracDiversity(gex, type = "Type", 
            sample = "Patient", by = "overall")

#tmp <- gex$junction_aa
tmp <- gex$cdr3
names(tmp) <- colnames(gex); head(tmp)

junc_dist <- pairwiseDist(tmp, dist_mat = getDNAMatrix(gap = 0))
#Error: vector memory exhausted (limit reached?)

cdr3_eq <- pairwiseEqual(tmp)

seq <- c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C", E="NTGGG")
d <- pairwiseEqual(seq)
rownames(d) <- colnames(d) <- seq
d

clones2 <- countClones(gex@meta.data, 
                      groups = c("patient_group", "cluster_cellType_manual_main"), 
                      clone = "cluster_no_airr")

### Quantify clonotypes scRepertoire

quantContig(
  combined_B,
  cloneCall = "gene+nt",
  chain = "both",
  scale = TRUE,
  group = "patient_group",
  exportTable = FALSE
)
head(combined_B)
quantContig(combined_B, cloneCall="gene+nt", scale = T)

abundanceContig(combined_B, cloneCall = "gene", scale = F)
lengthContig(combined_B, cloneCall="aa", chain = "IGL") 

clonalProportion(combined_B, cloneCall = "gene") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

occupiedscRepertoire(gex, x.axis = "seurat_clusters")

clonalOverlap(combined_B, cloneCall = "gene+nt", 
              method = "morisita")

combined2 <- expression2List(gex, split.by = "cluster_cellType_manual_fine")
combined2 <- expression2List(gex, split.by = "cluster_cellType_manual_main")
str(combined2)

clonalHomeostasis(combined2, cloneCall = "nt") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

clonalProportion(combined2, cloneCall = "nt") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

clonalDiversity(gex, cloneCall = "nt",group.by = "cluster_cellType_manual_fine")


clonalOverlap(combined2, cloneCall="aa", method="overlap") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

alluvialClonotypes(gex, cloneCall = "gene", 
                   y.axes = c("orig.ident", "seurat_clusters"), 
                   color = "seurat_clusters")

p1 <- alluvialClonotypes(gex, cloneCall = "gene", 
                   y.axes = c("orig.ident", "cluster_cellType_manual_main"), 
                   color = "cluster_cellType_manual_main")
ggsave2("../results/alluvialClonotypes_test.png", p1, height = 20, width = 30, units = "cm")

#results <- hierarchicalClones(imm, threshold=0.1)


### clonotype overlap correlation across samples and patient groups using compairr
cairr.aa.x <- as.data.frame.matrix(table(cairr.aa$cluster_no, cairr.aa$repertoire_id))
cairr.aa.x <- cairr.aa.x[apply(cairr.aa.x, 1, function(x){sum(x >= 1) >= 2}), ]
cairr.aa.x <- cairr.aa.x[order(-rowSums(cairr.aa.x)),]





```


#### mutation status per IGHV gene 

```{r}

vdj_gene_mu <- gex@meta.data
vdj_gene_mu$v_gene <- sub("\\*.*", "", vdj_gene_mu$v_call)

vdj_gene_mu  <- vdj_gene_mu %>%  
  group_by(v_gene, patient_group, orig.ident, cluster_cellType_manual_main) %>% 
  summarise(mu_mean = mean(mu_freq_total), n = n()) 

vdj_gene_mu <- vdj_gene_mu[vdj_gene_mu$cluster_cellType_manual_main == "DN" |
                           vdj_gene_mu$cluster_cellType_manual_main ==  "Memory" |
                           vdj_gene_mu$cluster_cellType_manual_main ==  "Naive" |
                           vdj_gene_mu$cluster_cellType_manual_main ==  "Transitional",]

vdj_gene_mu$patient_group_gene <- paste(vdj_gene_mu$patient_group, 
                                        vdj_gene_mu$v_gene, sep = "_")
vdj_gene$patient_group_gene <- paste(vdj_gene$patient_group, 
                                     vdj_gene$gene, sep = "_")

vdj <- left_join(vdj_gene_mu, vdj_gene, by = "patient_group_gene")
vdj <- vdj[which(vdj$gene %in% unique(vdj_gene$gene)[1:30]), ]

p <- ggplot(vdj, aes(x = reorder(v_gene, -seq_freq), 
                    y = mu_mean, 
                    fill = patient_group.x)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, 
                                     hjust = 1, 
                                     vjust = 1)) +
    scale_y_continuous(labels = percent) +
    ylab("mutated bp (%)") +
    xlab("") +
    geom_boxplot(position = "dodge", ) + 
      facet_grid(rows = vars(cluster_cellType_manual_main)) +
    scale_fill_discrete(name = "patient group")
  
    
 ggsave2(paste0("../results/VDJ_mu_per_patGroup_perCellType_test.png"),
          p ,
          width = 25, height = 15, unit = "cm")
 
  
```


### plot mutation status per celltypes and samples

```{r}

gex.meta <- gex@meta.data[gex$cluster_cellType_manual_main == "Memory" | 
                          gex$cluster_cellType_manual_main == "Naive" |
                          gex$cluster_cellType_manual_main == "DN" |
                          gex$cluster_cellType_manual_main == "Transitional", ]

####---- CDR3 mutation frequency - cell type fine - patient_group, 
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr3_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR3 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr3_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR3 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muCDR3_CellTypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- CDR3 mutation frequency - celltype main - patient group
# split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr3_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 mutations - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr3_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 mutations - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muCDR3_CelltypeMain_PatientGroup_rs.png"), 
          plot_grid(p1, p2, ncol = 2, labels = "AUTO", align = "h"),
          width = 25, height = 10, unit = "cm")


####---- CDR3 mutation frequency - celltypeMain - orig.ident 
# split by replacement/silent
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr3_r,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 mutations - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr3_s,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 mutations - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muCDR3_CelltypeMain_orig.ident_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = "AUTO", align = "h"),
          width = 25, height = 15, unit = "cm")


####---- CDR3 mutation count - celltypeMain - patient group
# split by replacement/silent

# !!!! FIX THIS !!!

as_tibble(gex.meta) %>% 
  group_by(cluster_cellType_manual_main, 
                                 patient_group, 
                                 mu_count_cdr3_r) %>% 
  
  summarise(cdr3_r = n() ) %>% 
  mutate(cdr3_r = cdr3_r/sum(cdr3_r)) %>%
  ggplot(aes(x = mu_count_cdr3_r, 
             y = cdr3_r, 
             fill = patient_group)) +
  theme_classic() +
  geom_col(position = "dodge") +
  scale_x_continuous(breaks = c(0:7)) +
  scale_y_continuous(labels = percent) +
  xlab("number of mutated CDR3 bp") + 
  ylab("fraction (%)") +
  facet_grid(cols = vars(cluster_cellType_manual_main))

colnames(gex@meta.data)


####---- VDJ mutation frequency up until bp 312 - celltypeFine - patient_group
p <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_v_total, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) 

ggsave2(paste0("../results/VDJ_mu312_CellTypeFine_patientGroup.png"), 
          p,
          width = 30, height = 10, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeFine - patient_group
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_v_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_v_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_CellTypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeMain - patient_group
p <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_v_total, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_celltype_patientGroup.png"), 
          p,
          width = 20, height = 8, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeMain - orig.ident
p <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_v_total,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_celltypeMain_orig.ident.png"), 
          p,
          width = 20, height = 8, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeMain - patient_group
# split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_v_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_v_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_celltypeMain_patientGroup_rs.png"), 
          plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"),
          width = 25, height = 10, unit = "cm")


####---- VDJ mutation frequency for CDR1-3 - celltypeMain - patient_group 
#split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_CDR_celltypeMain_patientGroup_rs.png"), 
          plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"),
          width = 25, height = 10, unit = "cm")


####---- VDJ mutation frequency for FWR1-4 celltypeMain - patient_group 
# split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_fwr_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_fwr_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_FWR_byCelltype_rs_perPatientGroup.png"), 
          plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"),
          width = 25, height = 10, unit = "cm")


####----  total VDJ mutation frequency for CDR1-3 - celltypeMain -  orig.ident 
# split by replacement/silent
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr_r,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr_s,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_CDR_celltypeMain_orig.ident_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 25, height = 18, unit = "cm")


####----  total VDJ mutation frequency for FWR1-4 - celltypeMain - orig.ident 
# split by replacement/silent
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_fwr_r,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_fwr_s,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_FWR_celltypeMain_orig.ident_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 25, height = 18, unit = "cm")


####---- VDJ mutation frequency for CDR1-3 - celltypeFine - patient_group
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_CDR_celltypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- total VDJ mutation frequency for FWR1-4  
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_fwr_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_fwr_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_FWR_celltypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- total VDJ mutation frequency - celltypeMain - patient_group
p <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_total, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_celltypeMain_patientGroup.png"), 
          p,
          width = 20, height = 6, unit = "cm")


####---- total VDJ mutation frequency - celltypeMain - orig.ident
p <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_total,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_celltypeMain_orig.ident.png"), 
          p,
          width = 30, height = 10, unit = "cm")

####---- total VDJ mutation frequency - celltypeFine - patient_group
p <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) 

ggsave2(paste0("../results/VDJ_muFull_celltypeFine_patient_group.png"), 
          p,
          width = 30, height = 10, unit = "cm")



gex.meta <- gex@meta.data[gex$cluster_cellType_manual_fine == "DN1_ID3" | 
                          gex$cluster_cellType_manual_fine == "DN2" |
                          gex$cluster_cellType_manual_fine == "DN2_CXCR3" |
                          gex$cluster_cellType_manual_fine == "DN2_ITGAX" |
                          gex$cluster_cellType_manual_fine == "DN4_IGHE" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                          gex$cluster_cellType_manual_fine == "Naive" |
                          gex$cluster_cellType_manual_fine == "Naive_IFN" |
                          gex$cluster_cellType_manual_fine == "Naive_Transitional", ]

####---- total VDJ mutation frequency - celltypeFine - orig.ident
p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
p[[i]] <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ], 
             aes(x = patient_group, 
                 y = mu_freq_total, 
                 fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1),
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      ggtitle(i) +
      scale_y_continuous(labels = percent) +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) 
}


ggsave2(paste0("../results/VDJ_muFull_celltypeFine_orig.ident_2.png"), 
          plot_grid(plotlist = p, ncol = 4, labels = NULL, align = "hv") + bgcolor("white"),
          width = 30, height = 30, unit = "cm")



#####---- percent mutated cells per seurat cluster
cluster_mut <- as.data.frame.matrix(table(gex$seurat_clusters,
                                          gex$mu_load))
cluster_mut$percmut <- 1 - cluster_mut$unmutated/rowSums(cluster_mut)
cluster_mut$orig.ident <- factor(rownames(cluster_mut), 
                                 levels = rownames(cluster_mut))

p <- ggplot(cluster_mut, aes(x = orig.ident, y = percmut)) +
      theme_classic() +
      ggtitle("perc cells with mutated VH per cluster") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("perc mutated cells") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_bar(stat='identity')

ggsave2(paste0("../results/VDJ_percMutCells_byCluster.png"), 
          p,
          width = 30, height = 20, unit = "cm")


####---- UMAP colored by categorical muLoad
p <- DimPlot(gex, reduction = "umap_2", 
                  pt.size = .01, 
                  group.by = "mu_load", 
                  raster = FALSE) + 
          ggtitle("Mutation load") +
          ylab("UMAP 2") +
          xlab("UMAP 1")

ggsave2(paste0("../results/UMAP_B-cells_muLoad_1.png"), 
          p,
          width = 25, height = 20, unit = "cm")

```


###  plot CDR3 aa attributes

```{r}
# "cdr3_aa_length"               
# "cdr3_aa_gravy"               
# "cdr3_aa_bulk"                 
# "cdr3_aa_aliphatic"
# "cdr3_aa_polarity"             
# "cdr3_aa_charge"              
# "cdr3_aa_basic"               
# "cdr3_aa_acidic"
# "cdr3_aa_aromatic"


#### ----- per cell type
#subeset gex to only include relevant cell types

gex.meta <- gex@meta.data[gex$cluster_cellType_manual_main == "Memory" | 
                          gex$cluster_cellType_manual_main == "Naive" |
                          gex$cluster_cellType_manual_main == "DN" |
                          gex$cluster_cellType_manual_main == "Transitional" , ]

#CDR3 aa length
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_length,
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa length") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("length (bp)") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#grand average of hydrophobicity
p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_gravy, #grand average of hydrophobicity
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa grand average of hydrophobicity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("hydrophobicity score") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#average bulkiness of amino acids
p3 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_bulk, #average bulkiness of amino acids
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa bulkiness") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average bulkiness of amino acids") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aliphatic
p4 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_aliphatic, #aliphatic index
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa aliphatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("aliphatic index") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_polarity
p5 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_polarity, #average polarity of amino acids
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa polarity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average polarity") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_charge
p6 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_charge, #net charge
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 charge") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("net charge") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_basic
p7 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_basic, #fract. of informative positions that are Arg, His or Lys
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa basic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos that are Arg, His or Lys") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_acidic
p8 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_acidic, #fraction of informative positions that are Asp or Glu
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa acidic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are Asp or Glu") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aromatic
p9 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_aromatic, #fract. of informative pos. that are His, Phe, Trp or Tyr
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa aromatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are His, Phe, Trp or Tyr") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)

ggsave2(paste0("../results/VDJ_attrCDR3_byCelltype_perSample_2.png"), 
          plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, p9, 
                    nrow = 3, ncol = 3, labels = "AUTO", align = "h"),
          width = 40, height = 30, unit = "cm")


#### ----- per cluster
#CDR3 aa length
p1 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_length,
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa length") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("length (bp)") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#grand average of hydrophobicity
p2 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_gravy,
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa grand average of hydrophobicity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("hydrophobicity score") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#average bulkiness of amino acids
p3 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_bulk, #average bulkiness of amino acids
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa bulkiness") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average bulkiness of amino acids") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aliphatic
p4 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aliphatic, #aliphatic index
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aliphatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("aliphatic index") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_polarity
p5 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_polarity, #average polarity of amino acids
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa polarity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average polarity") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_charge
p6 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_charge, #net charge
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa charge") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("net charge") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)

#cdr3_aa_basic
p7 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_basic, #fract. of informative positions that are Arg, His or Lys
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa basic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos that are Arg, His or Lys") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_acidic
p8 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_acidic, #fraction of informative positions that are Asp or Glu
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa acidic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are Asp or Glu") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aromatic
p9 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic, #fract. of informative pos. that are His, Phe, Trp or Tyr
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aromatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are His, Phe, Trp or Tyr") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)

#get legend
p_legend <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic,
                  color = patient_group)) +
      theme_classic() +
      geom_boxplot() 
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()


p_grid <- plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, p9, 
                    nrow = 3, ncol = 3, labels = "AUTO", align = "h")

ggsave2(paste0("../results/VDJ_attrCDR3_byCluster_perSample_2.png"), 
          plot_grid(p_grid, p_legend, ncol = 2, align = "none", rel_widths = c(12, 1)),
          width = 45, height = 30, unit = "cm")


```


###  plot IgH class, mu_load and clonetype per cluster

```{r}

####---- IgH per cluster
p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(9, "Set3"), "grey")) + 
              theme(legend.position = "none") + ggtitle("IGH c gene") + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank())

p2 <- ggplot(gex@meta.data, aes(x = cluster_cellType_manual_fine, fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") +
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      scale_fill_brewer(palette="Set3", na.value="grey", name = "c gene")

ggsave2("../results/cluster_c_gene_10x_abundance.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h") + bgcolor("white"), 
        width = 30, height = 14, unit = "cm")

p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", 
              ncol = 2, 
              raster = FALSE) + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank()) + 
  ggtitle("IGH c gene")
  
ggsave2("../results/UMAP_B-cells_IgHclass.png", 
        p1, 
        width = 30, height = 20, unit = "cm")


##---- mutation load per cluster
p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(4, "Dark2"), "grey")) +
              theme(legend.position = "none") +
              ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank())

p2 <- ggplot(gex@meta.data, aes(x = cluster_cellType_manual_fine, fill = mu_load)) +
          theme_classic() +
          geom_bar(position = "fill") + 
          ylab("proportion of cells") +
          xlab("") +
          facet_grid(rows = vars(patient_group)) + 
          theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
          scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "Mutation load")

ggsave2("../results/cluster_mu_load_abundance.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"), 
        width = 30, height = 14, unit = "cm")

p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", 
              ncol = 2, 
              raster = FALSE) +
              ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank())
              
ggsave2("../results/UMAP_B-cells_muLoad.png", 
        p1, 
        width = 30, height = 20, unit = "cm")


##---- cloneType per cluster based on compairr data

p1 <- DimPlot(gex, 
              group.by = "cloneType_airr", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(4, "Dark2"), "grey")) +
              theme(legend.position = "none") +
  ggtitle("clonotype") +  
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank())
table(gex$patient_group, useNA = "always")

p2 <- ggplot(gex@meta.data, aes(x = cluster_cellType_manual_fine, 
                                fill = cloneType_airr)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") + 
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      scale_fill_brewer(palette="Dark2", na.value = "grey", name = "clonotype")

ggsave2("../results/cluster_cloneType_abundance.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"), 
        width = 30, height = 14, unit = "cm")


```


## Save rds GEX_BCELLS

```{r}

saveRDS(gex, file = "../results/GEX7_BVDJ.rds")
#gex <- readRDS("../results/GEX7_BVDJ.rds")

```


## Print sessionInfo()

```{r}

sessionInfo()

```



