---
output: html_document
editor_options: 
  chunk_output_type: console
---

# VDJ analysis {#VDJ}
## Load packages

```{r}

source("./00_dependencies.R")

```


## Load Seurat object

```{r}

gex <- readRDS("../results/GEX6_BANNOT.rds"); gex

```


## Load pheno file

```{r}

pheno <- read.table("../suppl/pheno/pSS_pheno_210623.csv", sep = ";", header = TRUE)
colnames(pheno)[c(1, 3)] <- c("Sample_id", "orig.ident")
pheno$orig.ident <- gsub("P007[ab]", "P007", pheno$orig.ident)
pheno$orig.ident <- gsub("P008[ab]", "P008", pheno$orig.ident)
pheno <- unique(pheno[, c(1, 3, 5, 7:8, 11:12)])

pheno$patient_group <- paste0(ifelse(pheno$SSA == "YES", "SSA", ""), 
                            "_", 
                            ifelse(pheno$SSB == "YES", "SSB", ""))
pat_annot <- c(SSAB = "SSA_SSB", SSA = "SSA_", DNEG="NA_NA", DNEG="_")
pheno$patient_group <- names(pat_annot)[match(pheno$patient_group, pat_annot)]
pheno$patient_group[grep("^C00", pheno$orig.ident) ] <- "CTRL"

head(pheno); dim(pheno)

#Four non-overlapping groups
#1. SSAB - SSA+/SSB+
#2. SSA  - SSA+/SSB-
#3. DNEG - SSA-/SSB-
#4. CTRL - CTRL


```


## immcantation
### Load output from Immcantation and perform mutation analysis

```{r}

files_B <- list.files("../data/clono_B_VDJ_tmp", 
                      pattern = "*germ-pass.tsv", 
                      recursive = TRUE, 
                      full.names = TRUE)
ncpu <- cpuCount()
imm <- NULL
for (i in files_B) {
  
  cat("reading: ", print(i), "\n")
  tmp <- read_rearrangement(i)
  
  ####---- get sample name
  j <- gsub("_heavy_germ-pass.tsv", "", gsub(".*out/", "", i))
  cat("sample name: ", print(j), "\n")
  
  ####---- number of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = FALSE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_v_total'
  
  ####---- frequency of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = TRUE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_v_total'
  
  ####---- number of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")

  ####---- number of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
   
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_total'
  
  ####---- frequency of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_total'
 
  ####---- properties for CDR3 aa's
  tmp <- aminoAcidProperties(tmp, 
                             seq = "junction", 
                             nt = TRUE, 
                             trim = TRUE, 
                             label = "cdr3")
  
  tmp$cell_id <- paste0(j, "_", gsub("-1", "", tmp$cell_id))
  tmp$orig.ident <- gsub("a|b", "", gsub("_.*", "", tmp$cell_id))
  tmp$sequence_id <- paste0(j, "_", gsub("-1", "", tmp$sequence_id))
  rownames(tmp) <- tmp$cell_id
  
  if (is.null(imm)) {
    imm <- tmp
  } else {
    imm <- rbind(imm, tmp)
  }
  
}

head(imm)


```


### integrate immcantation VDJ data with seurat object

```{r}

#select immcantation columns
# cols <- c("sequence_id", "productive", "vj_in_frame", "stop_codon", "locus", 
#           "v_call", "d_call", "j_call", "c_call", 
#           "cell_id", "clone_id", "umi_count", 
#           "v_call_10x", "d_call_10x", "j_call_10x", 
#           "germline_v_call","germline_d_call", "germline_j_call",
#           "cdr3", "consensus_count",
#           "mu_count_total", "mu_freq_total", 
#           "mu_count_v_r_total", "mu_count_v_s_total", "mu_freq_v_r_total", "mu_freq_v_s_total",
#           "mu_freq_cdr1_r", "mu_freq_cdr1_s", "mu_freq_cdr2_r", "mu_freq_cdr2_s", "mu_freq_cdr3_r", "mu_freq_cdr3_s", "mu_freq_fwr1_r", "mu_freq_fwr1_s", "mu_freq_fwr2_r", "mu_freq_fwr2_s", "mu_freq_fwr3_r", "mu_freq_fwr3_s", "mu_freq_fwr4_r", "mu_freq_fwr4_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "cdr3_aa_length", "cdr3_aa_gravy", "cdr3_aa_bulk", "cdr3_aa_aliphatic", "cdr3_aa_polarity", "cdr3_aa_charge", "cdr3_aa_basic", "cdr3_aa_acidic", "cdr3_aa_aromatic")

colnames(imm)
#select immcantation columns not to be added to gex
cols <- c("sequence", "sequence_alignment", "germline_alignment", 
          "v_cigar", "d_cigar", "j_cigar",
          "fwr1", "fwr2", "fwr3", "fwr4", "cdr1", "cdr2", 
          "germline_alignment_d_mask")

imm_add <- imm[, !(colnames(imm) %in% cols)]
rownames(imm_add) <- imm_add$cell_id

gex <- AddMetaData(gex, imm_add)
colnames(gex@meta.data)


```


### add pheno data to immcantation object

```{r}

#imm$orig.ident <- gsub("a|b", "", gsub("_.*", "", imm$cell_id))
imm <- merge(imm, 
                 pheno[, c(2,8)], 
                 by.x = "orig.ident", 
                 by.y = "orig.ident", 
                 all.x = TRUE); head(imm)
imm <- imm[, c(2:ncol(imm), 1)]


```


### save immcantation VDJ data

```{r}

colnames(imm)[colnames(imm) == 'orig.ident'] <- 'repertoire_id'

write_rearrangement(imm, "../results/VDJ_immcantation_out.tsv")
#imm <- read_rearrangement("../results/VDJ_immcantation_out.tsv")


```


### clonal distance threshold estimation

```{r}

imm <- distToNearest(imm, model = "ham", 
                    normalize = "len", 
                    vCallColumn = "v_call",
                    progress = TRUE,
                    nproc = 4)
colnames(imm)
# Determine threshold
threshold <- findThreshold(imm$dist_nearest, method = "density")

p1 <- ggplot(data = subset(imm,!is.na(dist_nearest))) +
  theme_bw() +
  ggtitle("Distance to nearest: Hamming") +
  xlab("distance") +
  geom_histogram(
    aes(x = dist_nearest),
    binwidth = 0.025,
    fill = "steelblue",
    color = "white"
  )
plot(p1)

# plot the distribution
plot(threshold)

thr <- round(threshold@threshold, 2)
thr



imm2 <- imm
imm2 <- imm2[!(colnames(imm2) %in% c('clone_id', 'vj_group', 'vjl_group', 'junction_l', 'cdr3_col', 'clone_temp'))]

results <- identicalClones(imm2, 
                           method = "nt",
                           nproc = 4,
                           verbose = TRUE,
                           cell_id = "cell_id",
                           cdr3 = TRUE)

results_db <- as.data.frame(results)
glimpse(results_db)
plot(results, binwidth=0.02)
glimpse(summary(results))

results2 <- hierarchicalClones(imm2, threshold = 0.15)
results_db2 <- as.data.frame(results2)
glimpse(results_db2)
plot(results2, binwidth=0.02)
glimpse(summary(results2))



```


### VDJ mutation
#### add categorical variable for total mutation burden

```{r}

# unmutated
# low < 1%
# med < 2%
# high > 2%

gex$mu_load <- ifelse(gex$mu_freq_total == 0, "unmutated", 
                 ifelse(gex$mu_freq_total > 0 & gex$mu_freq_total <= 0.01, "low",
                        ifelse(gex$mu_freq_total > 0.01 & gex$mu_freq_total <= 0.02, "med",
                               ifelse(gex$mu_freq_total > 0.02, "high", NA))))


```


## scRepertiore BCR
### Read 10x BCR files

```{r}

# read 10x VDJ BCR files
samples_B <- list.files("../data/clono_B_VDJ")

clono_B <- lapply(samples_B, function(x){
  
  tmp <- read.csv(paste0("../data/clono_B_VDJ/", x, "/", x, "_B_filtered_contig_annotations.csv"), 
                  stringsAsFactors = FALSE)
  return(tmp)
  
})

names(clono_B) <- samples_B

#remove trailing -1
clono_B <- lapply(seq_along(clono_B), function(i){
  clono_B[[i]]$barcode <- gsub("-1", "", clono_B[[i]]$barcode)
  return(clono_B[[i]])
})

names(clono_B) <- samples_B

clono_B <- clono_B[order(names(clono_B))]

print(object.size(clono_B), units = "MB") #227.6 Mb
saveRDS(clono_B, paste0("../results/clono_B_list.rds"))
#clono_B <- readRDS("../results/clono_B_list.rds")


```


### generate a combined data frame for BCR sequences using scRepertoire

```{r}

#split combineBCR and do it per sample to avoid memory issues
#takes on average 20min/sample to run
combined_B <- lapply(1:length(clono_B), function(x) {
  
  print(names(clono_B[x]))
  print(nrow(clono_B[[x]]))
  print(Sys.time())
  
  tmp <- combineBCR(clono_B[[x]],
                    samples = names(clono_B[x]),
                    ID = NULL,
                    removeNA = TRUE,
                    removeMulti = TRUE)

  return(tmp)
  
})

#reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
combined_B <- lapply(1:length(combined_B), function(x){
  tmp <- combined_B[[x]][[1]]
  return(tmp)
})

#---- handle samples where libraries were created and sequenced twice
combined_B[["P007"]] <- rbind(combined_B[["P007a"]], combined_B[["P007b"]])
combined_B[["P007a"]] <- NULL
combined_B[["P007b"]] <- NULL

combined_B[["P008"]] <- rbind(combined_B[["P008a"]], combined_B[["P008b"]])
combined_B[["P008a"]] <- NULL
combined_B[["P008b"]] <- NULL

combined_B <- combined_B[order(names(combined_B))]

names_tmp <- names(combined_B)
combined_B <- lapply(seq_along(combined_B), function(i) {
  combined_B[[i]]$sample <- gsub("a|b", "", combined_B[[i]]$sample)
  return(combined_B[[i]])
})

names(combined_B) <- names_tmp
#----

print(object.size(combined_B), units = "MB") #181.3 Mb
saveRDS(combined_B, paste0("../results/combined_BCR_scRepertoire.rds"))
#combined_B <- readRDS("../results/combined_BCR_scRepertoire.rds")


```


### Integrate scRepertoir out with Seurat object

```{r}

gex <- combineExpression(combined_B, gex)
colnames(gex@meta.data)


```


## add 10x IgH call to Seurat object

```{r}

igh_10x <- lapply(1:length(clono_B), function(x) {
  
  tmp <- clono_B[[x]][clono_B[[x]]$chain == "IGH" & 
                        clono_B[[x]]$productive == "true" & 
                        !is.na(clono_B[[x]]$c_gene) & 
                        clono_B[[x]]$c_gene != "" &
                        clono_B[[x]]$c_gene != "IGLC1" &
                        clono_B[[x]]$c_gene != "IGLC2" , c(1, 10)]
  tmp$barcode <- paste0(names(clono_B)[x], "_", tmp$barcode)
  names(tmp)[2] <- "c_gene_10x"

  tmp <- tmp[!duplicated(tmp$barcode), ]
  
  rownames(tmp) <- tmp$barcode
  tmp$barcode <- NULL
  
  return(tmp)
  }
)

igh_10x <- do.call(rbind.data.frame, igh_10x)
table(igh_10x$c_gene_10x, useNA = "always")
gex <- AddMetaData(gex, igh_10x)


```



## TRUST4
### read TRUST4 files

```{r}

# samples_TRUST4 <- sub("_barcode_report.tsv", "", list.files("../data/clono_TRUST4"))
# clono_TRUST4 <- lapply(samples_TRUST4, function(x){
#   tmp <- read.table(paste0("../data/clono_TRUST4/", x, "_barcode_report.tsv"), header = TRUE, comment.char = "")
#   colnames(tmp) <- sub("X." , "", colnames(tmp))
#   return(tmp)
# })
# 
# names(clono_TRUST4) <- samples_TRUST4
# 
# # remove trailing -1
# # combineTRUST4 adds sample name to barcodes as prefix
# clono_TRUST4 <- lapply(seq_along(clono_TRUST4), function(i) {
#   clono_TRUST4[[i]][, 1] <- gsub("-1", "", clono_TRUST4[[i]][, 1])
#   print(names(clono_TRUST4[i]))
#   return(clono_TRUST4[[i]])
# })
# 
# names(clono_TRUST4) <- samples_TRUST4
# 
# saveRDS(clono_TRUST4, paste0("../results/clono_TRUST4_list.rds"))
# #clono_TRUST4 <- readRDS("../results/clono_TRUST4_list.rds")


```


### combine TRUST4 data using scRepertoire

```{r}

# combined_TRUST4_B <- lapply(1:(length(clono_TRUST4)-1), function(x) {
#   
#   print(names(clono_TRUST4[x]))
#   print(nrow(clono_TRUST4[[x]]))
#   print(Sys.time())
#   
#   tmp <- combineTRUST4(clono_TRUST4[[x]],
#                     samples = names(clono_TRUST4[x]),
#                     ID = NULL,
#                     cells = c("B"),
#                     removeNA = FALSE)
# 
#   return(tmp)
#   
# })
# 
# #reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
# combined_TRUST4_B <- lapply(1:length(combined_TRUST4_B), function(x){
#   tmp <- combined_TRUST4_B[[x]][[1]]
#   return(tmp)
# })
# 
# 
# names(combined_TRUST4_B) <- names(clono_TRUST4)[1:(length(clono_TRUST4)-1)]
# 
# #---- handle samples where libraries were created and sequenced twice
# combined_TRUST4_B[["P007"]] <- rbind(combined_TRUST4_B[["P007a"]], combined_TRUST4_B[["P007b"]])
# combined_TRUST4_B[["P007a"]] <- NULL
# combined_TRUST4_B[["P007b"]] <- NULL
# 
# combined_TRUST4_B[["P008"]] <- rbind(combined_TRUST4_B[["P008a"]], combined_TRUST4_B[["P008b"]])
# combined_TRUST4_B[["P008a"]] <- NULL
# combined_TRUST4_B[["P008b"]] <- NULL
# 
# combined_TRUST4_B <- combined_TRUST4_B[order(names(combined_TRUST4_B))]
# 
# names_tmp <- names(combined_TRUST4_B)
# combined_TRUST4_B <- lapply(seq_along(combined_TRUST4_B), function(i) {
#   combined_TRUST4_B[[i]]$sample <- gsub("a|b", "", combined_TRUST4_B[[i]]$sample)
#   return(combined_TRUST4_B[[i]])
# })
# 
# names(combined_TRUST4_B) <- names_tmp
# #----
# 
# saveRDS(combined_TRUST4_B, paste0("../results/combined_TRUST4_B_scRepertoire.rds"))
# #combined_TRUST4_B <- readRDS("../results/combined_TRUST4_B_scRepertoire.rds")


```


### Integrate TRUST4 out with Seurat object

```{r}

# gex <- combineExpression(combined_TRUST4_B, gex)
# colnames(gex@meta.data)

```


## Analyse and visualize immcantation and scRepertoir output
### VDJ gene usage
#### VDJ correlation

```{r}

#VDJ gene usage correlation
VDJ_cor <- function(i, #seurat object containing meta data for VDJ gene usage
                    correl, #which correlation to use i.e. spearman or pearson...
                    VDJgene #VDJ gene to plot
                    ){
  
  cat(VDJgene, "\n") 
  cat(correl, "\n")
  
  i@meta.data[, VDJgene] <- unlist(lapply(strsplit(i@meta.data[, VDJgene], split = ","),
                                          function(x) {
  ifelse(length(x) > 1 , 
         return(x[1]), 
         return(x))})) #keep only first allele if several exist
  
  #i$seurat_clusters_patient_group <-paste0(i$seurat_clusters, "_", i$patient_group)
  i$celltype_patient_group <-paste0(i$cluster_cellType_manual_fine, "-", i$patient_group)
  tmp <- table(list(i@meta.data[, VDJgene], i$celltype_patient_group))
  
  cor_tmp <- cor(tmp, method = correl)
  cor_tmp[is.na(cor_tmp)] <- 0

  p <- pheatmap::pheatmap(cor_tmp, 
                     annotation_col = data.frame(cell_type = factor(sub("-.*", "", colnames(cor_tmp))),
                                                 patient_group = factor(sub(".*-", "", colnames(cor_tmp))),
                                                  row.names = colnames(cor_tmp)),
                      breaks = seq(0, 1, length.out = 100),
                      filename = paste0("../results/VDJ_Cor_", VDJgene, "_", 
                                        correl, "_per_cluster_PatientGroup.png"),
                      fontsize = 8, 
                      fontsize_row = 4, 
                      fontsize_col = 4,
                      border_color = NA,
                      main = paste(VDJgene, correl, "correlation"),
                      color = colorRampPalette(c("lightblue", "gray90", "firebrick2"))(99))
  
}

VDJ_cor(gex, "spearman", "v_call")
VDJ_cor(gex, "pearson", "v_call")

VDJ_cor(gex, "spearman", "d_call")
VDJ_cor(gex, "pearson", "d_call")

VDJ_cor(gex, "spearman", "j_call")
VDJ_cor(gex, "pearson", "j_call")

VDJ_cor(gex, "spearman", "c_call")
VDJ_cor(gex, "pearson", "c_call")


```


#### top VDJ genes

```{r}

# get most used VDJ genes per patient subgroup

get_topVDJ <- function(i, #seurat object
                       VDJgene = "tmp" #VDJ gene to plot
  ) {
  
  ## !!! handle this better !!!
  vh_col <- unlist(lapply(strsplit(i@meta.data[, VDJgene], split = ","), function(x) {
  ifelse(length(x) > 1 , return(x[1]), return(x))})) #get only first allele if several exist
  
  vh <- as.data.frame.matrix(table(vh_col, i$patient_group))
  vh <- vh[order(rowSums(vh), decreasing = TRUE), ]
  
  write.csv(vh, paste0("VDJ_", VDJgene, "_genesPerPatientGroup.csv")) #export counts per patient group
  
  vh <-  as.data.frame(t(t(vh) / colSums(vh))) #get percentages of repertoir
  vh$gene <- rownames(vh)
  
  nvh <- ifelse(nrow(vh) <= 50, nrow(vh), 50)
  
  p <- as_tibble(vh[1:nvh, ]) %>% pivot_longer(
       cols = 1:4,
       names_to = "patient_group",
       names_prefix = "",
       values_to = "count",
       values_drop_na = TRUE) %>%  
          ggplot(aes(x = reorder(gene, -count), y = count, fill = patient_group)) +
              geom_col(position = "dodge") +
              theme_classic() +
              xlab("") +
              ylab("Percent of repertoire") +
              scale_y_continuous(labels = percent) +
              ggtitle(paste0(VDJgene, " genes, top ", nvh)) +
              theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
              facet_grid(rows = vars(patient_group))

    ggsave2(paste0("../results/VDJ_geneUsage_", VDJgene, "_geneUsage_perPatientGroup.png"), 
          p,
          width = 30, height = 14, unit = "cm")
  
   }

get_topVDJ(gex, "v_call")
get_topVDJ(gex, "d_call")
get_topVDJ(gex, "j_call")
get_topVDJ(gex, "c_call")


```


#### VDJ gene by alakazam countGenes()

```{r}

get_VDJperc <- function(i, #seurat object with imm cantation out in meta data
                       VDJgene = "tmp" #VDJ gene to plot
  ) {
  
  tmp <- i@meta.data
   
  ####---- 1 gene count per patient group per seurat cluster
  vdj_gene_1 <- countGenes(tmp, 
                           gene = VDJgene, 
                           groups = c("patient_group", "seurat_clusters"), 
                           mode = "gene")
  
  p <- ggplot(vdj_gene_1, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene Usage per cluster")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(seurat_clusters))

  ggsave2(paste0("../results/VDJ_countGenes_",VDJgene 
                 ,"_gene_perPatientGroup_perCluster.png"), 
          p,
          width = 15, height = 40, unit = "cm")
  
  
  ####---- 2 gene family count per patient group per seurat cluster
  vdj_family_1 <- countGenes(tmp, 
                             gene = VDJgene, 
                             groups = c("patient_group", "seurat_clusters"), 
                             mode = "family")
  
   p <- ggplot(vdj_family_1, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene family Usage per cluster")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(seurat_clusters))

  ggsave2(paste0("../results/VDJ_countGenes_",VDJgene ,
                 "_family_perPatientGroup_perCluster.png"),
          p,
          width = 15, height = 40, unit = "cm")

  ####---- 3 gene count per patient group per cell type
  vdj_gene_2 <- countGenes(tmp, 
                           gene = VDJgene, 
                           groups = c("patient_group", "cluster_cellType_manual_main"), 
                           mode = "gene")
  
  p <- ggplot(vdj_gene_2, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene Usage per cell type")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(cluster_cellType_manual_main))

  ggsave2(paste0("../results/VDJ_countGenes_", VDJgene 
                 ,"_gene_perPatientGroup_perCelltype.png"),
          p,
          width = 20, height = 15, unit = "cm")
  
  ####---- 4 gene family count per patient group per cell type
  vdj_family_2 <- countGenes(tmp, 
                             gene = VDJgene, 
                             groups = c("patient_group", 
                                        "cluster_cellType_manual_main"), 
                             mode = "family")
  
  p <- ggplot(vdj_family_2, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene family Usage per cell type")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(cluster_cellType_manual_main))

  ggsave2(paste0("../results/VDJ_countGenes_", VDJgene 
                 ,"_family_perPatientGroup_perCelltype.png"),
          p,
          width = 20, height = 15, unit = "cm")
  
}

get_VDJperc(gex, "v_call")
get_VDJperc(gex, "d_call")
get_VDJperc(gex, "j_call")
get_VDJperc(gex, "c_call")

```


#### Fig5B

```{r}

####---- make Fig5B
VDJgene <- "v_call"
vdj_gene <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group"), 
                       mode = "gene",
                       fill = TRUE)

# vdj_gene <- vdj_gene[!(vdj_gene$gene %in% 
#                          (names(table(vdj_gene$gene))[table(vdj_gene$gene) < 4])), ]

vdj_gene$fc <- NA

table(vdj_gene$gene)
for (i in 1:nrow(vdj_gene)){
  ifelse(vdj_gene$patient_group[i] == "CTRL", 
         vdj_gene$fc[i] <- NA,
         vdj_gene$fc[i] <- log2((vdj_gene$seq_freq[i] * 100 + 1)/ (vdj_gene$seq_freq[vdj_gene$gene == vdj_gene$gene[i] & vdj_gene$patient_group == "CTRL"] *100  + 1)))
}

vdj_gene <- vdj_gene[order(vdj_gene$seq_count, decreasing = TRUE), ]
tmp <- vdj_gene[which(vdj_gene$gene %in% unique(vdj_gene$gene)[1:40]), ]

tmp <- vdj_gene

p1 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = fc, 
                          colour = patient_group)) +
  theme_classic() +
      ggtitle(paste0("IGHV gene usage per patient group")) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
            legend.position = "none",
            plot.margin = unit(c(0.2,0,0,0.2), "cm")) +
      ylab("log2FC to CTRL") +
      xlab("") +
      geom_point() +
  geom_hline(yintercept = 0)

p2 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, 
                                       vjust = 1), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0,0,0,0.2), "cm")) +
      ylab("Percent of repertoire") +
      xlab("") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

p_legend <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
  scale_fill_discrete(name = "Patient group") +
      geom_col(position = "dodge")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p3 <- plot_grid(p1, p2, nrow = 2, labels = NULL, align = "v", rel_heights = c(1, 2))


 ggsave2(paste0("../results/VDJ_countGenes_", VDJgene 
                 ,"_test.png"),
          plot_grid(p3, p_legend, 
                    ncol = 2, 
                    labels = NULL, 
                    align = "v", 
                    rel_widths = c(12, 2)) + bgcolor("white"),
          width = 20, height = 15, unit = "cm")


```


### top clonotypes

```{r}
sort(table(gex$CTaa), decreasing = T)[1:5]
sort(table(gex$junction_aa), decreasing = T)[1:10]
sort(table(gex$junction), decreasing = T)[1:10]
colnames(gex.meta)
# gex$CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF <- factor(gex$CTaa == "CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF")
# 
# for (i in 1:15){
# print(sort(table(gex$CTaa), decreasing = T)[i])
# print(table(factor(gex$CTaa == names(sort(table(gex$CTaa), decreasing = T)[i])), gex$orig.ident))
# }
# 
# table(gex$CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF, gex$orig.ident)
# 
# p1 <- DimPlot(gex, reduction = "umap_2", 
#         group.by = "CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF", 
#         cols = c("gray80", "firebrick", "gray80"), order=TRUE, pt.size = 0.5, split.by = "orig.ident", ncol = 2)
# 
# ggsave2("../results/clono_test.png", p1, height = 30, width = 10, units = "cm")
# gex$seurat_clusters_patient_group <-paste0(gex$seurat_clusters, "_", gex$patient_group)
# 
# 
# unlist(lapply(strsplit(gex@meta.data$v_call, split = ","), function(x) {
#   ifelse(length(x) > 1 , return(x[1]), return(x))
#   }
# ))


```


### top clonotypes

```{r}

colnames(gex@meta.data)
"junction"                     "junction_aa"                 

 [79] "cdr3"

tmp <- gex$junction_aa
tmp <- gex$cdr3
names(tmp) <- colnames(gex)
head(tmp)

junc_dist <- pairwiseDist(tmp, dist_mat = getDNAMatrix(gap=0))
#Error: vector memory exhausted (limit reached?)

cdr3_eq <- pairwiseEqual(tmp)


seq <- c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C", E="NTGGG")
d <- pairwiseEqual(seq)
rownames(d) <- colnames(d) <- seq
d


```


### Quantify clonotypes scRepertoire

```{r}

quantContig(
  combined_B,
  cloneCall = "gene+nt",
  chain = "both",
  scale = TRUE,
  group = "patient_group",
  exportTable = FALSE
)
head(combined_B)
quantContig(combined_B, cloneCall="gene+nt", scale = T)

abundanceContig(combined_B, cloneCall = "gene", scale = F)
lengthContig(combined_B, cloneCall="aa", chain = "IGL") 

clonalProportion(combined_B, cloneCall = "gene") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

occupiedscRepertoire(gex, x.axis = "seurat_clusters")

clonalOverlap(combined_B, cloneCall = "gene+nt", 
              method = "morisita")

head(combined_B)
colnames(gex@meta.data)
table(gex$clone_id)
head(gex$CTaa)
for_paulo <- gex@meta.data[ , c("orig.ident", "patient_group", "cluster_cellType_manual_fine", "cluster_cellType_manual_main", "CTnt", "CTaa")]
head(for_paulo)
write.csv(for_paulo, "../results/for_clono.csv")


alluvialClonotypes(gex, cloneCall = "gene", 
                   y.axes = c("orig.ident", "seurat_clusters"), 
                   color = "seurat_clusters")

p1 <- alluvialClonotypes(gex, cloneCall = "gene", 
                   y.axes = c("orig.ident", "cluster_cellType_manual_main"), 
                   color = "cluster_cellType_manual_main")
ggsave2("../results/alluvialClonotypes_test.png", p1, height = 20, width = 30, units = "cm")

combined2 <- expression2List(gex, split.by = "cluster_cellType_manual_fine")
combined2 <- expression2List(gex, split.by = "cluster_cellType_manual_main")
str(combined2)

clonalHomeostasis(combined2, cloneCall = "nt") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

clonalProportion(combined2, cloneCall = "nt") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

clonalOverlap(combined2, cloneCall="aa", method="overlap") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

table(gex$orig.ident[gex$seurat_clusters == 22])
table(gex$mu_load[gex$seurat_clusters == 22])

clonalDiversity(gex, cloneCall = "nt",group.by = "cluster_cellType_manual_fine")

#results <- hierarchicalClones(imm, threshold=0.1)




```


### clonotype analysis

# !!! FIX THIS !!!

```{r}


clones <- countClones(gex@meta.data, 
                      groups = c("patient_group", "cluster_cellType_manual_main"), 
                      clone = "clone_id")

clones2 <- countClones(gex@meta.data, 
                      groups = c("cluster_cellType_manual_main"), 
                      clone = "clone_id")

ggplot(gex@meta.data, aes(y = clone_id, x = patient_group, fill = clone_id)) +
      theme_bw() +
      ggtitle("Vh Family Usage per group") +
      #theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Clonal abundance") +
      xlab("") +
      #scale_y_continuous(labels = percent) +
      geom_col()

ggplot(gex@meta.data, y = )


png("../results/clonotype_piecharts.png", height = 1600, width = 3200, res = 300)
par(mfrow = c(1, 4))
for (i in unique(gex@meta.data$patient_group)){
  
  clones_tmp <- sort(table(gex@meta.data$clone_id[gex@meta.data$patient_group == i]), decreasing = TRUE)
  
  pie(clones_tmp[1:1000], clockwise = T, main = i, labels = names(clones_tmp)[1:9],radius = .6,border = NA,xpd=T,line=0)
  
}
par(mfrow = c(1, 1))
dev.off()



clones_tmp <- sort(table(gex@meta.data$clone_id[gex@meta.data$patient_group == i]), decreasing = TRUE)
head(clones_tmp)
#,col = mycol,

#pie chart
pie(clones_tmp[1:1000], clockwise = T, main = paste0("\n\n",j," diversity"), labels = names(clones_tmp)[1:9],radius = .6,border = NA,xpd=T,line=0)

arrange(clones, patient_group, seq_freq) %>% 
ggplot(aes(x = clone_id, y = seq_freq)) +
      theme_bw() +
      geom_col(aes(color = patient_group))

dev.off()
summary(clones)
head(clones2)

ggplot(clones, aes(x = clone_id, y = seq_freq, fill = patient_group)) +
      theme_bw() +
      ggtitle("Vh Family Usage per group") +
      #theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Clonal abundance") +
      xlab("") +
      #scale_y_continuous(labels = percent) +
      geom_col()

geom_line(aes(color = patient_group), size = 2, alpha = 0.8)

#abundance estimation
cu <- estimateAbundance(outfile, group = "patient_group", 
                        ci = 0.95, nboot = 100, clone = "clone_id")
#Error: vector memory exhausted (limit reached?)
#nboot = 100 works locally, nboot = 200 does not

#plot(cu, legend_title = "patient_group")

png("../results/VDJ_clonalAbundances_perPatientGroup.png", 
    width = 20, height = 20, units = "cm", res = 200)
plot(cu, 
     #colors = sample_colors, 
     main_title = "clonal abundances", 
     legend_title = "Patient_group")
dev.off()


#diversity curve
#per patient group
sample_curve <- alphaDiversity(outfile, group = "patient_group", clone = "clone_id",
                               min_q = 0, max_q = 4, step_q = 0.1,
                               ci = 0.95, nboot = 200)

png("../results/VDJ_diversityCurve_perPatientGroup.png", 
    width = 20, height = 20, units = "cm", res = 200)
plot(sample_curve, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "Patient_group")
dev.off()


#per c_call
#remove c_call classes with too few cells
tmp <- outfile[!(outfile$c_call == "IGHE" | 
                   outfile$c_call == "IGLC1" |
                   outfile$c_call == "IGLC2"), ]

isotype_curve <- alphaDiversity(tmp, group = "c_call", clone = "clone_id",
                                min_q = 0, max_q = 4, step_q = 0.1,
                                ci = 0.95, nboot = 200)

png("../results/VDJ_diversityCurve_perPatientGroup.png", 
    width = 20, height = 20, units = "cm", res = 200)
plot(isotype_curve, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "c_call")
dev.off()



```


####----- get v genes with most mutations
#  !!!! fix this !!!

```{r}
# tmp <- gex@meta.data[which(gex@meta.data$v_call %in% unique(vdj_gene$gene)[1:40]), ]
# table(gex@meta.data$germline_v_call)
# colnames(gex@meta.data)
# ggplot(gex@meta.data, 
#              aes(x = v_call, 
#                  y = mu_freq_v_r_total, 
#                  color = patient_group)) +
#       theme_classic() +
#       ggtitle("VDJ mutations, full length - replacement") +
#       theme(axis.text.x = element_text(angle = 45, 
#                                        hjust = 1, vjust = 1)) +
#       ylab("mutated bp (%)") +
#       xlab("") +
#       scale_y_continuous(labels = percent) +
#       geom_boxplot(size = 0.4, alpha = 0.8) 
# 
#   facet_grid(rows = vars(cluster_cellType_manual_main)) 

  
  vdj_gene_mu <- gex@meta.data
  #sort(table(vdj_gene_mu$v_call), decreasing = T)
  vdj_gene_mu$v_gene <- sub("\\*.*", "", vdj_gene_mu$v_call)
  
  #sort(table(vdj_gene_mu$v_gene), decreasing = T)
  #table(vdj_gene_mu$v_gene)
  
  #vdj_gene_mu$v_gene <- gsub("IGHV3-30-3", "IGHV3-30", vdj_gene_mu$v_gene)
  #vdj_gene_mu$v_gene <- gsub("IGHV4-38-2", "IGHV4-38", vdj_gene_mu$v_gene)
  #vdj_gene_mu$v_gene <- gsub("IGHV5-10-1", "IGHV5-10", vdj_gene_mu$v_gene)
  #vdj_gene_mu$v_gene <- gsub("IGHV4-30-4", "IGHV4-30", vdj_gene_mu$v_gene)
  #vdj_gene_mu$v_gene <- gsub("IGHV4-30-2", "IGHV4-30", vdj_gene_mu$v_gene)
  vdj_gene_mu$v_gene <- gsub("IGHV7-4-1", "IGHV7-4", vdj_gene_mu$v_gene)
  vdj_gene_mu$v_gene <- gsub("IGHV3-43D", "IGHV3-43", vdj_gene_mu$v_gene)
  vdj_gene_mu$v_gene <- gsub("IGHV1-69-2", "IGHV1-69", vdj_gene_mu$v_gene)
  vdj_gene_mu$v_gene <- gsub("IGHV3-69-1", "IGHV3-69", vdj_gene_mu$v_gene)
  vdj_gene_mu$v_gene <- gsub("IGHV2-70D", "IGHV2-70", vdj_gene_mu$v_gene)

# vdj_gene_mu  <- vdj_gene_mu %>%  group_by(v_gene, patient_group) %>%
#     summarise(mu_mean = mean(mu_freq_cdr_r), n = n()) 

vdj_gene_mu  <- vdj_gene_mu %>%  group_by(v_gene, patient_group, orig.ident, cluster_cellType_manual_main) %>%
    summarise(mu_mean = mean(mu_freq_total), n = n()) 

# vdj_gene_mu  <- vdj_gene_mu %>%  group_by(v_gene, patient_group) %>%
#     summarise(mu_mean = mean(mu_freq_cdr3_r), n = n()) 

vdj_gene_mu <- vdj_gene_mu[vdj_gene_mu$cluster_cellType_manual_main == "DN" |
                           vdj_gene_mu$cluster_cellType_manual_main ==  "Memory" |
                           vdj_gene_mu$cluster_cellType_manual_main ==  "Naive" |
                           vdj_gene_mu$cluster_cellType_manual_main ==  "Transitional",]

# vdj_gene_mu$cluster_cellType_manual_main <- factor(vdj_gene_mu$cluster_cellType_manual_main, levels = unique(vdj_gene_mu$cluster_cellType_manual_main))
# 
# table(vdj_gene_mu$cluster_cellType_manual_main)

vdj_gene_mu$patient_group_gene <- paste(vdj_gene_mu$patient_group, 
                                        vdj_gene_mu$v_gene, sep = "_")
vdj_gene$patient_group_gene <- paste(vdj_gene$patient_group, 
                                     vdj_gene$gene, sep = "_")

#vdj <- merge(vdj_gene, vdj_gene_mu, by = "patient_group_gene")

vdj <- left_join(vdj_gene_mu, vdj_gene, by = "patient_group_gene")
vdj <- vdj[which(vdj$gene %in% unique(vdj_gene$gene)[1:40]), ]

p    <- ggplot(vdj, aes(x = reorder(v_gene, -seq_freq), 
                    y = mu_mean, 
                    fill = patient_group.x)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
    scale_y_continuous(labels = percent) +
    geom_boxplot(position = "dodge", ) + 
      facet_grid(rows = vars(cluster_cellType_manual_main))
    
 ggsave2(paste0("../results/VDJ_mu_per_patGroup_perCellType_test.png"),
          p ,
          width = 25, height = 15, unit = "cm")

  # colnames(gex@meta.data)
  # colour = patient_group.x
                              
 
  
```


### plot mutation status

```{r}

gex.meta <- gex@meta.data[gex$cluster_cellType_manual_main == "Memory" | 
                          gex$cluster_cellType_manual_main == "Naive" |
                          gex$cluster_cellType_manual_main == "DN" |
                          gex$cluster_cellType_manual_main == "Transitional", ]

####---- CDR3 mutation frequency - cell type fine - patient_group, 
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr3_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR3 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr3_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR3 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muCDR3_CellTypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- CDR3 mutation frequency - celltype main - patient group
# split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr3_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 mutations - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr3_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 mutations - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muCDR3_CelltypeMain_PatientGroup_rs.png"), 
          plot_grid(p1, p2, ncol = 2, labels = "AUTO", align = "h"),
          width = 25, height = 10, unit = "cm")


####---- CDR3 mutation frequency - celltypeMain - orig.ident 
# split by replacement/silent
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr3_r,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 mutations - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr3_s,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 mutations - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muCDR3_CelltypeMain_orig.ident_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = "AUTO", align = "h"),
          width = 25, height = 15, unit = "cm")


####---- CDR3 mutation count - celltypeMain - patient group
# split by replacement/silent

# !!!! FIX THIS !!!

as_tibble(gex.meta) %>% 
  group_by(cluster_cellType_manual_main, 
                                 patient_group, 
                                 mu_count_cdr3_r) %>% 
  
  summarise(cdr3_r = n() ) %>% 
  mutate(cdr3_r = cdr3_r/sum(cdr3_r)) %>%
  ggplot(aes(x = mu_count_cdr3_r, 
             y = cdr3_r, 
             fill = patient_group)) +
  theme_classic() +
  geom_col(position = "dodge") +
  scale_x_continuous(breaks = c(0:7)) +
  scale_y_continuous(labels = percent) +
  xlab("number of mutated CDR3 bp") + 
  ylab("fraction (%)") +
  facet_grid(cols = vars(cluster_cellType_manual_main))

colnames(gex@meta.data)


####---- VDJ mutation frequency up until bp 312 - celltypeFine - patient_group
p <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_v_total, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) 

ggsave2(paste0("../results/VDJ_mu312_CellTypeFine_patientGroup.png"), 
          p,
          width = 30, height = 10, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeFine - patient_group
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_v_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_v_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_CellTypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeMain - patient_group
p <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_v_total, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_celltype_patientGroup.png"), 
          p,
          width = 20, height = 8, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeMain - orig.ident
p <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_v_total,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_celltypeMain_orig.ident.png"), 
          p,
          width = 20, height = 8, unit = "cm")


####---- VDJ mutation frequency up until bp 312 - celltypeMain - patient_group
# split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_v_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_v_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, IGHV until bp 312 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_mu312_celltypeMain_patientGroup_rs.png"), 
          plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"),
          width = 25, height = 10, unit = "cm")


####---- VDJ mutation frequency for CDR1-3 - celltypeMain - patient_group 
#split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_cdr_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_CDR_celltypeMain_patientGroup_rs.png"), 
          plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"),
          width = 25, height = 10, unit = "cm")


####---- VDJ mutation frequency for FWR1-4 celltypeMain - patient_group 
# split by replacement/silent
p1 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_fwr_r, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_fwr_s, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_FWR_byCelltype_rs_perPatientGroup.png"), 
          plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"),
          width = 25, height = 10, unit = "cm")


####----  total VDJ mutation frequency for CDR1-3 - celltypeMain -  orig.ident 
# split by replacement/silent
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr_r,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_cdr_s,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_CDR_celltypeMain_orig.ident_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 25, height = 18, unit = "cm")


####----  total VDJ mutation frequency for FWR1-4 - celltypeMain - orig.ident 
# split by replacement/silent
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_fwr_r,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - replacement") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_fwr_s,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - silent") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_FWR_celltypeMain_orig.ident_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 25, height = 18, unit = "cm")


####---- VDJ mutation frequency for CDR1-3 - celltypeFine - patient_group
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_cdr_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, CDR1-3 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_CDR_celltypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- total VDJ mutation frequency for FWR1-4  
# split by replacement/silent
p1 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_fwr_r, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - replacement") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

p2 <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_fwr_s, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations, FWR1-4 - silent") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_FWR_celltypeFine_patientGroup_rs.png"), 
          plot_grid(p1, p2, nrow = 2, labels = NULL, align = "h"),
          width = 30, height = 15, unit = "cm")


####---- total VDJ mutation frequency - celltypeMain - patient_group
p <- ggplot(gex.meta,
            aes(x = cluster_cellType_manual_main, 
                y = mu_freq_total, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_celltypeMain_patientGroup.png"), 
          p,
          width = 20, height = 6, unit = "cm")


####---- total VDJ mutation frequency - celltypeMain - orig.ident
p <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = mu_freq_total,
                  fill = orig.ident)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75))

ggsave2(paste0("../results/VDJ_muFull_celltypeMain_orig.ident.png"), 
          p,
          width = 20, height = 8, unit = "cm")


####---- total VDJ mutation frequency - celltypeFine - patient_group
p <- ggplot(gex@meta.data, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) 

ggsave2(paste0("../results/VDJ_muFull_celltypeFine_patient_group.png"), 
          p,
          width = 30, height = 10, unit = "cm")


#####---- percent mutated cells per seurat cluster
cluster_mut <- as.data.frame.matrix(table(gex$seurat_clusters,
                                          gex$mu_load))
cluster_mut$percmut <- 1 - cluster_mut$unmutated/rowSums(cluster_mut)
cluster_mut$orig.ident <- factor(rownames(cluster_mut), 
                                 levels = rownames(cluster_mut))

p <- ggplot(cluster_mut, aes(x = orig.ident, y = percmut)) +
      theme_classic() +
      ggtitle("perc cells with mutated VH per cluster") +
      theme(axis.text.x = element_text(angle = 45, 
                                       hjust = 1, vjust = 1)) +
      ylab("perc mutated cells") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_bar(stat='identity')

ggsave2(paste0("../results/VDJ_percMutCells_byCluster.png"), 
          p,
          width = 30, height = 20, unit = "cm")


####---- UMAP colored by categorical muLoad
p <- DimPlot(gex, reduction = "umap_2", 
                  pt.size = .01, 
                  group.by = "mu_load", 
                  raster = FALSE) + 
          ggtitle("Mutation load") +
          ylab("UMAP 2") +
          xlab("UMAP 1")

ggsave2(paste0("../results/UMAP_B-cells_muLoad_1.png"), 
          p,
          width = 25, height = 20, unit = "cm")

```


###  plot CDR3 aa attributes

```{r}
# "cdr3_aa_length"               
# "cdr3_aa_gravy"               
# "cdr3_aa_bulk"                 
# "cdr3_aa_aliphatic"
# "cdr3_aa_polarity"             
# "cdr3_aa_charge"              
# "cdr3_aa_basic"               
# "cdr3_aa_acidic"
# "cdr3_aa_aromatic"


#### ----- per cell type
#subeset gex to only include relevant cell types

gex.meta <- gex@meta.data[gex$cluster_cellType_manual_main == "Memory" | 
                          gex$cluster_cellType_manual_main == "Naive" |
                          gex$cluster_cellType_manual_main == "DN" |
                          gex$cluster_cellType_manual_main == "Transitional" , ]

#CDR3 aa length
p1 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_length,
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa length") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("length (bp)") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#grand average of hydrophobicity
p2 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_gravy, #grand average of hydrophobicity
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa grand average of hydrophobicity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("hydrophobicity score") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#average bulkiness of amino acids
p3 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_bulk, #average bulkiness of amino acids
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa bulkiness") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average bulkiness of amino acids") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aliphatic
p4 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_aliphatic, #aliphatic index
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa aliphatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("aliphatic index") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_polarity
p5 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_polarity, #average polarity of amino acids
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa polarity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average polarity") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_charge
p6 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_charge, #net charge
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 charge") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("net charge") +
      xlab("") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_basic
p7 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_basic, #fract. of informative positions that are Arg, His or Lys
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa basic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos that are Arg, His or Lys") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_acidic
p8 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_acidic, #fraction of informative positions that are Asp or Glu
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa acidic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are Asp or Glu") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aromatic
p9 <- ggplot(gex.meta,
              aes(x = patient_group, 
                  y = cdr3_aa_aromatic, #fract. of informative pos. that are His, Phe, Trp or Tyr
                  color = orig.ident)) +
      theme_classic() +
      ggtitle("CDR3 aa aromatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are His, Phe, Trp or Tyr") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_boxplot(size = 0.3, alpha = 0.8)

ggsave2(paste0("../results/VDJ_attrCDR3_byCelltype_perSample_2.png"), 
          plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, p9, 
                    nrow = 3, ncol = 3, labels = "AUTO", align = "h"),
          width = 40, height = 30, unit = "cm")


#### ----- per cluster
#CDR3 aa length
p1 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_length,
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa length") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("length (bp)") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#grand average of hydrophobicity
p2 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_gravy,
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa grand average of hydrophobicity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("hydrophobicity score") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#average bulkiness of amino acids
p3 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_bulk, #average bulkiness of amino acids
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa bulkiness") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average bulkiness of amino acids") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aliphatic
p4 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aliphatic, #aliphatic index
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aliphatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("aliphatic index") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_polarity
p5 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_polarity, #average polarity of amino acids
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa polarity") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("average polarity") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_charge
p6 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_charge, #net charge
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa charge") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("net charge") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)

#cdr3_aa_basic
p7 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_basic, #fract. of informative positions that are Arg, His or Lys
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa basic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos that are Arg, His or Lys") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_acidic
p8 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_acidic, #fraction of informative positions that are Asp or Glu
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa acidic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are Asp or Glu") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aromatic
p9 <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic, #fract. of informative pos. that are His, Phe, Trp or Tyr
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aromatic") +
      theme(axis.text.x=element_text(angle = 45, 
                                     hjust = 1, vjust = 1), 
            legend.position = "none") +
      ylab("% of pos. that are His, Phe, Trp or Tyr") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)

#get legend
p_legend <- ggplot(gex@meta.data,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic,
                  color = patient_group)) +
      theme_classic() +
      geom_boxplot() 
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()


p_grid <- plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, p9, 
                    nrow = 3, ncol = 3, labels = "AUTO", align = "h")

ggsave2(paste0("../results/VDJ_attrCDR3_byCluster_perSample_2.png"), 
          plot_grid(p_grid, p_legend, ncol = 2, align = "none", rel_widths = c(12, 1)),
          width = 45, height = 30, unit = "cm")


```


###  plot IgH class, mu_load and clonetype per cluster

```{r}

##---- IgH per cluster
# res <- table(list(gex$seurat_clusters, gex$c_gene_10x))
# res <- t(res)
# res <- res[,order(colSums(res),decreasing = T)]
# res2 <- t( t(res) / colSums(res))
# res3 <- res / rowSums(res)
# res4 <- t( t(res3) / colSums(res3) )
# mypal <- RColorBrewer::brewer.pal(nrow(res2),"Set3")
# 
# #plot proportion of cells per cluster
# png(file = paste0("../results/cluster_IgHclass_abundance.png"),
#     width = 20, height = 8, units = "cm", res = 200)
# mypar(1,1,mar = c(2,4,2,5), xpd=TRUE, cex.axis = 0.7)
# barplot(res2*100, las = 1, col = mypal, main = "IgH class", ylab = "Proportion of cells\n")
# legend("topright", legend = rownames(res2), bty = "n",fill = mypal, inset=c(-0.15,0))
# dev.off()


p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(9, "Set3"), "grey")) + 
              theme(legend.position = "none") + ggtitle("IGH c gene") + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank())

p2 <- ggplot(gex@meta.data, aes(x = cluster_cellType_manual_fine, fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") +
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      scale_fill_brewer(palette="Set3", na.value="grey", name = "c gene")

ggsave2("../results/cluster_c_gene_10x_abundance.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h") + bgcolor("white"), 
        width = 30, height = 14, unit = "cm")

p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", 
              ncol = 2, 
              raster = FALSE) + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank()) + 
  ggtitle("IGH c gene")
  

ggsave2("../results/UMAP_B-cells_IgHclass.png", 
        p1, 
        width = 30, height = 20, unit = "cm")


##---- mutation load per cluster
# res <- table(list(gex$seurat_clusters, gex$mu_load))
# res <- t(res)
# res <- res[,order(colSums(res),decreasing = T)]
# res2 <- t(t(res)/colSums(res))
# res3 <- res / rowSums(res)
# res4 <- t( t(res3) / colSums(res3) )
# mypal <- RColorBrewer::brewer.pal(nrow(res4),"Set3")
# 
# #plot proportion of cells per cluster
# png(file = paste0("../results/cluster_muLoad_abundance.png"),
#     width = 20, height = 8, units = "cm", res = 200)
# mypar(1,1,mar = c(2,4,2,6), xpd=TRUE, cex.axis = 0.7)
# barplot(res2*100, las = 1, col = mypal, main = "Mutation load", ylab = "Proportion of cells\n")
# legend("topright", legend = rownames(res2), bty = "n",fill = mypal, inset=c(-0.18,0))
# dev.off()

p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(4, "Dark2"), "grey")) +
              theme(legend.position = "none") +
              ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank())

p2 <- ggplot(gex@meta.data, aes(x = cluster_cellType_manual_fine, fill = mu_load)) +
          theme_classic() +
          geom_bar(position = "fill") + 
          ylab("proportion of cells") +
          xlab("") +
          facet_grid(rows = vars(patient_group)) + 
          theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
          scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "Mutation load")

ggsave2("../results/cluster_mu_load_abundance.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"), 
        width = 30, height = 14, unit = "cm")

p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", 
              ncol = 2, 
              raster = FALSE) +
              ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank())
              
ggsave2("../results/UMAP_B-cells_muLoad.png", 
        p1, 
        width = 30, height = 20, unit = "cm")


##---- cloneType per cluster
ggplot(gex@meta.data, aes(x = seurat_clusters, fill = c_call)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      facet_grid(rows = vars(patient_group))

res <- table(gex$seurat_clusters, gex$cloneType, gex$patient_group)

res <- t(res)
res <- res[,order(colSums(res),decreasing = T)]
res2 <- t(t(res)/colSums(res))
res3 <- res / rowSums(res)
res4 <- t( t(res3) / colSums(res3) )
mypal <- RColorBrewer::brewer.pal(nrow(res4),"Set3")

#plot proportion of cells per cluster
png(file = paste0("../results/cluster_cloneType_abundance.png"),
    width = 20, height = 8, units = "cm", res = 200)
mypar(1,1,mar = c(2,4,2,12), xpd=TRUE, cex.axis = 0.5)
barplot(res2*100, las = 1, col = mypal, main = "clonotype", ylab = "Proportion of cells\n")
legend("topright", legend = rownames(res2), bty = "n",fill = mypal, inset=c(-0.5,0))
dev.off()


####---- cloneTypes
p1 <- DimPlot(gex, 
              group.by = "cloneType", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(4, "Dark2"), "grey")) +
              theme(legend.position = "none") +
  ggtitle("clonotype") +  
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank())

p2 <- ggplot(gex@meta.data, aes(x = cluster_cellType_manual_fine, fill = cloneType)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") + 
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      scale_fill_brewer(palette="Dark2", na.value = "grey", name = "clonotype")

ggsave2("../results/cluster_cloneType_abundance.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h"), 
        width = 30, height = 14, unit = "cm")


```


## Save rds GEX_BCELLS

```{r}

saveRDS(gex, file = "../results/GEX7_BVDJ.rds")
#gex <- readRDS("../results/GEX7_BVDJ.rds")

```


## Print sessionInfo()

```{r}

sessionInfo()

```



