---
output: html_document
editor_options: 
  chunk_output_type: console
---

# VDJ analysis {#VDJ}
## Load packages

```{r}

source("./00_dependencies.R")

```


## Load Seurat object

```{r}

gex <- readRDS("../results/GEX6_BANNOT.rds")

```


## Load pheno file

```{r}

pheno <- read.table("../suppl/pheno/pSS_pheno_211208.csv", sep = ",", header = TRUE)
colnames(pheno)[c(1, 3)] <- c("Sample_id", "orig.ident")
pheno$orig.ident <- gsub("P007[ab]", "P007", pheno$orig.ident)
pheno$orig.ident <- gsub("P008[ab]", "P008", pheno$orig.ident)
pheno <- unique(pheno[, c(1, 3, 5, 7:8, 11:12)])

pheno$patient_group <- paste0(ifelse(pheno$SSA == "YES", "SSA", ""), 
                            "_", 
                            ifelse(pheno$SSB == "YES", "SSB", ""))
pat_annot <- c(SSAB = "SSA_SSB", SSA = "SSA_", DNEG="NA_NA", DNEG="_")
pheno$patient_group <- names(pat_annot)[match(pheno$patient_group, pat_annot)]
pheno$patient_group[grep("^C00", pheno$orig.ident) ] <- "CTRL"

head(pheno); dim(pheno)



#Four non-overlapping groups
#1. SSAB - SSA+/SSB+
#2. SSA  - SSA+/SSB-
#3. DNEG - SSA-/SSB-
#4. CTRL - CTRL


```


## immcantation
### Load output from Immcantation and perform mutation analysis

```{r}

files_B <- list.files("../data/clono_B_VDJ_tmp", 
                      pattern = "*germ-pass.tsv", 
                      recursive = TRUE, 
                      full.names = TRUE)
ncpu <- cpuCount()
imm <- NULL
for (i in files_B) {
  
  cat("reading: ", print(i), "\n")
  tmp <- read_rearrangement(i)
  
  ####---- get sample name
  j <- gsub("_heavy_germ-pass.tsv", "", gsub(".*out/", "", i))
  cat("sample name: ", print(j), "\n")
  
  ####---- number of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = FALSE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_v_total'
  
  ####---- frequency of mutations per cell for V up untill base 312
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = TRUE, 
                           combine = TRUE,
                           nproc = ncpu,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_v_total'
  
  ####---- number of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for V up untill base 312 
  # per segment per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per region (CDR1, CDR2, CDR3, FWR1, FWR2, FWR3 and FWR4) per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")

  ####---- number of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- frequency of mutations per cell for full VDJ 
  # per CDR/FWR per replacement/silent
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  ####---- number of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
   
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_total'
  
  ####---- frequency of mutations per cell for full VDJ
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = TRUE,
                          nproc = ncpu,
                          cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_total'
 
  ####---- properties for CDR3 aa's
  tmp <- aminoAcidProperties(tmp, 
                             seq = "junction", 
                             nt = TRUE, 
                             trim = TRUE, 
                             label = "cdr3")
  
  tmp$cell_id <- paste0(j, "_", gsub("-1", "", tmp$cell_id))
  tmp$orig.ident <- gsub("a|b", "", gsub("_.*", "", tmp$cell_id))
  tmp$sequence_id <- paste0(j, "_", gsub("-1", "", tmp$sequence_id))
  rownames(tmp) <- tmp$cell_id
  
  if (is.null(imm)) {
    imm <- tmp
  } else {
    imm <- rbind(imm, tmp)
  }
  
}


```


### integrate immcantation VDJ data with seurat object

```{r}

#select immcantation columns
# cols <- c("sequence_id", "productive", "vj_in_frame", "stop_codon", "locus", 
#           "v_call", "d_call", "j_call", "c_call", 
#           "cell_id", "clone_id", "umi_count", 
#           "v_call_10x", "d_call_10x", "j_call_10x", 
#           "germline_v_call","germline_d_call", "germline_j_call",
#           "cdr3", "consensus_count",
#           "mu_count_total", "mu_freq_total", 
#           "mu_count_v_r_total", "mu_count_v_s_total", "mu_freq_v_r_total", "mu_freq_v_s_total",
#           "mu_freq_cdr1_r", "mu_freq_cdr1_s", "mu_freq_cdr2_r", "mu_freq_cdr2_s", "mu_freq_cdr3_r", "mu_freq_cdr3_s", "mu_freq_fwr1_r", "mu_freq_fwr1_s", "mu_freq_fwr2_r", "mu_freq_fwr2_s", "mu_freq_fwr3_r", "mu_freq_fwr3_s", "mu_freq_fwr4_r", "mu_freq_fwr4_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "cdr3_aa_length", "cdr3_aa_gravy", "cdr3_aa_bulk", "cdr3_aa_aliphatic", "cdr3_aa_polarity", "cdr3_aa_charge", "cdr3_aa_basic", "cdr3_aa_acidic", "cdr3_aa_aromatic")

#select immcantation columns NOT to be added to gex
cols <- c("sequence", "sequence_alignment", "germline_alignment", 
          "v_cigar", "d_cigar", "j_cigar",
          "fwr1", "fwr2", "fwr3", "fwr4", "cdr1", "cdr2", 
          "germline_alignment_d_mask")

imm_add <- imm[, !(colnames(imm) %in% cols)]
rownames(imm_add) <- imm_add$cell_id

gex <- AddMetaData(gex, imm_add)

```


### add pheno data to immcantation object

```{r}

#imm$orig.ident <- gsub("a|b", "", gsub("_.*", "", imm$cell_id))
imm <- merge(imm, pheno[, c(2,8)], 
             by.x = "orig.ident", 
             by.y = "orig.ident", 
             all.x = TRUE); head(imm)
imm <- imm[, c(2:ncol(imm), 1)]

```


### save immcantation VDJ data

```{r}

colnames(imm)[colnames(imm) == 'orig.ident'] <- 'repertoire_id'
imm$duplicate_count <- imm$umi_count
colnames(imm)[colnames(imm) == 'clone_id'] <- 'clone_id_perSample'

write_rearrangement(imm, "../results/VDJ_immcantation_out.tsv")
#imm <- read_rearrangement("../results/VDJ_immcantation_out.tsv")

#imm_out <- imm[, c("sequence_id", "v_call", "j_call", "junction")]
#write_rearrangement(imm_out, "../results/VDJ_immcantation_out_forDefineClones.tsv")

```


## read compairr out

```{r}

cairr.aa <- read.table("../results/VDJ_compairr_out_aa.tsv", header = TRUE)
cairr.aa$barcode <- gsub("_contig_1", "", gsub("_contig_2", "", gsub("_contig_3", "", cairr.aa$sequence_id)))
rownames(cairr.aa) <- cairr.aa$barcode

colnames(cairr.aa)[1:2] <- paste0(colnames(cairr.aa)[1:2], "_airr")

#add to gex
gex <- AddMetaData(gex, cairr.aa[, 1:2])

```


## VDJ mutation
### add categorical variable for total mutation burden

```{r}

# unmutated
# low < 1%
# med < 2%
# high > 2%

gex$mu_load <- ifelse(gex$mu_freq_total == 0, "unmutated", 
                 ifelse(gex$mu_freq_total > 0 & gex$mu_freq_total <= 0.01, "low",
                        ifelse(gex$mu_freq_total > 0.01 & gex$mu_freq_total <= 0.02, "med",
                               ifelse(gex$mu_freq_total > 0.02, "high", NA))))

gex$mu_load <- factor(gex$mu_load, levels = c("high", "med", "low", "unmutated"))


```


### add categorical variable for clonotype expansion

```{r}

# "Hyperexpanded (100 < X <= 500)"
# "Large (20 < X <= 100)"
# "Medium (5 < X <= 20)"
# "Small (1 < X <= 5)"
# "Single (0 < X <= 1)"

table(gex$cluster_size_airr, useNA ="always")
factor(slot(seurat, "meta.data")$cloneType, 
                levels = c("Hyperexpanded (100 < X <= 500)", 
                           "Large (20 < X <= 100)", 
                           "Medium (5 < X <= 20)", 
                           "Small (1 < X <= 5)", 
                           "Single (0 < X <= 1)", 
                           NA))

gex$cloneType_airr <- factor(ifelse(is.na(gex$cluster_size_airr), NA, 
                             ifelse(gex$cluster_size_airr == 1, "Single (0 < X <= 1)", 
                                    ifelse((gex$cluster_size_airr > 1) & (gex$cluster_size_airr <= 5), "Small (1 < X <= 5)", ifelse((gex$cluster_size_airr > 5) & (gex$cluster_size_airr <= 20), "Medium (5 < X <= 20)", ifelse((gex$cluster_size_airr > 20) & (gex$cluster_size_airr <= 100), "Large (20 < X <= 100)", "Hyperexpanded (100 < X)"))))), 
                             levels = c("Hyperexpanded (100 < X <= 500)",
                                        "Large (20 < X <= 100)", 
                                        "Medium (5 < X <= 20)", 
                                        "Small (1 < X <= 5)", 
                                        "Single (0 < X <= 1)", NA))

```


## scRepertiore BCR
### Read 10x BCR files

```{r}

# read 10x VDJ BCR files
samples_B <- list.files("../data/clono_B_VDJ")

clono_B <- lapply(samples_B, function(x){
  
  tmp <- read.csv(paste0("../data/clono_B_VDJ/", x, "/", x, "_B_filtered_contig_annotations.csv"), 
                  stringsAsFactors = FALSE)
  return(tmp)
  
})

names(clono_B) <- samples_B

#remove trailing -1
clono_B <- lapply(seq_along(clono_B), function(i){
  clono_B[[i]]$barcode <- gsub("-1", "", clono_B[[i]]$barcode)
  return(clono_B[[i]])
})

names(clono_B) <- samples_B

clono_B <- clono_B[order(names(clono_B))]

print(object.size(clono_B), units = "MB") #227.6 Mb
saveRDS(clono_B, paste0("../results/clono_B_list.rds"))
#clono_B <- readRDS("../results/clono_B_list.rds")


```


### generate a combined data frame for BCR sequences using scRepertoire

```{r}

#split combineBCR and do it per sample to avoid memory issues
#takes on average 20min/sample to run
combined_B <- lapply(1:length(clono_B), function(x) {
  
  print(names(clono_B[x]))
  print(nrow(clono_B[[x]]))
  print(Sys.time())
  
  tmp <- combineBCR(clono_B[[x]],
                    samples = names(clono_B[x]),
                    ID = NULL,
                    removeNA = TRUE,
                    removeMulti = TRUE)

  return(tmp)
  
})

#reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
combined_B <- lapply(1:length(combined_B), function(x){
  tmp <- combined_B[[x]][[1]]
  return(tmp)
})

#---- handle samples where libraries were created and sequenced twice
combined_B[["P007"]] <- rbind(combined_B[["P007a"]], combined_B[["P007b"]])
combined_B[["P007a"]] <- NULL
combined_B[["P007b"]] <- NULL

combined_B[["P008"]] <- rbind(combined_B[["P008a"]], combined_B[["P008b"]])
combined_B[["P008a"]] <- NULL
combined_B[["P008b"]] <- NULL

combined_B <- combined_B[order(names(combined_B))]

names_tmp <- names(combined_B)
combined_B <- lapply(seq_along(combined_B), function(i) {
  combined_B[[i]]$sample <- gsub("a|b", "", combined_B[[i]]$sample)
  return(combined_B[[i]])
})

names(combined_B) <- names_tmp

saveRDS(combined_B, paste0("../results/combined_BCR_scRepertoire.rds"))
#combined_B <- readRDS("../results/combined_BCR_scRepertoire.rds")


```


### Integrate scRepertoir out with Seurat object

```{r}

gex <- combineExpression(combined_B, gex)

```


## add IgH call and VDJ genes without allele info to Seurat object

```{r}

igh_10x <- lapply(1:length(clono_B), function(x) {
  
  tmp <- clono_B[[x]][clono_B[[x]]$chain == "IGH" & 
                        clono_B[[x]]$productive == "true" & 
                        !is.na(clono_B[[x]]$c_gene) & 
                        clono_B[[x]]$c_gene != "" &
                        clono_B[[x]]$c_gene != "IGLC1" &
                        clono_B[[x]]$c_gene != "IGLC2" , c(1, 10)]
  tmp$barcode <- paste0(names(clono_B)[x], "_", tmp$barcode)
  names(tmp)[2] <- "c_gene_10x"

  tmp <- tmp[!duplicated(tmp$barcode), ]
  
  rownames(tmp) <- tmp$barcode
  tmp$barcode <- NULL
  
  return(tmp)
  }
)

igh_10x <- do.call(rbind.data.frame, igh_10x)
table(igh_10x$c_gene_10x, useNA = "always")
gex <- AddMetaData(gex, igh_10x)

gex$c_gene_10x <- factor(gex$c_gene_10x, levels = sort(unique(gex$c_gene_10x)))
gex$c_gene_10x <- as.character(gex$c_gene_10x)


#get  IGH _call genes without allele info
gex@meta.data$IGHV <- unlist(lapply(strsplit(gex@meta.data$v_call, 
                                             split = "\\*"), function(x) {return(x[1])}))

gex@meta.data$IGHD <- unlist(lapply(strsplit(gex@meta.data$d_call, 
                                             split = "\\*"), function(x) {return(x[1])}))

gex@meta.data$IGHJ <- unlist(lapply(strsplit(gex@meta.data$j_call, 
                                             split = "\\*"), function(x) {return(x[1])}))


```

## subset gex@meta for plotting

```{r}

gex.meta <- gex@meta.data[(gex$cluster_cellType_manual_fine == "DN1_ID3" |
                          gex$cluster_cellType_manual_fine == "DN2" |
                          gex$cluster_cellType_manual_fine == "DN2_CXCR3" |
                          gex$cluster_cellType_manual_fine == "DN2_ITGAX" |
                          gex$cluster_cellType_manual_fine == "DN4" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_CD1C" |
                          gex$cluster_cellType_manual_fine == "Naive" |
                          gex$cluster_cellType_manual_fine == "Naive_IFN" |
                          gex$cluster_cellType_manual_fine == "Naive_Transitional") &
                          gex$orig.ident !="P024", ]

table(gex.meta$orig.ident, gex.meta$cluster_cellType_manual_fine)


```



## immcantation and scRepertoir output visualization
### VDJ gene usage
#### plot FiG3B VDJ correlation and Supplementary Figure S11A-C

```{r}

colors_cellTypeFine2 <- c(brewer.pal(5, "Blues"), 
  brewer.pal(4, "YlOrRd"),
  brewer.pal(3, "Greens")[c(3,2,1)])

names(colors_cellTypeFine2) <- sort(unique(gex.meta$cluster_cellType_manual_fine))

#VDJ gene usage correlation
VDJ_cor <- function(i, #seurat object meta data containing data for VDJ gene usage
                    correl, #which correlation to use i.e. spearman or pearson...
                    VDJgene #VDJ gene to plot
                    ){
  
  cat(VDJgene, "\n") 
  cat(correl, "\n")
  
  #remove allele information
  i[, VDJgene] <- unlist(lapply(strsplit(i[, VDJgene], 
                                         split = "\\*"),
                                function(x) {return(x[1])}))
  
  i$celltype_patient_group <- paste0(i$cluster_cellType_manual_fine, 
                                     "-", 
                                     i$patient_group)
  
  tmp <- as.data.frame.matrix(table(list(i[, VDJgene],
                    i$celltype_patient_group)))
  
  cor_tmp <- cor(tmp, method = correl)
  print(head(cor_tmp))
  cor_tmp[is.na(cor_tmp)] <- 0

  ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                     DNEG = "#7CAE00", 
                                     SSA = "#00BFC4", 
                                     SSAB = "#C77CFF"),
                   #cluster = scales::hue_pal()(15),
                   cell_type = colors_cellTypeFine2)
  
  titles <- data.frame(gexnames = c("v_call", "d_call", "j_call", "c_gene_10x"),
                       outnames = c("V", "D", "J", "IGH c"))
  
  print(titles$outnames[titles$gexnames == VDJgene])
  
  p <- pheatmap::pheatmap(cor_tmp, 
                     annotation_col = data.frame(cell_type = factor(sub("-.*", 
                                                                        "", 
                                                                        colnames(cor_tmp))),
                                                 patient_group = factor(sub(".*-", 
                                                                            "",
                                                                            colnames(cor_tmp))),
                                                 row.names = colnames(cor_tmp)),
                     annotation_colors = ann_cols[1:2],
                     breaks = seq(0, 1, length.out = 100),
                     filename = paste0("../results/VDJ_Cor_", VDJgene, "_", 
                                        correl, "_per_cluster_PatientGroup_4.png"),
                     fontsize = 8, 
                     fontsize_row = 4, 
                     fontsize_col = 4,
                     border_color = NA,
                     show_rownames = FALSE,
                     show_colnames = FALSE,
                     main = paste0(titles$outnames[titles$gexnames == VDJgene], 
                                   " genes, ", correl, " correlation"),
                     color = c(colorRampPalette(c("lightblue1", 
                                                  "lightblue2", 
                                                  "firebrick2"))(99)))
  
  return(p)
  
}


sort(unique(gex.meta$cluster_cellType_manual_fine))
cor.s.v_call <- VDJ_cor(gex.meta, "spearman", "v_call")
cor.p.v_call <- VDJ_cor(gex.meta, "pearson", "v_call")

cor.s.d_call <- VDJ_cor(gex.meta, "spearman", "d_call")
cor.p.d_call <- VDJ_cor(gex.meta, "pearson", "d_call")

cor.s.j_call <- VDJ_cor(gex.meta, "spearman", "j_call")
cor.p.j_call <- VDJ_cor(gex.meta, "pearson", "j_call")

cor.s.c_call <- VDJ_cor(gex.meta, "spearman", "c_gene_10x")
cor.p.c_call <- VDJ_cor(gex.meta, "pearson", "c_gene_10x")


```


#### plot top VDJ genes per patient subgroup, Supplementary Figure S10 (SFig10A-C)

```{r}

get_topVDJ <- function(i, #seurat object
                       VDJgene = "tmp" #VDJ gene to plot
                       ) {
  
  #remove allele information
  vh_col <- unlist(lapply(strsplit(i@meta.data[, VDJgene], 
                                                   split = "\\*"),
                                          function(x) {return(x[1])})) 
  
  vh <- as.data.frame.matrix(table(vh_col, i$patient_group))
  vh <- vh[order(rowSums(vh), decreasing = TRUE), ]
  
  #export counts per patient group
  #write.csv(vh, paste0("../results/VDJ_geneUsage_", VDJgene, "_genesPerPatientGroup.csv")) 
  
  #get percentages of repertoir
  vh <-  as.data.frame(t(t(vh) / colSums(vh))) 
  vh$gene <- rownames(vh)
  
  nvh <- ifelse(nrow(vh) <= 50, nrow(vh), 50)
  
  titles <- data.frame(gexnames = c("v_call", "d_call", "j_call", "c_gene_10x"),
                       outnames = c("V", "D", "J", "IGH constant"))
  
  p <- as_tibble(vh[1:nvh, ]) %>% pivot_longer(
       cols = 1:4,
       names_to = "patient_group",
       names_prefix = "",
       values_to = "count",
       values_drop_na = TRUE) %>%  
          ggplot(aes(x = reorder(gene, -count), y = count, fill = patient_group)) +
              geom_col(position = "dodge") +
              theme_classic() +
              xlab("") +
              ylab("Percent of repertoire") +
              scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
              scale_fill_discrete(name = "Patient group") +
              ggtitle(paste0(titles$outnames[titles$gexnames == VDJgene], " genes")) +
              theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
                    axis.text.y = element_text(size = 6),
                    strip.text.y = element_text(size = 8)) +
              facet_grid(rows = vars(patient_group))

    ggsave2(paste0("../results/VDJ_geneUsage_", VDJgene, "_geneUsage_perPatientGroup_2.png"), 
          p,
          width = 30, height = 14, unit = "cm")
    
    return(p)
  
   }

pv <- get_topVDJ(gex, "v_call")
pd <- get_topVDJ(gex, "d_call")
pj <- get_topVDJ(gex, "j_call")
pc <- get_topVDJ(gex, "c_gene_10x")

ggsave2(paste0("../results/SFig10.png"), 
          plot_grid(pc, pd, pj, align = "hv", labels = "AUTO", ncol = 1) +
          bgcolor("white") + 
          border("white"),
          width = 15, height = 25, unit = "cm", dpi = 300)


```


#### plot IGHV usage per patient group, Figure 3C (Fig3C)

```{r}

vdj_gene <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group"), 
                       mode = "gene",
                       fill = TRUE)

vdj_gene$fc <- NA

#calculate FC relative to CTRL
table(vdj_gene$gene)
for (i in 1:nrow(vdj_gene)){
  ifelse(vdj_gene$patient_group[i] == "CTRL", 
         vdj_gene$fc[i] <- NA,
         vdj_gene$fc[i] <- log2((vdj_gene$seq_freq[i] * 100 + 1)/ (vdj_gene$seq_freq[vdj_gene$gene == vdj_gene$gene[i] & vdj_gene$patient_group == "CTRL"] *100  + 1)))
}

vdj_gene <- vdj_gene[order(vdj_gene$seq_count, decreasing = TRUE), ]
tmp <- vdj_gene[which(vdj_gene$gene %in% unique(vdj_gene$gene)[1:45]), ]

p1 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                      y = fc, 
                      colour = patient_group)) +
  theme_classic() +
      ggtitle(paste0("IGHV gene usage per patient group")) +
      theme(axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2, 0, 0, 0), "cm")) +
          ylab("log2FC to CTRL") +
          xlab("") +
          geom_point() +
      geom_hline(yintercept = 0)

p2 <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0, 0, 0, 0.2), "cm")) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

p_legend <- ggplot(tmp, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
  theme_classic() +
  scale_fill_discrete(name = "Patient group") +
  geom_col(position = "dodge")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p3 <- plot_grid(p1, p2, nrow = 2, labels = NULL,
                align = "v", axis = "l", rel_heights = c(1, 2.8))


 ggsave2(paste0("../results/VDJ_countGenes_", VDJgene 
                 ,".png"),
          plot_grid(p3, p_legend, 
                    ncol = 2, 
                    labels = NULL, 
                    align = "v", 
                    rel_widths = c(12, 2)) + bgcolor("white") + ,
          width = 20, height = 12, unit = "cm")


```


#### allele usage for differentially used IGHV genes

```{r}

vdj_allele <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group"), 
                       mode = "allele",
                       fill = TRUE)

#IGHV1-69
vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]

p1 <- ggplot(tmp1, aes(x = reorder(gene, -seq_freq), 
                       y = seq_freq, 
                       fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2, 0, 0, 0.2), "cm")) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-69 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup.png"), 
          plot_grid(p1, ncol = 1, labels = NULL, align = "h"),
          width = 15, height = 10, unit = "cm")

vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]


#
vdj_allele <- countGenes(gex@meta.data, 
                       gene = "v_call", 
                       groups = c("patient_group", "cluster_cellType_manual_main"),
                       mode = "allele",
                       fill = TRUE)

vdj_allele1 <- vdj_allele[grepl("^IGHV1-69*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp1 <- vdj_allele1[which(vdj_allele1$gene %in% unique(vdj_allele1$gene)[1:10]), ]


p1 <- ggplot(tmp1, aes(x = reorder(gene, -seq_freq), 
                       y = seq_freq, 
                       fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            axis.text.y = element_text(hjust = 1, 
                                       vjust = 0.5,
                                       size = 6),
            strip.background = element_blank(), 
            #strip.text.y = element_blank(),
            #legend.position = "none",
            plot.margin = unit(c(0.2, 0, 0, 0.2), "cm"),
            strip.text.y = element_text(angle = 0, vjust = 0)) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-69 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(cluster_cellType_manual_main))

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perCelltypeMain.png"), 
          plot_grid(p1, ncol = 1, labels = NULL, align = "h"),
          width = 15, height = 10, unit = "cm")


gex@meta.data$IGHV1_69_umi <- 0
gex@meta.data$IGHV1_69_umi[which(gex@meta.data$IGHV == "IGHV1-69")] <- gex@meta.data$umi_count[which(gex@meta.data$IGHV == "IGHV1-69")]

# featureplots for IGHV1_69 UMIs
p <- list()
for (i in sort(unique(gex$patient_group))){
p[[i]] <- FeaturePlot(gex %>% tidyseurat::filter(patient_group == i), 
            features = "IGHV1_69_umi",
            reduction = "umap_2",  
            raster = FALSE, 
            pt.size = 0.1, 
            ncol = 1, 
            order = TRUE,
            by.col = TRUE) +
  xlab("UMAP 1") +
  ylab("UMAP 2") +
  ggtitle(paste0(i, " IGHV1-69 UMI ")) +
  theme(axis.text = element_blank(),
          axis.ticks = element_blank())
  
}

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup_featureplot.png"), 
          plot_grid(plotlist = p, ncol = 2, labels = NULL, align = "h"),
          width = 20, height = 20, unit = "cm")


#percent of IGHV-69 cells per patient group per celltype fine
tmp <- as.data.frame.matrix(table(gex@meta.data$IGHV, gex@meta.data$cluster_cellType_manual_fine))
gex@meta.data$cluster_cellType_manual_fine <- as.factor(gex@meta.data$cluster_cellType_manual_fine)

p <- as_tibble(gex@meta.data) %>% 
  select(orig.ident, cluster_cellType_manual_fine, IGHV, patient_group) %>% 
  filter(IGHV == "IGHV1-69")  %>%  
  add_count(orig.ident, cluster_cellType_manual_fine, .drop = FALSE) %>% 
  distinct() %>%
  group_by(orig.ident) %>%
  mutate(per =  prop.table(n)) %>% 
  ggplot(aes(x = cluster_cellType_manual_fine,
             y = per,
             fill = patient_group)) +
  scale_y_continuous(labels = percent) +
  theme_classic() +
  xlab("") +
  ylab("percent of IGHV1-69 cells per patient group") +
  ggtitle("IGHV1-69") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  geom_boxplot()




gex.meta <- gex@meta.data[(gex$cluster_cellType_manual_fine == "DN1_ID3" |
                          gex$cluster_cellType_manual_fine == "DN2" |
                          gex$cluster_cellType_manual_fine == "DN2_CXCR3" |
                          gex$cluster_cellType_manual_fine == "DN2_ITGAX" |
                          gex$cluster_cellType_manual_fine == "DN4_IGHE" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_CD1C" |
                          gex$cluster_cellType_manual_fine == "Naive" |
                          gex$cluster_cellType_manual_fine == "Naive_IFN" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl_CD69_str" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl_str" |
                          gex$cluster_cellType_manual_fine == "Memory_Platlet" |
                          gex$cluster_cellType_manual_fine == "Naive_Transitional") &
                          gex$orig.ident !="P024", ]
gex.meta$cluster_cellType_manual_fine <- as.factor(gex.meta$cluster_cellType_manual_fine)
gex.meta$IGHV <- as.factor(gex.meta$IGHV)
gex.meta$orig.ident <- as.factor(gex.meta$orig.ident)

p <- select(gex.meta, orig.ident, cluster_cellType_manual_fine, IGHV) %>% 
  group_by(orig.ident, cluster_cellType_manual_fine, IGHV, .drop = FALSE) %>%
  summarise(n = n()) %>%
  mutate(percent = n/sum(n)) %>%
  filter(IGHV == "IGHV1-69")  %>% 
  left_join(select(gex.meta, orig.ident, patient_group) %>% distinct()) %>%
  ggplot(aes(x = cluster_cellType_manual_fine,
             y = percent,
             fill = patient_group)) +
  scale_y_continuous(labels = percent) +
  theme_classic() +
  xlab("") +
  ylab("percent IGHV1-69 cells per patient group") +
  ggtitle("IGHV1-69") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)

ggsave2(paste0("../results/VDJ_geneUsage_IGHV-69_perPatientGroupPerCelltype.png"), 
          p,
          width = 20, height = 12, unit = "cm")



####---- per sample

tmp <- data.frame(orig.ident = gex.meta$orig.ident, patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, levels = tmp$orig.ident)


p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
p[[i]] <- select(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ], 
                 orig.ident, 
                 IGHV) %>% 
  group_by(orig.ident,
           IGHV, 
           .drop = FALSE) %>%
  summarise(n = n()) %>%
  mutate(percent = n/sum(n)) %>%
  filter(IGHV == "IGHV1-69")  %>% 
  left_join(select(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ], 
                   orig.ident, 
                   patient_group) %>% 
              distinct()) %>%
  ggplot(aes(x = orig.ident,
             y = percent,
             fill = patient_group)) +
  scale_y_continuous(labels = percent) +
  theme_classic() +
  xlab("") +
  ylab("percent IGHV1-69 cells") +
  ggtitle(paste0(i, " IGHV1-69")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  geom_bar()
}







## get legend
p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == unique(gex$cluster_cellType_manual_fine)[1], ], 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "Patient group")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

# make plot
p1 <- plot_grid(plotlist = p, ncol = 3, labels = NULL, align = "hv")

ggsave2(paste0("../results/VDJ_muFull_celltypeFine_orig.ident_3.png"), 
          plot_grid(p1, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) +
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")


add_count(orig.ident, cluster_cellType_manual_fine, .drop = FALSE) %>%
  
ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup_perCelltypeFine.png"), 
          p,
          width = 16, height = 12, unit = "cm")


p <- as_tibble(gex@meta.data) %>% 
  select(orig.ident, cluster_cellType_manual_main, IGHV, patient_group) %>% 
  filter(IGHV == "IGHV1-69")  %>%  
  add_count(orig.ident, cluster_cellType_manual_main, .drop = FALSE) %>% 
  distinct() %>%
  group_by(orig.ident) %>%
  mutate(per =  prop.table(n)) %>% 
  ggplot(aes(x = cluster_cellType_manual_main,
             y = per,
             fill = patient_group)) +
  scale_y_continuous(labels = percent) +
  theme_classic() +
  xlab("") +
  ylab("percent of IGHV1-69 cells per patient group") +
  ggtitle("IGHV1-69") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  geom_boxplot()

ggsave2(paste0("../results/VDJ_geneUsage_alleles_IGHV-69_perPatientGroup_perCelltypeMain.png"), 
          p,
          width = 12, height = 12, unit = "cm")


# make pie chart with celltypeMain for each patient_group
# !!!! fix this !!!!
p <- as_tibble(gex@meta.data) %>% 
  select(cluster_cellType_manual_main, IGHV, patient_group) %>% 
  filter(IGHV == "IGHV1-69")  %>%  
  add_count(patient_group, cluster_cellType_manual_main, .drop = FALSE) %>% 
  distinct() %>%
  group_by(patient_group) %>%
  mutate(perc =  prop.table(n))

ggplot(p, aes(x=1,y = perc, fill = cluster_cellType_manual_main)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1) +
  theme_void() +
  geom_text(aes(label = c(paste0(round(p$perc[1:5]*100, digits = 2), "%"), 
                rep("", times = nrow(p) - 5))), 
            position = position_stack(vjust = 0.5, reverse = FALSE),
            size = 3.2)+
  
  xlim(0.5, 2.5) +
  ggtitle("") +
  scale_fill_discrete(name = "Cell type main") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5))

 
#IGHV1-18
vdj_allele2 <- vdj_allele[grepl("^IGHV1-18*", vdj_allele$gene), ] %>% arrange(desc(seq_count))
tmp2 <- vdj_allele2[which(vdj_allele2$gene %in% unique(vdj_allele2$gene)[1:10]), ]

p2 <- ggplot(tmp2, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0,0,0.2), "cm")) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV1-18 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))

#IGHV3-30-3
vdj_allele3 <- vdj_allele[grepl("^IGHV3-30-3*", vdj_allele$gene),] %>% arrange(desc(seq_count))
tmp3 <- vdj_allele3[which(vdj_allele3$gene %in% unique(vdj_allele3$gene)[1:10]), ]

p3 <- ggplot(tmp3, aes(x = reorder(gene, -seq_freq), 
                          y = seq_freq, 
                          fill = patient_group)) +
      theme_classic() +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5,
                                       size = 8), 
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0,0,0.2), "cm")) +
      ylab("Percent of allele repertoire") +
      xlab("") +
      ggtitle("IGHV3-30-3 alleles") +
      #scale_y_continuous(labels = percent) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(patient_group))


```


## clonotype analysis
### clonotype diversity

```{r}

####---------------------------------------------------------------------
####---- clonal diversity
####---------------------------------------------------------------------

####---- abundance estimation
patient_group_estimateAbundance <- estimateAbundance(gex@meta.data, 
                                              group = "patient_group", 
                                              clone = "cluster_no_airr",
                                              ci = 0.95, 
                                              nboot = 100)
saveRDS(patient_group_estimateAbundance, "../results/VDJ_clono_patient_group_estimateAbundance.rds")
patient_group_estimateAbundance <- readRDS("../results/VDJ_clono_patient_group_estimateAbundance.rds")

#Error: vector memory exhausted (limit reached?) 
#nboot = 100 works locally, nboot = 200 does not

#plotAbundanceCurve(patient_group_estimateAbundance) + theme_classic()

png("../results/VDJ_clono_Abundances_patient_group.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_abundances, 
     #colors = sample_colors, 
     main_title = "Clonal abundances", 
     legend_title = "patient group")
dev.off()

rm(patient_group_estimateAbundance); gc()

####---- diversity curve
####---- per orig.ident
orig.ident_alphaDiversity <- alphaDiversity(gex@meta.data, 
                                   group = "orig.ident", 
                                   clone = "cluster_no_airr",
                                   min_q = 0, 
                                   max_q = 4, 
                                   step_q = 0.1,
                                   ci = 0.95, 
                                   nboot = 50)

saveRDS(orig.ident_alphaDiversity, "../results/VDJ_clono_orig.ident_alphaDiversity.rds")
#orig.ident_alphaDiversity <- readRDS("../results/VDJ_clono_orig.ident_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_orig.ident.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(orig.ident_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "Clonal diversity", 
     legend_title = "Sample")
dev.off()

plotDiversityCurve(orig.ident_alphaDiversity,main_title = "Clonal diversity", 
     legend_title = "Sample") + theme_classic()


rm(orig.ident_alphaDiversity); gc()

####---- per patient group
patient_group_alphaDiversity <- alphaDiversity(gex@meta.data, 
                                   group = "patient_group", 
                                   clone = "cluster_no_airr",
                                   min_q = 0, 
                                   max_q = 4, 
                                   step_q = 0.1,
                                   ci = 0.95, 
                                   nboot = 100)

#saveRDS(patient_group_alphaDiversity, "../results/VDJ_clono_patient_group_alphaDiversity.rds")
patient_group_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_alphaDiversity.rds")
head(patient_group_alphaDiversity@tests)

patient_group_alphaDiversity@tests[patient_group_alphaDiversity@tests$pvalue < 0.1,]

png("../results/VDJ_clono_alphaDiversity_patient_group.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "Clonal diversity", 
     legend_title = "Patient group")
dev.off()

rm(patient_group_alphaDiversity); gc()

####---- per c_call
# remove c_call classes with too few cells
tmp <- gex@meta.data
tmp <- tmp[!is.na(tmp$c_call),]
tmp <- tmp[!(tmp$c_call == "IGHE" | 
             tmp$c_call == "IGLC1" |
             tmp$c_call == "IGLC2"), ]

c_call_alphaDiversity <- alphaDiversity(tmp, 
                                        group = "c_call", 
                                        clone = "cluster_no_airr",
                                        min_q = 0, 
                                        max_q = 4, 
                                        step_q = 0.1,
                                        ci = 0.95, 
                                        nboot = 100)

saveRDS(c_call_alphaDiversity, 
        "../results/VDJ_clono_c_call_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_c_call.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(c_call_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "c call")
dev.off()

####---- diversity per patient group and c_call
tmp$patient_group_c_call <- paste0(tmp$patient_group, "_", tmp$c_call)
patient_group_c_call_alphaDiversity <- alphaDiversity(tmp, 
                                        group = "patient_group_c_call", 
                                        clone = "cluster_no_airr",
                                        min_q = 0, 
                                        max_q = 4, 
                                        step_q = 0.1,
                                        ci = 0.95, 
                                        nboot = 100)

saveRDS(patient_group_c_call_alphaDiversity, 
        "../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")
patient_group_c_call_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_patient_group_c_call.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_c_call_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "patient group\nc call")
dev.off()

plotDiversityCurve(patient_group_c_call_alphaDiversity) + theme_classic()


####---- diversity per patient group and c_call
tmp$patient_group_c_call <- paste0(tmp$patient_group, "_", tmp$c_call)
patient_group_c_call_alphaDiversity <- alphaDiversity(tmp, 
                                        group = "patient_group_c_call", 
                                        clone = "cluster_no_airr",
                                        min_q = 0, 
                                        max_q = 4, 
                                        step_q = 0.1,
                                        ci = 0.95, 
                                        nboot = 100)

saveRDS(patient_group_c_call_alphaDiversity, 
        "../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")
patient_group_c_call_alphaDiversity <- readRDS("../results/VDJ_clono_patient_group_c_call_alphaDiversity.rds")

png("../results/VDJ_clono_alphaDiversity_patient_group_c_call.png", 
    width = 10, height = 8, units = "cm", res = 200)
plot(patient_group_c_call_alphaDiversity, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "patient group\nc call")
dev.off()

plotDiversityCurve(patient_group_c_call_alphaDiversity) 


####---- Calculate clonal diversity indices
#four samples, 2 DNEG, 2 SSA and 1 SSAB have less repertoir diversity!
#in which celltype?
#in cells expressing which IGHC?

tmp <- gex
tmp$patient_group_c_call <- paste0(tmp$patient_group, "_", tmp$c_call)

combined2 <- expression2List(tmp, split.by = "patient_group")
p <- clonalDiversity(combined2, cloneCall = "cluster_no_airr", group.by = "orig.ident", x.axis = "patient_group") 

p1 <- p + 
  theme_classic() + 
  xlab("") + 
  geom_text(check_overlap = TRUE,
            size = 2.5,
            aes(label = orig.ident)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

ggsave2(paste0("../results/VDJ_clono_diversityIndex_plots.png"),
        p1,
        width = 25, height = 12, unit = "cm")

```


### common clonotypes across samples

```{r}

####---------------------------------------------------------------------
####---- make pie charts
####---------------------------------------------------------------------

imm2 <- gex@meta.data[,c("cluster_no_airr", "orig.ident")]
imm3 <- table(imm2)
imm4 <- data.frame(table(rowSums(imm3 > 0)))
colnames(imm4) <- c("clonocount", "cellcount")
imm4$perc <- round(imm4$cellcount / sum(imm4$cellcount) * 100, digits = 5)
head(imm4)

####---- pie1 - all clonotypes
p1 <- ggplot(imm4, aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1) +
  geom_text(aes(label = c(paste0(round(imm4$perc[1], digits = 2), "%"), 
                rep("", times = nrow(imm4) - 1))), 
            position = position_stack(vjust = 0.5, reverse = TRUE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(format(sum(imm4$cellcount), big.mark = " "), " clonotypes")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5))

####---- get legend
p_legend <- ggplot(imm4, aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity", 
           #color = "white"
           ) +
  coord_polar(theta = "y", start = 0)+
  theme_void() +
  theme(legend.position = "bottom", legend.box = "horizontal") +
  scale_fill_discrete(name = "Samples per \nclonotype")

p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

####---- get colors
g <- ggplot_build(p1)
c.p1 <- unique(g$data[[1]]["fill"])

####---- pie2  for all but unique clonotypes
p2 <- ggplot(imm4[2:nrow(imm4), ], 
             aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1)+
  geom_text(aes(label = c(paste0(round(imm4$perc[2:5], digits = 2), "%"), 
                rep("", times = nrow(imm4) - 5))), 
            position = position_stack(vjust = 0.75, reverse = FALSE),
            size = 3.2)+
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(format(sum(imm4$cellcount[2:nrow(imm4)]), big.mark = " "), " clonotypes")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5)
        ) +
  scale_fill_manual(values = c.p1$fill[2:nrow(c.p1)])


####---- pie3  clonotypes present in â‰¥ 6 samples
p3 <- ggplot(imm4[6:nrow(imm4), ], 
             aes(x = 2, y = cellcount, fill = clonocount)) +
  geom_bar(stat = "identity") +
  coord_polar(theta = "y", start = 0, direction = 1)+
  geom_text(aes(label = c(paste0(round(imm4$cellcount[6:9], digits = 3)
                                 #, "%"
                                 ), 
                rep("", times = nrow(imm4) - 9))), 
            position = position_stack(vjust = 0.6, 
                                      reverse = FALSE),
            size = 3.2) +
  theme_void() +
  xlim(0.5, 2.5) +
  ggtitle(paste0(sum(imm4$cellcount[6:nrow(imm4)]), 
                 " clonotypes\npresent in â‰¥ 6 samples")) +
  scale_fill_discrete(name = "Samples per clonotype") +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5)) +
  scale_fill_manual(values = c.p1$fill[6:nrow(c.p1)])

p4 <- plot_grid(p1, p2, p3, nrow = 1, labels = NULL,
                align = "vh", axis = "l")

####---- save pies
ggsave2(paste0("../results/VDJ_clono_clonotypeDistribution_perSample_2.png"),
          plot_grid(p4, p_legend, 
                    ncol = 1, 
                    labels = NULL, 
                    align = "v", 
                    rel_heights = c(5, 2)) + bgcolor("white"),
          width = 20, height = 12, unit = "cm")


```


### heatmap of top clonotypes

```{r}

####--------------------------------------------------------------------- 
####---- get clonotypes present in â‰¥ 6 samples using output from compairr
####---------------------------------------------------------------------

gex.meta <- gex@meta.data[gex@meta.data$orig.ident != "P024", ]
pheno_2 <- unique(gex.meta[, c("orig.ident", "patient_group")])
annot_junc <- unique(gex.meta[, c("cluster_no_airr", "junction_aa")])

tmp <- as.data.frame.matrix(table(gex.meta$cluster_no_airr,
                                  gex.meta$orig.ident))

tmp <- tmp[rowSums(tmp > 0) > 5, ]
tmp <- tmp[, order(pheno_2$patient_group[match(colnames(tmp), 
                                               pheno_2$orig.ident)])]

tmp2 <- (tmp > 0) * 1
pheno3 <- as.data.frame.matrix(table(pheno_2$orig.ident, pheno_2$patient_group))
tmp3 <- as.matrix(tmp2) %*% as.matrix(pheno3)
tmp3 <- t(t(tmp3)/c(table(pheno_2$patient_group)))
h.1 <- hclust(dist(tmp3, method = "binary"), method = "complete")
#h.1 %>% as.dendrogram %>% plot

clr <- cutree(h.1, h = 0.2) # number of row clusters #15

table(clr)

ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                   DNEG = "#7CAE00", 
                                   SSA = "#00BFC4", 
                                   SSAB = "#C77CFF"),
                 cluster = scales::hue_pal()(15))

pheatmap(tmp,
         cluster_cols = FALSE,
         cluster_rows = h.1,
         #border_color = NA,
         cutree_rows = 5,
         annotation_col = data.frame(patient_group = factor(pheno_2$patient_group[match(colnames(tmp), pheno_2$orig.ident)]), row.names = colnames(tmp)),
         clustering_method = "ward.D2",
         color = c("gray95", colorRampPalette(c("gray70", 
                                                "orange3", 
                                                "firebrick"))(90)),
         fontsize_row = 6,
         fontsize_col = 8,
         main = paste0(nrow(tmp), " clonotypes occuring in â‰¥ 6 samples"),
         #labels_row = annot_junc$junction_aa[match(rownames(tmp), 
         #                                         annot_junc$cluster_no_airr)],
         show_rownames = FALSE,
         annotation_colors = ann_cols[1],
         legend = TRUE,
         filename = "../results/VDJ_clono_clonotypeDistribution_perSample_topClonotypes75_3.png"
         )  


```


### most prevelent clonotypes across samples

```{R}
gex.meta <- gex@meta.data[gex@meta.data$orig.ident != "P024", ]
pheno_2 <- unique(gex.meta[, c("orig.ident", "patient_group")])
tmp <- as.data.frame.matrix(table(gex.meta$cluster_no_airr,
                                  gex.meta$orig.ident))

#get clonotypes present in â‰¥ 2 samples
tmp <- tmp[rowSums(tmp > 0) > 1, ]

#export info for 1799 clonotypes
 # outfile <- gex.meta[gex.meta$cluster_no_airr %in% rownames(tmp), c("barcode", "orig.ident", "cluster_cellType_manual_fine", "cluster_cellType_manual_main", "cluster_no_airr", "v_call", "d_call", "j_call", "c_call", "IGHV", "c_gene_10x", "junction", "junction_aa", "cdr3", "patient_group", "mu_load", "mu_freq_total")]
 # write.csv(outfile, "../results/1799_clonotypes.csv", sep = ",", col.names = TRUE, row.names = FALSE, quote = TRUE)

#tmp <- read.csv("../results/1799_clonotypes.csv")

#make the clonotype data frame per sample binary
tmp <- (tmp > 0) * 1

#get number of samples for each clonotype within each subgroup
pheno3 <- as.data.frame.matrix(table(pheno_2$orig.ident, pheno_2$patient_group))
tmp3 <- as.matrix(tmp) %*% as.matrix(pheno3)

#normalize to number of samples in subgroup
tmp3 <- t(t(tmp3)/c(table(pheno_2$patient_group))) 
h.3 <- hclust(dist(tmp3, method = "binary"), method = "complete")
#saveRDS(h.3, "../results/h.3.rds")

#plot(as.dendrogram(h.3)); abline(h = 0.2, col = "red")
clr <- cutree(h.3, h = 0.2) # number of row clusters #15

tmp4 <- rowsum(tmp, clr) #number of clonotypes per sample in each row clade
h.4 <- hclust(dist(t(tmp4), method = "binary"), method = "complete")
#plot(as.dendrogram(h.4)); abline(h = 0.2, col ="red") 
clc <- cutree(h.4, h = 0.2) #number of colum clusters #4

ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                   DNEG = "#7CAE00", 
                                   SSA = "#00BFC4", 
                                   SSAB = "#C77CFF"),
                 cluster = scales::hue_pal()(15))

p1 <- pheatmap(tmp,
               cluster_cols = h.4,
               cluster_rows = h.3,
               #cluster_cols = FALSE,
               #border_color = NA,
               cutree_rows = length(unique(clr)),
               cutree_cols = length(unique(clc)),
               annotation_col = data.frame(patient_group = 
                         factor(pheno_2$patient_group[match(colnames(tmp),
                                                  pheno_2$orig.ident)], 
                                levels = c("CTRL", "DNEG", "SSA", "SSAB")), 
                                           row.names = colnames(tmp)),
               annotation_row = data.frame(cluster = factor(clr), row.names = names(clr)),
               annotation_names_row = TRUE,
               annotation_colors = ann_cols[1],
               color = c("gray95", colorRampPalette(c("gray70", 
                                                      "orange3", 
                                                      "firebrick"))(90)),
               legend = FALSE,
               fontsize_col = 6,
               treeheight_row = 40,
               treeheight_col = 30,
               main = paste0(nrow(tmp), " clonotypes occuring in â‰¥ 2 samples"),
               filename = "../results/VDJ_clono_clonotypeDistribution_perSample_topClonotypes_3.png",
               show_rownames = FALSE
         )


# library(dendextend)
# unique(p1$tree_row$labels[p1$tree_row$order])
# dendr <- h.3  %>% as.dendrogram
# dendc <- h.4  %>% as.dendrogram
# gplots::heatmap.2(tmp,
#                   dendrogram = "both",
#                   Rowv = dendr,
#                   Colv = dendc)
# dev.off()


unique(clr[p1$tree_row$order])
table(clr[p1$tree_row$order])

#Clusters containing CTRL
sum(table(clr[p1$tree_row$order])[c(1,4,6,8,9,12,12,15)]) #396

#pSS clusters, i.e. without clonotypes also in CTRL 
sum(table(clr[p1$tree_row$order]))-sum(table(clr[p1$tree_row$order])[c(1,4,6,8,9,12,12,15)]) #[1] 1403

#SSA/SSAB clusters
sum(table(clr[p1$tree_row$order])[c(5)]) #[1] 457




#### ---- venn diagram
tmp2 <- as.data.frame(t(rowsum(t(tmp), group = pheno$patient_group)))
tmp2 <- as.data.frame((tmp2 > 0) * 1)
tmp2$ident <- as.character(rownames(tmp2))

venn.list <- list()[1:4]
names(venn.list) <- colnames(tmp2)[1:4]
for (i in colnames(tmp2)[1:4]){
  for (j in 1:nrow(tmp2))
  venn.list[[i]][j] <- ifelse(tmp2[j, i] == 1, tmp2$ident[j], NA)
  #venn.list[[i]][venn.list[[i]] == "0"] <- NA
}

ann_cols <- list(patient_group = c(CTRL = "#F8766D", 
                                     DNEG = "#7CAE00", 
                                     SSA = "#00BFC4", 
                                     SSAB = "#C77CFF"))
venn.diagram(venn.list, 
             filename = "../results/Fig5B_I.png", 
             na = "none", 
             print.mode = c("raw", "percent"), 
             category.names = names(venn.list),
             disable.logging = TRUE,
             sigdigs = 1,
             cat.fontfamily = "sans",
             fontfamily ="sans",
             fill = ann_cols$patient_group,
             height = 15,
             width = 15, 
             resolution = 200,
             units = "cm",
             imagetype = "png",
             main = paste0(nrow(tmp2), " unique clonotypes present in â‰¥ 2 samples"),
             #sub = paste(""),
             main.fontfamily = "sans",
             main.cex = 1.5,
             cex = 1,
             cat.cex =1.2
             #sub.fontfamily = "sans"
             )






### clonotype hamming distance

```{R}

### !!! FIX this !!! there are more a clonotype can contain several junction_aa or junction_nt


annot_junc_aa <- unique(gex.meta[, c("cluster_no_airr", "junction_aa")])
#annot_junc_nt <- unique(gex.meta[, c("cluster_no_airr", "junction")])

xxx <- annot_junc_aa$junction_aa[annot_junc_aa$cluster_no_airr %in% rownames(tmp)]
names(xxx) <- annot_junc_aa$cluster_no_airr[annot_junc_aa$cluster_no_airr %in% rownames(tmp)]
length(xxx)
xxx <- annot_junc_nt$junction[annot_junc_nt$cluster_no_airr %in% rownames(tmp)]
names(xxx) <- annot_junc_nt$cluster_no_airr[annot_junc_nt$cluster_no_airr %in% rownames(tmp)]

my_order <- h.3$labels[h.3$order]
my_order <- setNames(1:length(my_order),my_order)
aa_order <- xxx[order(my_order[names(xxx)])]

res <- msa::msa(AAStringSet(xxx), method = "ClustalOmega", order = "input")
hemoAln2 <- msaConvert(res, type = "seqinr::alignment")
hemoAln2$nam <- xxx
d <- seqinr::dist.alignment(hemoAln2, "identity")
d[is.na(d)] <- 1

summary(d)
class(d)
class(as.matrix(d))
as_tibble(as.matrix(d))
head(as.matrix(d))
head(as.data.frame(as.matrix(d)))
class(as.data.frame(as.matrix(d)))

head(as.data.frame(as.matrix(d))[aa_order, aa_order])

as.matrix(d)[1:5,1:5]
as_tibble(as.matrix(d))[1:10, 1:10]
as_tibble(as.matrix(d))[aa_order, aa_order][1:10, 1:10]

rownames(as.data.frame(as.matrix(d)))
colnames(as.data.frame(as.matrix(d)))


 h.5 <- hclust(d, method = "ward.D2") 
 plot(as.dendrogram(h.5)); abline(h = 0.2, col ="red")

 h.6 <- hclust(as.matrix(d), method = "ward.D2")
# plot(as.dendrogram(h.6)); abline(h = 0.2, col ="red")

head(as.data.frame(as.matrix(d))[aa_order, aa_order][1:10, 1:10])
p3 <- pheatmap(as.data.frame(as.matrix(d))[aa_order, aa_order][1:100, 1:100],
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         #cluster_cols = FALSE,
         #cluster_rows = FALSE,
         #clustering_method = "ward.D2",
         #cutree_cols = length(unique(cl)),
         show_colnames = FALSE, 
         show_rownames = FALSE,
         #cutree_rows = length(unique(cl)),
         filename = "../results/VDJ_clono_clonotypeDistribution_hamm.dist_6.png",
         scale = "row",
         #color = c(colorRampPalette(c("navy", "gray95", "firebrick"))(90))
         color = c(colorRampPalette(c("firebrick", "gray95", "white"))(90)) 
        )

#[aa_order, aa_order]

p3 <- pheatmap(as.matrix(d),
         cluster_cols = h.5,
         cluster_rows = h.5,
         #cluster_cols = TRUE,
         #cluster_rows = TRUE,
         #clustering_method = "ward.D2",
         #cutree_cols = length(unique(cl)),
         show_colnames = FALSE, 
         show_rownames = FALSE,
         cutree_rows = 9,
         cutree_cols = 9,
         #cutree_rows = length(unique(cl)),
         filename = "../results/VDJ_clono_clonotypeDistribution_hamm.dist_10.png",
         scale = "row",
         color = c(colorRampPalette(c("navy", "gray95", "firebrick"))(90)) 
        )





#h <- njs(d)
#mypar(1,3)
# plot(h,cex=.6)
# 
# h2 <- hclust(d, method = "ward.D2")
# par(cex=.6, mar=c(4,2,1,30))
# mypar(1,1)
# plot(as.dendrogram(h2),horiz=T,axes=T,xpd=T, main="aa", xlab="distance")
# par(cex=1)
# add_letter(figlabels[1]); figlabels <- figlabels[-1]

```


###IGHV genes among most prevelent clonotypes

```{R}

tmp2 <- as.data.frame.matrix(table(gex.meta$cluster_no_airr,
                                   gex.meta$IGHV))
tmp2 <- tmp2[rownames(tmp), ]
tmp2 <- (tmp2 > 0) * 1 #make matrix binary

#number of clonotypes per cluster
tmp7 <- rowsum(tmp2, clr)
tmp7 <- tmp7[, colSums((tmp2 > 0) * 1) > 30]

tmp8 <- (tmp7 > 0) * 1
h.8 <- hclust(dist(t(tmp8), method = "binary"), method = "complete")

p4 <- pheatmap((tmp7[unique(clr[h.3$order ]), ]), 
         cluster_rows = FALSE,
         cluster_cols = h.8,
         color = c("gray95", colorRampPalette(c("gray70", 
                                                "orange3", 
                                                "firebrick"))(90)),
         treeheight_col = 30,
         border_color = NA,
         filename = "../results/VDJ_clono_clonotypeDistribution_IGHV_topClonotypes_counts.png")

#frequency of clonotypes per cluster
tmp9 <- rowsum(tmp2, clr)
tmp9 <- apply(as.matrix(tmp9), 2, function(x){x/table(clr)})
tmp9 <- tmp9[, colSums((tmp2 > 0) * 1) > 30]

tmp10 <- (tmp9 > 0) * 1
h.9 <- hclust(dist(t(tmp10), method = "binary"), method = "complete")


#binary heatmap for IGHV genes corresponding to clonotype heatmap
p5 <- pheatmap((tmp9[unique(clr[h.3$order ]), ]), 
         cluster_rows = FALSE,
         cluster_cols = h.9,
         color = c("gray95", colorRampPalette(c("gray70", 
                                                "orange3", 
                                                "firebrick"))(90)),
         treeheight_col = 30,
         border_color = NA,
         filename = "../results/VDJ_clono_clonotypeDistribution_IGHV_topClonotypes_clustfract.png")


p6 <- pheatmap(tmp2[, colSums((tmp2 > 0) * 1) > 30],
         cluster_cols = h.8,
         cluster_rows = h.3,
         #treeheight_row = 0,
         #border_color = NA,
         #annotation_col = data.frame(patient_group = factor(pheno_2$patient_group[match(colnames(tmp), pheno_2$orig.ident)]), row.names = colnames(tmp2)),
         #clustering_method = "ward.D2",
         color = c("gray95", colorRampPalette(c("gray70",
                                                "orange3",
                                                "firebrick"))(90)),
         fontsize_row = 5,
         fontsize_col = 6,
         legend = FALSE,
         show_rownames = FALSE,
         treeheight_row = 40,
         treeheight_col = 30,
         cutree_rows = length(unique(clr)),
         main = paste0(nrow(tmp2), " clonotypes occuring in â‰¥ 2 samples"),
         filename = "../results/VDJ_clono_clonotypeDistribution_IGHV_topClonotypes_2.png",
         # labels_row = annot_junc$junction_aa[match(rownames(tmp2),
         #                                           annot_junc$cluster_no_airr)]

         )

# dev.off()
# plot_list = list()
# plot_list[['p1']] = p1$gtable
# plot_list[['p2']] = p2$gtable

ggsave2(paste0("../results/VDJ_clono_clonotypeDistribution_topClonotypes_2.png"),
        plot_grid(plotlist = plot_list, 
                  ncol = 2, 
                  labels = NULL, 
                  #align = "h",
                  greedy = FALSE,
                  rel_widths = c(1, 1)) + bgcolor("white"),
        width = 25, height = 20, unit = "cm")


```


### clonotype correlation

```{R}

####---------------------------------------------------------------------
####---- clonotype correlation using cluster_no_airr
####---------------------------------------------------------------------
tmp <- gex@meta.data
tmp$celltype_patient_group <- paste0(tmp$cluster_cellType_manual_main, 
                                     "-", 
                                     tmp$patient_group)

tmp2 <- table(list(tmp$cluster_no_airr, 
                    tmp$celltype_patient_group))
  
  cor_tmp <- cor(tmp2, method = "spearman")
  cor_tmp[is.na(cor_tmp)] <- 0
  
p <-pheatmap(cor_tmp, 
                     annotation_col = data.frame(cell_type = factor(sub("-.*", "", colnames(cor_tmp))),
                                                 patient_group = factor(sub(".*-", "", colnames(cor_tmp))),
                                                  row.names = colnames(cor_tmp)),
                      breaks = seq(0, 1, length.out = 100),
                      filename = paste0("../results/VDJ_clono_spearman_per_cluster_PatientGroup.png"),
                      fontsize = 8, 
                      fontsize_row = 4, 
                      fontsize_col = 4,
                      border_color = NA,
                      main = paste("clonotype spearman correlation"),
                      color = colorRampPalette(c("lightblue", "gray30","firebrick2", "red"))(99))












StartracDiversity(gex, type = "Type", 
            sample = "Patient", by = "overall")

#tmp <- gex$junction_aa
tmp <- gex$cdr3
names(tmp) <- colnames(gex); head(tmp)

junc_dist <- pairwiseDist(tmp, dist_mat = getDNAMatrix(gap = 0))
#Error: vector memory exhausted (limit reached?)

cdr3_eq <- pairwiseEqual(tmp)

seq <- c(A="ATGGC", B="ATGGG", C="ATGGG", D="AT--C", E="NTGGG")
d <- pairwiseEqual(seq)
rownames(d) <- colnames(d) <- seq
d

clones2 <- countClones(gex@meta.data, 
                      groups = c("patient_group", "cluster_cellType_manual_main"), 
                      clone = "cluster_no_airr")

### Quantify clonotypes scRepertoire

quantContig(
  combined_B,
  cloneCall = "gene+nt",
  chain = "both",
  scale = TRUE,
  group = "patient_group",
  exportTable = FALSE
)
head(combined_B)
quantContig(combined_B, cloneCall="gene+nt", scale = T)

abundanceContig(combined_B, cloneCall = "gene", scale = F)
lengthContig(combined_B, cloneCall="aa", chain = "IGL") 

clonalProportion(combined_B, cloneCall = "gene") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

occupiedscRepertoire(gex, x.axis = "seurat_clusters")

clonalOverlap(combined_B, cloneCall = "gene+nt", 
              method = "morisita")

combined2 <- expression2List(gex, split.by = "cluster_cellType_manual_fine")
combined2 <- expression2List(gex, split.by = "cluster_cellType_manual_main")
str(combined2)

clonalHomeostasis(combined2, cloneCall = "nt") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

clonalProportion(combined2, cloneCall = "nt") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

clonalDiversity(gex, cloneCall = "nt",group.by = "cluster_cellType_manual_fine")


clonalOverlap(combined2, cloneCall="aa", method="overlap") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

alluvialClonotypes(gex, cloneCall = "gene", 
                   y.axes = c("orig.ident", "seurat_clusters"), 
                   color = "seurat_clusters")

p1 <- alluvialClonotypes(gex, cloneCall = "gene", 
                   y.axes = c("orig.ident", "cluster_cellType_manual_main"), 
                   color = "cluster_cellType_manual_main")
ggsave2("../results/alluvialClonotypes_test.png", p1, height = 20, width = 30, units = "cm")

#results <- hierarchicalClones(imm, threshold=0.1)


### clonotype overlap correlation across samples and patient groups using compairr
cairr.aa.x <- as.data.frame.matrix(table(cairr.aa$cluster_no, cairr.aa$repertoire_id))
cairr.aa.x <- cairr.aa.x[apply(cairr.aa.x, 1, function(x){sum(x >= 1) >= 2}), ]
cairr.aa.x <- cairr.aa.x[order(-rowSums(cairr.aa.x)),]


```


### clonotype sequence identity

```{r}
# READ IN DATA

colnames(gex@meta.data)
head(gex@meta.data$cdr3)

a <- read.delim('/Volumes/GoogleDrive/My Drive/pSS_manuscript/clono/VDJ_immcantation_out.tsv')
as.character(a$germline_alignment[1])
as.character(a$sequence_alignment[1])
as.character(a$sequence[1])
as.character(a$junction[1])

# REFORMAT STRUCTURE
xxx <- as.character( unlist(c(droplevels(a[1,c("sequence","sequence_alignment","germline_alignment","junction",
             "cdr1","cdr2","cdr3","fwr1","fwr2","fwr3","fwr4","junction_10x",
             "germline_alignment_d_mask")]))) )
names(xxx) <- c("sequence","sequence_alignment","germline_alignment","junction",
                "cdr1","cdr2","cdr3","fwr1","fwr2","fwr3","fwr4","junction_10x",
                "germline_alignment_d_mask")

# CONVERT TO SEQUENCE ALIGNMENT
res <- msa::msa(DNAStringSet(xxx,use.names = T),method = "ClustalW",order = "input")

# PRINT ON CONSOLE
print(res,show="complete")

# PRINT ON A PDF FILE
msaPrettyPrint(res,file = "ClustalW_aligment.pdf",output="pdf",
               showNames="left", showNumbering="none", showLogo="top",
               showConsensus="bottom", logoColors="rasmol",
               verbose=FALSE, askForOverwrite=FALSE)

# SAVE TO CSV
tmp <- as.character(res@unmasked)
writeLines(paste0( names(tmp),";",tmp ),sep = "\n",con = "ClustalW_aligment.csv")

# SAVE TO FASTA
write.fasta(sequences = tmp, names = names(tmp), file.out = "ClustalW_aligment.fasta")
```


### clonotype phylogeny

```{r}
library(phangorn)
#install.packages("phangorn")
BiocManager::install("phangorn")

fdir <- system.file("extdata/trees", package = "phangorn")
primates <- read.phyDat(file.path(fdir, "primates.dna"),
                        format = "interleaved")
tree <- pratchet(primates, trace = 0) |> acctran(primates)
parsimony(tree, primates)

library(seqLogo)
#install.packages("seqLogo")
BiocManager::install("seqLogo")

seqLogo()
seqLogo(t(subset(anc.mpr, getRoot(tree), 1:20)[[1]]), ic.scale = FALSE)


```


## mutation status
### plot mutation status per celltypes and samples
#### plot Fig4B, total VDJ mutation frequency - celltypeFine - patient_group

```{r}

p1 <- ggplot(gex.meta, 
            aes(x = cluster_cellType_manual_fine, 
                y = mu_freq_total, 
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5),
            plot.title = element_text(size = 12)) +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) 

ggsave2(paste0("../results/Fig4B.png"), 
          p1,
          width = 30, height = 10, unit = "cm")

```


#### plot Fig4C, CDR3 mutation count ratio R/S - cell type fine - patient_group,

```{r}
gex.meta$mu_count_cdr3_RSratio <- log2((gex.meta$mu_count_cdr3_r + 1) / 
                                         (gex.meta$mu_count_cdr3_s + 1))

#remove cells with no mutations
#Removed 181168 rows containing non-finite values (stat_ydensity). 
gex.meta$mu_count_cdr3_RSratio[(gex.meta$mu_count_cdr3_r + 
                                  gex.meta$mu_count_cdr3_s) == 0] <- NA

nrow(gex.meta[!is.na(gex.meta$mu_count_cdr3_RSratio), ])

p2 <- ggplot(gex.meta, 
             aes(x = cluster_cellType_manual_fine, 
                 y = mu_count_cdr3_RSratio, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0("CDR3 mutation ratio Replacement/Silent for ", 
                     nrow(gex.meta[!is.na(gex.meta$mu_count_cdr3_RSratio), ]),
                     " mutated cells")) +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 1, 
                                       vjust = 0.5),
            plot.title = element_text(size = 12)) +
      ylab("mutation count ratio R/S (log2)") +
      xlab("") +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)

ggsave2(paste0("../results/Fig4C.png"), 
          p2,
          width = 30, height = 10, unit = "cm")


```


#### plot Fig4D, fraction of CDR3 mutated cells per sample and celltypeFine

```{r}
p3 <- as_tibble(gex.meta) %>% 
  select(cluster_cellType_manual_fine, 
         patient_group,
         orig.ident,
         mu_count_cdr3_r, 
         mu_count_cdr3_s) %>% 
  replace_na(list(mu_count_cdr3_r = 0, mu_count_cdr3_s = 0)) %>%
  mutate(mu_cdr3_binary = mu_count_cdr3_r #+ mu_count_cdr3_s
         ,
         mu_cdr3_binary = (mu_cdr3_binary > 0) * 1,
         mu_cdr3_binary = recode(mu_cdr3_binary, `0` = "Unmutated", `1` = "Mutated")) %>%
  group_by(cluster_cellType_manual_fine,
           patient_group,
           orig.ident, 
           mu_cdr3_binary) %>%
  summarise(cdr3_n = n()) %>%
  mutate(cdr3_freq = cdr3_n / sum(cdr3_n)) %>% 
  filter(mu_cdr3_binary == "Mutated") %>%
  ggplot(aes(x = cluster_cellType_manual_fine, 
             y = cdr3_freq, 
             fill = patient_group)) +
  geom_violin(alpha = 0.8, 
              scale = "width", 
              draw_quantiles = c(0.25, 0.5, 0.75),
              trim = TRUE) +
  ylab("Proportion of cells with \nreplacement mutations within CDR3") +
  xlab("") +
  ggtitle("Replacement mutations in CDR3") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5),
        plot.title = element_text(size = 12)) +
  stat_compare_means(method = "anova", hide.ns = TRUE, label = "p.signif") +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1L)) 

ggsave2(paste0("../results/Fig4D.png"), 
          plot_grid(p3, labels = NULL, align = "h"),
          width = 30, height = 10, unit = "cm")

ggsave2(paste0("../results/Fig4B-D.png"), 
          plot_grid(p1, p2, p3, labels = NULL, align = "hv", ncol = 1) + 
          bgcolor("white") + 
          border("white"),
          width = 15, height = 30, unit = "cm")

```


#### plot SFig12A, total VDJ mutation frequency - celltypeMain - orig.ident 

```{r}

tmp <- data.frame(orig.ident = gex.meta$orig.ident, patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, levels = tmp$orig.ident)


p <- ggplot(gex.meta,
            aes(x = orig.ident, 
                y = mu_freq_total,
                fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5), 
            legend.position = "none") +
      ylab("mutated bp (%)") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)

## get legend
p_legend <- ggplot(gex.meta, 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "Patient group")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

ggsave2(paste0("../results/SFig12A.png"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 12, unit = "cm")


```


####  plot SFig12B, total VDJ mutation frequency - celltypeFine - orig.ident

```{r}

# relevel orig.ident to plot based on patient group order
tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
p[[i]] <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ], 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      theme_classic() +
      ggtitle("VDJ mutations") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            legend.position = "none") +
      ggtitle(i) +
      scale_y_continuous(labels = scales::percent_format(accuracy = 1L),
                         breaks = seq(0, 20, 5), 
                         limits = c(0, 20)) +
      scale_fill_discrete(name = "patient group") +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE)  
}

y_label <- ggdraw() + 
  draw_label("mutated bp (%)", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

## get legend
p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == unique(gex$cluster_cellType_manual_fine)[1], ], 
             aes(x = orig.ident, 
                 y = mu_freq_total, 
                 fill = patient_group)) +
      geom_violin(alpha = 0.8, 
                  scale = "width", 
                  draw_quantiles = c(0.25, 0.5, 0.75),
                  trim = TRUE) +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_discrete(name = "Patient group")
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

p1 <- plot_grid(plotlist = p, ncol = 3, labels = NULL, align = "hv") + 
          bgcolor("white") + 
          border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/SFig12B_2.png"),
        plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")

```


##  plot Fig3A IgH class per cell type, Figure 3A

```{r}

####---- IgH per cluster
p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE,
              order = TRUE,
              cols = c(brewer.pal(9, "Set3")),
              na.value = "white") + 
  ggtitle("IGH constant gene") + 
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          legend.position = "none")

p2 <- ggplot(gex.meta %>% filter(!is.na(c_gene_10x)), 
             aes(x = cluster_cellType_manual_fine,
                 fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") +
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      scale_fill_brewer(palette="Set3", na.value="grey", name = "c gene",  drop = FALSE)

ggsave2("../results/Fig3A_2.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h") + 
          bgcolor("white") + 
          border("white"), 
        width = 30, height = 14, unit = "cm")

```


##  plot Fig3D cloneType per cell type fine based on compairr data, Figure 3D

```{r}

p1 <- DimPlot(gex, 
              group.by = "cloneType_airr", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE,
              order = c("Single (0 < X <= 1)",
                        "Small (1 < X <= 5)",
                        "Medium (5 < X <= 20)",
                        "Large (20 < X <= 100)"), 
              cols = c(brewer.pal(4, "Dark2")),
              na.value = "white") +
  ggtitle("clonotype") +  
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none")

p2 <- ggplot(gex.meta %>% filter(!is.na(cloneType_airr)), 
             aes(x = cluster_cellType_manual_fine, 
                 fill = cloneType_airr)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      xlab("") + 
      facet_grid(rows = vars(patient_group)) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      scale_fill_brewer(palette="Dark2", na.value = "grey", name = "clonotype")

ggsave2("../results/Fig3D_2.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h") + 
          bgcolor("white") + 
          border("white"), 
        width = 30, height = 14, unit = "cm")

table(gex$orig.ident, gex$cloneType_airr)

```


##  plot Fig4A, mutation load per cell type, Figure 4A

```{r}

p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.6, ncol = 2, 
              raster = TRUE, 
              cols = c(brewer.pal(4, "Dark2")),
              order = c("unmutated", "low", "med", "high", NA),
              na.value = "white") +
  ggtitle("Mutation load") +
  xlab("UMAP 1") + 
  ylab("UMAP 2") + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "none")


p2 <- ggplot(gex.meta %>% filter(!is.na(mu_load)), 
             aes(x = cluster_cellType_manual_fine, 
                 fill = mu_load)) +
          theme_classic() +
          geom_bar(position = "fill") + 
          ylab("proportion of cells") +
          xlab("") +
          facet_grid(rows = vars(patient_group)) + 
          theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
          scale_fill_brewer(palette = "Dark2", na.value = "grey", name = "Mutation load")

ggsave2("../results/Fig4A_4.png", 
        plot_grid(p1, p2, ncol = 2, labels = NULL, align = "h") + 
          bgcolor("white") + 
          border("white"), 
        width = 30, height = 14, unit = "cm")


gex.meta$v_call_tmp <-  unlist(lapply(strsplit(gex.meta$v_call, 
                                         split = "\\*"),
                                function(x) {return(x[1])}))
table(gex.meta$v_call_tmp, gex.meta$mu_load)

```


## plot SFig13A, IgH class per cellTypeMain per patient

```{r}

#gex.meta$c_gene_10x <- factor(gex.meta$c_gene_10x, levels = sort(unique(gex.meta$c_gene_10x)))

tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- ggplot(gex.meta %>% filter(!is.na(c_gene_10x)),
            aes(x = orig.ident,
                fill = c_gene_10x)) +
      theme_classic() +
      ggtitle("IGH constant") +
      theme(axis.text.x = element_text(angle = 90, 
                                       hjust = 0.5, 
                                       vjust = 0.5), 
            legend.position = "none") +
      ylab("proportion of cells") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      #scale_fill_discrete(name = "patient group") +
      facet_grid(cols = vars(cluster_cellType_manual_main)) +
      geom_bar(position = "fill") +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "c gene", drop = FALSE)

## get legend
p_legend <- ggplot(gex.meta, 
             aes(x = orig.ident,
                 fill = c_gene_10x)) +
      geom_bar(position = "fill") +
      theme_classic() + 
      theme(legend.position = "bottom", legend.box = "horizontal") +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "c gene", drop = FALSE)
  
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()

ggsave2(paste0("../results/SFig13A.png"), 
          plot_grid(p, p_legend, 
                    ncol = 1, labels = NULL, 
                    align = "hv", rel_heights = c(1.5, 0.2)) + 
          bgcolor("white") + 
          border("white"),
          width = 30, height = 12, unit = "cm")



```


## plot SFig13B, IgH class per cellTypeFine per patient

```{r}

#gex.meta$c_gene_10x <- factor(gex.meta$c_gene_10x, levels = sort(unique(gex.meta$c_gene_10x)))

tmp <- data.frame(orig.ident = gex.meta$orig.ident, 
                  patient_group = gex.meta$patient_group)
tmp <- tmp[order(tmp$patient_group), ]
tmp <- unique(tmp)
gex.meta$orig.ident <- factor(gex.meta$orig.ident, 
                              levels = tmp$orig.ident)

p <- list()
for (i in sort(unique(gex.meta$cluster_cellType_manual_fine))){
p[[i]] <-  ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == i, ] %>% 
                    filter(!is.na(c_gene_10x)), 
                  aes(x = orig.ident,
                      fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ggtitle(i) +
      scale_x_discrete(drop = FALSE) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            legend.position = "none") +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "c gene", drop = FALSE) 

print(table(gex.meta$c_gene_10x[gex.meta$cluster_cellType_manual_fine == i]))

}

p_legend <- ggplot(gex.meta[gex.meta$cluster_cellType_manual_fine == unique(gex$cluster_cellType_manual_fine)[1], ], 
             aes(x = orig.ident,
                 fill = c_gene_10x)) +
      geom_bar(position = "fill") +
      theme_classic() + 
      theme(legend.position = "bottom", 
            legend.box = "horizontal", 
            panel.border = NULL) +
      scale_fill_brewer(palette = "Set3", na.value = "grey", name = "c gene", drop = FALSE)

p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot() 

y_label <- ggdraw() + 
  draw_label("proportion of cells", 
             fontface = 'bold',
             #x = 0.1,
             hjust = 0.5, 
             angle = 90,vjust = 0)

p1 <- plot_grid(plotlist = p, ncol = 3, labels = NULL, align = "hv")  + 
  bgcolor("white") + 
  border("white")

p2 <- plot_grid(y_label, p1, 
          ncol = 2, labels = NULL, 
          align = "hv", 
          rel_widths = c(0.03, 1)) + 
  bgcolor("white") + 
  border("white")

ggsave2(paste0("../results/SFig13B_2.png"), 
          plot_grid(p2, p_legend, ncol = 1, labels = NULL, align = "hv", rel_heights = c(1, 0.1)) +
          bgcolor("white") + 
          border("white"),
          width = 30, height = 25, unit = "cm")


```


##  plot SFig14, CDR3 aa attributes Supplementary Figure 14 

```{r}

# "cdr3_aa_length"               
# "cdr3_aa_gravy"               
# "cdr3_aa_bulk"                 
# "cdr3_aa_aliphatic"
# "cdr3_aa_polarity"             
# "cdr3_aa_charge"              
# "cdr3_aa_basic"               
# "cdr3_aa_acidic"
# "cdr3_aa_aromatic"

#### ----- subeset gex to only include relevant cell types
gex.meta <- gex@meta.data[(gex$cluster_cellType_manual_fine == "DN1_ID3" |
                          gex$cluster_cellType_manual_fine == "DN2" |
                          gex$cluster_cellType_manual_fine == "DN2_CXCR3" |
                          gex$cluster_cellType_manual_fine == "DN2_ITGAX" |
                          gex$cluster_cellType_manual_fine == "DN4" |
                          gex$cluster_cellType_manual_fine == "Memory_Cl" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                          gex$cluster_cellType_manual_fine == "Memory_IgM_CD1C" |
                          gex$cluster_cellType_manual_fine == "Naive" |
                          gex$cluster_cellType_manual_fine == "Naive_IFN" |
                          gex$cluster_cellType_manual_fine == "Naive_Transitional") &
                          gex$orig.ident !="P024", ]

#CDR3 aa length
p1 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_length,
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa length") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("length (bp)") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#grand average of hydrophobicity
p2 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_gravy,
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa grand average of hydrophobicity") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("hydrophobicity score") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#average bulkiness of amino acids
p3 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_bulk, #average bulkiness of amino acids
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa bulkiness") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("average bulkiness of amino acids") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aliphatic
p4 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aliphatic, #aliphatic index
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aliphatic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("aliphatic index") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_polarity
p5 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_polarity, #average polarity of amino acids
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa polarity") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("average polarity") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_charge
p6 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_charge, #net charge
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa charge") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("net charge") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)

#cdr3_aa_basic
p7 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_basic, #fract. of informative positions that are Arg, His or Lys
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa basic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("% of pos that are Arg, His or Lys") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_acidic
p8 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_acidic, #fraction of informative positions that are Asp or Glu
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa acidic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8),
            legend.position = "none") +
      ylab("% of pos. that are Asp or Glu") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)


#cdr3_aa_aromatic
p9 <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic, #fract. of informative pos. that are His, Phe, Trp or Tyr
                  color = patient_group)) +
      theme_classic() +
      ggtitle("CDR3 aa aromatic") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            axis.title.y = element_text(size = 8), 
            legend.position = "none") +
      ylab("% of pos. that are His, Phe, Trp or Tyr") +
      xlab("") +
      geom_boxplot(size = 0.3, alpha = 0.8)

#get legend
p_legend <- ggplot(gex.meta,
              aes(x = cluster_cellType_manual_fine, 
                  y = cdr3_aa_aromatic,
                  color = patient_group)) +
      theme_classic() +
      geom_boxplot() +
      theme(legend.position="bottom")
p_legend <- get_legend(p_legend) %>% ggpubr::as_ggplot()


p_grid <- plot_grid(p1, p2, p3, p4, p5, p6, p7, p8, p9, 
                    nrow = 3, ncol = 3, labels = "AUTO", align = "hv")

ggsave2(paste0("../results/SFig14.png"), 
          plot_grid(p_grid, p_legend, ncol = 1, align = "none", rel_heights = c(18, 1)) +
          bgcolor("white") + 
          border("white") ,
          width = 35, height = 30, unit = "cm")


```


## Save rds GEX7_VDJ

```{r}

#saveRDS(gex, file = "../results/GEX7_BVDJ.rds")
#gex <- readRDS("../results/GEX7_BVDJ.rds")

```


## Subset and save slim rds GEX7_VDJ

```{r}
gex.slim <- gex

print(object.size(gex.slim), humanReadable = TRUE, units = "GB") #7.2 Gb

#gex.slim@assays$RNA@counts <- matrix()
gex.slim@assays$RNA@scale.data <- matrix()
to_remove <- c("pca_1", "pca_2", 
               "scpred","scpred_projection", 
               "harmony_1", "harmony_2", 
               "umap_1", "umap_2_3d")
gex.slim@reductions <- gex.slim@reductions[!(names(gex.slim@reductions) %in% to_remove)]
#gex.slim <- gex.slim[, gex.slim$orig.ident != "P024"]
print(object.size(gex.slim), humanReadable = TRUE, units = "GB") #4.3 Gb
#saveRDS(gex.slim, file = "../results/GEX7_BVDJ_slim.rds")

gex.slim.fine <- gex.slim[, (gex.slim$cluster_cellType_manual_fine == "DN1_ID3" |
                              gex.slim$cluster_cellType_manual_fine == "DN2" |
                              gex.slim$cluster_cellType_manual_fine == "DN2_CXCR3" |
                              gex.slim$cluster_cellType_manual_fine == "DN2_ITGAX" |
                              gex.slim$cluster_cellType_manual_fine == "DN4" |
                              gex.slim$cluster_cellType_manual_fine == "Memory_Cl" |
                              gex.slim$cluster_cellType_manual_fine == "Memory_IgM" |
                              gex.slim$cluster_cellType_manual_fine == "Memory_IgM_ALOX5" |
                              gex.slim$cluster_cellType_manual_fine == "Memory_IgM_CD1C" |
                              gex.slim$cluster_cellType_manual_fine == "Naive" |
                              gex.slim$cluster_cellType_manual_fine == "Naive_IFN" |
                              gex.slim$cluster_cellType_manual_fine == "Naive_Transitional")]
print(object.size(gex.slim.fine), humanReadable = TRUE, units = "GB") #3.1 Gb


#saveRDS(gex.slim.fine, file = "../results/GEX7_BVDJ_slim_fine.rds")



```


## Print sessionInfo()

```{r}

sessionInfo()

```



