---
output: html_document
editor_options: 
  chunk_output_type: console
---

# VDJ analysis pt1 {#VDJ1}
## Load packages

```{r}

source("./00_dependencies.R")

```


## Load Seurat object

```{r}

gex <- readRDS("../results/GEX6_BANNOT.rds"); gex

```


## Load pheno file

```{r}

pheno <- read.table("../suppl/pheno/pSS_pheno_210623.csv", sep = ";", header = TRUE)
colnames(pheno)[c(1, 3)] <- c("Sample_id", "orig.ident")
pheno$orig.ident <- gsub("P007[ab]", "P007", pheno$orig.ident)
pheno$orig.ident <- gsub("P008[ab]", "P008", pheno$orig.ident)
pheno <- unique(pheno[, c(1, 3, 5, 7:8, 11:12)])

pheno$patient_group <- paste0(ifelse(pheno$SSA == "YES", "SSA", ""), 
                            "_", 
                            ifelse(pheno$SSB == "YES", "SSB", ""))
pat_annot <- c(SSAB = "SSA_SSB", SSA = "SSA_", DNEG="NA_NA", DNEG="_")
pheno$patient_group <- names(pat_annot)[match(pheno$patient_group, pat_annot)]
pheno$patient_group[grep("^C00", pheno$orig.ident) ] <- "CTRL"

head(pheno); dim(pheno)

#Four non-overlapping groups
#1. SSAB - SSA+/SSB+
#2. SSA  - SSA+/SSB-
#3. DNEG - SSA-/SSB-
#4. CTRL - CTRL


```


## immcantation
### Load output from Immcantation and perform mutation analysis

```{r}

files_B <- list.files("../data/clono_B_VDJ_tmp", 
                      pattern = "*germ-pass.tsv", 
                      recursive = TRUE, 
                      full.names = TRUE)

outfile <- NULL

for (i in files_B) {
  
  cat("reading: ", print(i), "\n")
  
  tmp <- read_rearrangement(i)
  
  #get sample name
  j <- gsub("_heavy_germ-pass.tsv", "", gsub(".*out/", "", i))
  cat("sample name: ", print(j), "\n")
  
  #Get number of VDJ mutations per cell
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = FALSE, 
                           combine = TRUE,
                           nproc = 4,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_count'] <- 'mu_count_total'
  
  #Get frequency of VDJ mutations per cell
  tmp <- observedMutations(tmp, 
                           sequenceColumn = "sequence_alignment", 
                           germlineColumn = "germline_alignment_d_mask",
                           frequency = TRUE, 
                           combine = TRUE,
                           nproc = 4,
                           cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq'] <- 'mu_freq_total'

  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = FALSE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = 4,
                          cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_count_v_r'] <- 'mu_count_v_r_total'
  colnames(tmp)[colnames(tmp) == 'mu_count_v_s'] <- 'mu_count_v_s_total'
  
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_V_BY_SEGMENTS, 
                          combine = FALSE,
                          nproc = 4,
                          cloneColumn = "clone_id")
  
  colnames(tmp)[colnames(tmp) == 'mu_freq_v_r'] <- 'mu_freq_v_r_total'
  colnames(tmp)[colnames(tmp) == 'mu_freq_v_s'] <- 'mu_freq_v_s_total'
  
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ_BY_REGIONS, 
                          combine = FALSE,
                          nproc = 4,
                          cloneColumn = "clone_id")
  
  tmp <- observedMutations(tmp, 
                          sequenceColumn = "sequence_alignment", 
                          germlineColumn = "germline_alignment_d_mask",
                          frequency = TRUE, 
                          regionDefinition = IMGT_VDJ, 
                          combine = FALSE,
                          nproc = 4,
                          cloneColumn = "clone_id")
  
  #amino acid properties for CDR3 aa's
  tmp <- aminoAcidProperties(tmp, seq = "junction", nt = TRUE, 
                             trim = TRUE, label = "cdr3")
  
  #tmp$sample <- gsub("a|b", "", j)
  
  tmp$cell_id <- paste0(j, "_", gsub("-1", "", tmp$cell_id))
  rownames(tmp) <- tmp$cell_id

  
  if (is.null(outfile)) {
    outfile <- tmp
  } else {
    outfile <- rbind(outfile, tmp)
  }
  
}

head(outfile)

  
```


### save immcantation VDJ data

```{r}

write_rearrangement(outfile, "../results/VDJ_immcantation_out.tsv")
#outfile <- read_rearrangement("../results/VDJ_immcantation_out.tsv")

```


### add pheno data to immcantation object

```{r}

outfile$orig.ident <- gsub("a|b", "", gsub("_.*", "", outfile$cell_id))
outfile <- merge(outfile, 
                 pheno[, c(2,8)], 
                 by.x = "orig.ident", 
                 by.y = "orig.ident", 
                 all.x = TRUE); head(outfile)
outfile <- outfile[, c(2:ncol(outfile), 1)]
head(outfile)


```


### integrate immcantation VDJ data with seurat object

```{r}

#select immcantation columns
colnames(outfile)
cols <- c("sequence_id", "productive", "vj_in_frame", "stop_codon", "locus", 
          "v_call", "d_call", "j_call", "c_call", 
          "cell_id", "clone_id", "umi_count", 
          "v_call_10x", "d_call_10x", "j_call_10x", 
          "germline_v_call","germline_d_call", "germline_j_call",
          "cdr3",
          "mu_count_total", "mu_freq_total", 
          "mu_count_v_r_total", "mu_count_v_s_total", "mu_freq_v_r_total", "mu_freq_v_s_total",
          "mu_freq_cdr1_r", "mu_freq_cdr1_s", "mu_freq_cdr2_r", "mu_freq_cdr2_s", "mu_freq_cdr3_r", "mu_freq_cdr3_s", "mu_freq_fwr1_r", "mu_freq_fwr1_s", "mu_freq_fwr2_r", "mu_freq_fwr2_s", "mu_freq_fwr3_r", "mu_freq_fwr3_s", "mu_freq_fwr4_r", "mu_freq_fwr4_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "mu_freq_cdr_r", "mu_freq_cdr_s", "mu_freq_fwr_r", "mu_freq_fwr_s", "cdr3_aa_length", "cdr3_aa_gravy", "cdr3_aa_bulk", "cdr3_aa_aliphatic", "cdr3_aa_polarity", "cdr3_aa_charge", "cdr3_aa_basic", "cdr3_aa_acidic", "cdr3_aa_aromatic")
outfile_add <- outfile[, cols]
rownames(outfile_add) <- outfile_add$cell_id

gex <- AddMetaData(gex, outfile_add)

sort(table(gex@meta.data$cdr3),decreasing = T)[1:20]


```


### VDJ mutation
#### add categorical variable for mutation burden

```{r}

# unmutated
# low < 1%
# med < 2%
# high > 2%

gex$mu_load <- ifelse(gex$mu_freq_total == 0, "unmutated", 
                 ifelse(gex$mu_freq_total > 0 & gex$mu_freq_total <= 0.01, "low",
                        ifelse(gex$mu_freq_total > 0.01 & gex$mu_freq_total <= 0.02, "med",
                               ifelse(gex$mu_freq_total > 0.02, "high", NA))))


```


## scRepertiore BCR
### Read 10x BCR files

```{r}

# read 10x VDJ BCR files
samples_B <- list.files("../data/clono_B_VDJ")

clono_B <- lapply(samples_B, function(x){
  
  tmp <- read.csv(paste0("../data/clono_B_VDJ/", x, "/", x, "_B_filtered_contig_annotations.csv"), 
                  stringsAsFactors = FALSE)
  return(tmp)
  
})

names(clono_B) <- samples_B

#remove trailing -1
clono_B <- lapply(seq_along(clono_B), function(i){
  clono_B[[i]]$barcode <- gsub("-1", "", clono_B[[i]]$barcode)
  return(clono_B[[i]])
})

names(clono_B) <- samples_B

clono_B <- clono_B[order(names(clono_B))]

print(object.size(clono_B), units = "MB") #227.6 Mb
saveRDS(clono_B, paste0("../results/clono_B_list.rds"))
#clono_B <- readRDS("../results/clono_B_list.rds")
str(clono_B)


```


### add 10x IgH call to Seurat object

```{r}

igh_10x <- lapply(1:length(clono_B), function(x) {
  
  tmp <- clono_B[[x]][clono_B[[x]]$chain == "IGH" & 
                        clono_B[[x]]$productive == "true" & 
                        !is.na(clono_B[[x]]$c_gene) & 
                        clono_B[[x]]$c_gene != "" &
                        clono_B[[x]]$c_gene != "IGLC1" &
                        clono_B[[x]]$c_gene != "IGLC2" , c(1, 10)]
  tmp$barcode <- paste0(names(clono_B)[x], "_", tmp$barcode)
  names(tmp)[2] <- "c_gene_10x"

  tmp <- tmp[!duplicated(tmp$barcode), ]
  
  rownames(tmp) <- tmp$barcode
  tmp$barcode <- NULL
  
  return(tmp)
  }
)

igh_10x <- do.call(rbind.data.frame, igh_10x)
table(igh_10x$c_gene_10x, useNA = "always")
gex <- AddMetaData(gex, igh_10x)


```

### generate a combined data frame for BCR sequences using scRepertoire

```{r}

#split combineBCR and do it per sample to avoid memory issues
#takes on average 20min/sample to run
combined_B <- lapply(1:length(clono_B), function(x) {
  
  print(names(clono_B[x]))
  print(nrow(clono_B[[x]]))
  print(Sys.time())
  
  tmp <- combineBCR(clono_B[[x]],
                    samples = names(clono_B[x]),
                    ID = NULL,
                    removeNA = TRUE,
                    removeMulti = TRUE)

  return(tmp)
  
})

#reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
combined_B <- lapply(1:length(combined_B), function(x){
  tmp <- combined_B[[x]][[1]]
  return(tmp)
})

#---- handle samples where libraries were created and sequenced twice
combined_B[["P007"]] <- rbind(combined_B[["P007a"]], combined_B[["P007b"]])
combined_B[["P007a"]] <- NULL
combined_B[["P007b"]] <- NULL

combined_B[["P008"]] <- rbind(combined_B[["P008a"]], combined_B[["P008b"]])
combined_B[["P008a"]] <- NULL
combined_B[["P008b"]] <- NULL

combined_B <- combined_B[order(names(combined_B))]

names_tmp <- names(combined_B)
combined_B <- lapply(seq_along(combined_B), function(i) {
  combined_B[[i]]$sample <- gsub("a|b", "", combined_B[[i]]$sample)
  return(combined_B[[i]])
})

names(combined_B) <- names_tmp
#----

print(object.size(combined_B), units = "MB") #181.3 Mb
saveRDS(combined_B, paste0("../results/combined_BCR_scRepertoire.rds"))
#combined_B <- readRDS("../results/combined_BCR_scRepertoire.rds")
head(combined_B[[1]])


```


## Integrate scRepertoir out with Seurat object

```{r}

gex <- combineExpression(combined_B, gex)
colnames(gex@meta.data)
table(gex@meta.data$CTgene, useNA = "ifany")
gex@meta.data$cloneType
head(gex@meta.data[, c(77:84)])

```


## TRUST4
### read TRUST4 files

```{r}

# samples_TRUST4 <- sub("_barcode_report.tsv", "", list.files("../data/clono_TRUST4"))
# clono_TRUST4 <- lapply(samples_TRUST4, function(x){
#   tmp <- read.table(paste0("../data/clono_TRUST4/", x, "_barcode_report.tsv"), header = TRUE, comment.char = "")
#   colnames(tmp) <- sub("X." , "", colnames(tmp))
#   return(tmp)
# })
# 
# names(clono_TRUST4) <- samples_TRUST4
# 
# # remove trailing -1
# # combineTRUST4 adds sample name to barcodes as prefix
# clono_TRUST4 <- lapply(seq_along(clono_TRUST4), function(i) {
#   clono_TRUST4[[i]][, 1] <- gsub("-1", "", clono_TRUST4[[i]][, 1])
#   print(names(clono_TRUST4[i]))
#   return(clono_TRUST4[[i]])
# })
# 
# names(clono_TRUST4) <- samples_TRUST4
# 
# saveRDS(clono_TRUST4, paste0("../results/clono_TRUST4_list.rds"))
# #clono_TRUST4 <- readRDS("../results/clono_TRUST4_list.rds")


```


### combine TRUST4 data using scRepertoire

```{r}

# combined_TRUST4_B <- lapply(1:(length(clono_TRUST4)-1), function(x) {
#   
#   print(names(clono_TRUST4[x]))
#   print(nrow(clono_TRUST4[[x]]))
#   print(Sys.time())
#   
#   tmp <- combineTRUST4(clono_TRUST4[[x]],
#                     samples = names(clono_TRUST4[x]),
#                     ID = NULL,
#                     cells = c("B"),
#                     removeNA = FALSE)
# 
#   return(tmp)
#   
# })
# 
# #reformat output to fecilitate downstream usage by scRepertoire and integration with Seurat object
# combined_TRUST4_B <- lapply(1:length(combined_TRUST4_B), function(x){
#   tmp <- combined_TRUST4_B[[x]][[1]]
#   return(tmp)
# })
# 
# 
# names(combined_TRUST4_B) <- names(clono_TRUST4)[1:(length(clono_TRUST4)-1)]
# 
# #---- handle samples where libraries were created and sequenced twice
# combined_TRUST4_B[["P007"]] <- rbind(combined_TRUST4_B[["P007a"]], combined_TRUST4_B[["P007b"]])
# combined_TRUST4_B[["P007a"]] <- NULL
# combined_TRUST4_B[["P007b"]] <- NULL
# 
# combined_TRUST4_B[["P008"]] <- rbind(combined_TRUST4_B[["P008a"]], combined_TRUST4_B[["P008b"]])
# combined_TRUST4_B[["P008a"]] <- NULL
# combined_TRUST4_B[["P008b"]] <- NULL
# 
# combined_TRUST4_B <- combined_TRUST4_B[order(names(combined_TRUST4_B))]
# 
# names_tmp <- names(combined_TRUST4_B)
# combined_TRUST4_B <- lapply(seq_along(combined_TRUST4_B), function(i) {
#   combined_TRUST4_B[[i]]$sample <- gsub("a|b", "", combined_TRUST4_B[[i]]$sample)
#   return(combined_TRUST4_B[[i]])
# })
# 
# names(combined_TRUST4_B) <- names_tmp
# #----
# 
# saveRDS(combined_TRUST4_B, paste0("../results/combined_TRUST4_B_scRepertoire.rds"))
# #combined_TRUST4_B <- readRDS("../results/combined_TRUST4_B_scRepertoire.rds")


```


## Analyse and visualize immcantation and scRepertoir output
### VDJ gene usage
#### VDJ correlation

```{r}

#VDJ gene usage correlation
VDJ_cor <- function(i, #seurat object containing meta data for VDJ gene usage
                    correl, #which correlation to use i.e. spearman or pearson...
                    VDJgene #VDJ gene to plot
                    ){
  
  cat(VDJgene, "\n") 
  cat(correl, "\n")
  
      
  i@meta.data[, VDJgene] <- unlist(lapply(strsplit(i@meta.data[, VDJgene], split = ","), function(x) {
  ifelse(length(x) > 1 , return(x[1]), return(x))})) #keep only first allele if several exist
  
  i$seurat_clusters_patient_group <-paste0(i$seurat_clusters, "_", i$patient_group)
  tmp <- table(list(i@meta.data[, VDJgene], i$seurat_clusters_patient_group))
  
  cor_tmp <- cor(tmp, method = correl)
  cor_tmp[is.na(cor_tmp)] <- 0
  
  pheatmap::pheatmap( cor_tmp, 
                      annotation_col = data.frame(cell_type = i$BcelltypeMain[match(sub("_.*", "", colnames(cor_tmp)), i$seurat_clusters)],
                        cluster = factor(sub("_.*", "", colnames(cor_tmp))),
                                                 patient_group = factor(sub(".*_", "", colnames(cor_tmp))),
                                                  row.names = colnames(cor_tmp)),
                      breaks = seq(0, 1, length.out = 100),
                      filename = paste0("../results/VDJ_Cor_", VDJgene, "_", correl, "_per_cluster_PatientGroup.pdf"),
                      fontsize = 8, 
                      fontsize_row = 5, 
                      fontsize_col = 5,
                      main = paste(VDJgene, correl, "correlation"),
                      color = colorRampPalette(c("lightblue", "firebrick"))(99))
  
}

VDJ_cor(gex, "spearman", "v_call")
VDJ_cor(gex, "pearson", "v_call")

VDJ_cor(gex, "spearman", "d_call")
VDJ_cor(gex, "pearson", "d_call")

VDJ_cor(gex, "spearman", "j_call")
VDJ_cor(gex, "pearson", "j_call")

VDJ_cor(gex, "spearman", "c_call")
VDJ_cor(gex, "pearson", "c_call")


```


#### top VDJ genes

```{r}

# get most used VDJ genes per patient subgroup

get_topVDJ <- function(i, #seurat object
                       VDJgene = "tmp" #VDJ gene to plot
  ) {
  
  vh_col <- unlist(lapply(strsplit(i@meta.data[, VDJgene], split = ","), function(x) {
  ifelse(length(x) > 1 , return(x[1]), return(x))})) #get only first allele if several exist
  
  vh <- as.data.frame.matrix(table(vh_col, i$patient_group))
  vh <- vh[order(rowSums(vh), decreasing = TRUE), ]
  
  write.csv(vh, paste0("VDJ_", VDJgene, "_genesPerPatientGroup.csv")) #export counts per patient group
  
  vh <-  as.data.frame(t(t(vh) / colSums(vh))) #get percentages of repertoir
  vh$gene <- rownames(vh)
  
  p <- as_tibble(vh[1:50, ]) %>% pivot_longer(
       cols = 1:4,
       names_to = "patient_group",
       names_prefix = "",
       values_to = "count",
       values_drop_na = TRUE) %>%  
          ggplot(aes(x = reorder(gene, -count), y = count, fill = patient_group)) +
              geom_col(position = "dodge") +
              theme_classic() +
              xlab(VDJgene) +
              ylab("Percent of repertoire") +
              scale_y_continuous(labels = percent) +
              ggtitle(paste0(VDJgene, " genes, top 50")) +
              theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
              facet_grid(rows = vars(patient_group))

    ggsave2(paste0("../results/VDJ_geneUsage_", VDJgene, "_geneUsage_perPatientGroup.png"), 
          p,
          width = 30, height = 20, unit = "cm")
  
   }

get_topVDJ(gex, "v_call")
get_topVDJ(gex, "d_call")
get_topVDJ(gex, "j_call")
get_topVDJ(gex, "c_call")


```


#### VDJ gene by alakazam countGenes()

```{r}

get_VDJperc <- function(i, #seurat object with imm cantation out in meta data
                       VDJgene = "tmp" #VDJ gene to plot
  ) {
  
  tmp <- i@meta.data
   
  #1 gene count per patient group per seurat cluster
  vdj_gene_1 <- countGenes(tmp, 
                               gene = VDJgene, 
                               groups = c("patient_group", "seurat_clusters"), 
                               mode = "gene")
  
  p <- ggplot(vdj_gene_1, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene Usage per cluster")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(seurat_clusters))

  ggsave2(paste0("../results/VDJ_countGenes_",VDJgene ,"_perPatientGroup_perCluster.png"), 
          p,
          width = 15, height = 50, unit = "cm")
  
  
  #2 gene family count per patient group per seurat cluster
  vdj_family_1 <- countGenes(tmp, 
                                 gene = VDJgene, 
                                 groups = c("patient_group", "seurat_clusters"), 
                                 mode = "family")
  
   p <- ggplot(vdj_family_1, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene family Usage per cluster")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(seurat_clusters))

  ggsave2(paste0("../results/VDJ_countGenes_",VDJgene ,"_perPatientGroup_perCluster.png"),
          p,
          width = 15, height = 50, unit = "cm")

  
  
  #tmp <- tmp[tmp$BcelltypeMain == "Memory" | tmp$BcelltypeMain == "Naive", ]

  #3 gene count per patient group per cell type
  vdj_gene_2 <- countGenes(tmp, 
                               gene = VDJgene, 
                               groups = c("patient_group", "BcelltypeMain"), 
                               mode = "gene")
  
  p <- ggplot(vdj_gene_2, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene Usage per cell type")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(BcelltypeMain))

  ggsave2(paste0("../results/VDJ_countGenes_", VDJgene ,"_perPatientGroup_perCelltype.png"),
          p,
          width = 20, height = 25, unit = "cm")
  
  #4 gene family count per patient group per cell type
  vdj_family_2 <- countGenes(tmp, 
                                 gene = VDJgene, 
                                 groups = c("patient_group", "BcelltypeMain"), 
                                 mode = "family")
  
  p <- ggplot(vdj_family_2, aes(x = gene, y = seq_freq, fill = patient_group)) +
      theme_classic() +
      ggtitle(paste0(VDJgene, " Gene family Usage per cell type")) +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Percent of repertoire") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_col(position = "dodge") + 
      facet_grid(rows = vars(BcelltypeMain))

  ggsave2(paste0("../results/VDJ_countGenes_", VDJgene ,"_perPatientGroup_perCelltype.png"),
          p,
          width = 20, height = 25, unit = "cm")
  
}

get_VDJperc(gex, "v_call")
get_VDJperc(gex, "d_call")
get_VDJperc(gex, "j_call")
get_VDJperc(gex, "c_call")


```


### top clonotypes

```{r}
# sort(table(gex$CTaa), decreasing = T)[1:5]
# gex$CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF <- factor(gex$CTaa == "CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF")
# 
# for (i in 1:15){
# print(sort(table(gex$CTaa), decreasing = T)[i])
# print(table(factor(gex$CTaa == names(sort(table(gex$CTaa), decreasing = T)[i])), gex$orig.ident))
# }
# 
# table(gex$CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF, gex$orig.ident)
# 
# p1 <- DimPlot(gex, reduction = "umap_2", 
#         group.by = "CARVAESRYYDFWHGFQSSYYFDYW_CNSYTSSRSWVF", 
#         cols = c("gray80", "firebrick", "gray80"), order=TRUE, pt.size = 0.5, split.by = "orig.ident", ncol = 2)
# 
# ggsave2("../results/clono_test.png", p1, height = 30, width = 10, units = "cm")
# gex$seurat_clusters_patient_group <-paste0(gex$seurat_clusters, "_", gex$patient_group)
# 
# 
# unlist(lapply(strsplit(gex@meta.data$v_call, split = ","), function(x) {
#   ifelse(length(x) > 1 , return(x[1]), return(x))
#   }
# ))


```


### Quantify clonotypes scRepertoire

```{r}

quantContig(
  combined_B,
  cloneCall = "gene+nt",
  chain = "both",
  scale = TRUE,
  group = NULL,
  exportTable = FALSE
)


```


### clonotype analysis

```{r}


clones <- countClones(gex@meta.data, groups = c("patient_group", "BcelltypeMain"), clone = "clone_id")

arrange(clones, patient_group, seq_freq) %>% 
ggplot(aes(x = clone_id, y = seq_freq)) +
      theme_bw() +
      geom_col(aes(color = patient_group))
  
summary(clones)
head(sort(table(clones$clone_id), decreasing = TRUE))
head(clones)

ggplot(clones, aes(x = clone_id, y = seq_freq, fill = patient_group)) +
      theme_bw() +
      ggtitle("Vh Family Usage per group") +
      #theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("Clonal abundance") +
      xlab("") +
      #scale_y_continuous(labels = percent) +
      geom_col()

geom_line(aes(color = patient_group), size = 2, alpha = 0.8)

#abundance estimation
cu <- estimateAbundance(outfile, group = "patient_group", 
                        ci = 0.95, nboot = 100, clone = "clone_id")
#Error: vector memory exhausted (limit reached?)
#nboot = 100 works locally, nboot = 200 does not
plot(cu, legend_title = "patient_group")

png("../results/VDJ_clonalAbundances_perPatientGroup.png", 
    width = 20, height = 20, units = "cm", res = 200)
plot(cu, 
     #colors = sample_colors, 
     main_title = "clonal abundances", 
     legend_title = "Patient_group")
dev.off()


#diversity curve
#per patient group
sample_curve <- alphaDiversity(outfile, group = "patient_group", clone = "clone_id",
                               min_q = 0, max_q = 4, step_q = 0.1,
                               ci = 0.95, nboot = 200)
png("../results/VDJ_diversityCurve_perPatientGroup.png", 
    width = 20, height = 20, units = "cm", res = 200)
plot(sample_curve, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "Patient_group")
dev.off()


#per c_call
#remove c_call classes with too few cells
tmp <- outfile[!(outfile$c_call == "IGHE" | 
                   outfile$c_call == "IGLC1" |
                   outfile$c_call == "IGLC2"), ]

isotype_curve <- alphaDiversity(tmp, group = "c_call", clone = "clone_id",
                                min_q = 0, max_q = 4, step_q = 0.1,
                                ci = 0.95, nboot = 200)

png("../results/VDJ_diversityCurve_perPatientGroup.png", 
    width = 20, height = 20, units = "cm", res = 200)
plot(isotype_curve, 
     #colors = sample_colors, 
     main_title = "clonal diversity", 
     legend_title = "c_call")
dev.off()



```


### plot CDR3 properties

```{r}


```


### plot mutation status

```{r}

p5 <- ggplot(gex@meta.data, aes(x = seurat_clusters, y = mu_freq_total)) +
      theme_classic() +
      ggtitle("Vh mutations") +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("perc mutated bp VH") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_boxplot(aes(color = patient_group), size = 0.4, alpha = 0.8)

ggsave2(paste0("../results/VDJ_Vh_mutationstatus_byCluster.png"), 
          p5,
          width = 30, height = 20, unit = "cm")

p6 <- ggplot(gex@meta.data, aes(x = orig.ident, y = mu_freq_total)) +
      theme_classic() +
      ggtitle("Vh mutations") +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("perc mutated bp VH") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_boxplot(aes(color = patient_group), size = 0.4, alpha = 0.8)

ggsave2(paste0("../results/VDJ_Vh_mutationstatus_bySample.png"), 
          p6,
          width = 30, height = 20, unit = "cm")

cluster_mut <- as.data.frame.matrix(table(gex$seurat_clusters, gex$mu_load))
cluster_mut$percmut <- 1-cluster_mut$unmutated/rowSums(cluster_mut)
cluster_mut$orig.ident <- factor(rownames(cluster_mut), levels = rownames(cluster_mut))

p7 <- ggplot(cluster_mut, aes(x = orig.ident, y = percmut)) +
      theme_classic() +
      ggtitle("perc cells with mutated VH per cluster") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("perc mutated cells") +
      xlab("") +
      scale_y_continuous(labels = percent) +
      geom_bar(stat='identity')

ggsave2(paste0("../results/VDJ_Vh_percMutCells_byCluster.png"), 
          p7,
          width = 30, height = 20, unit = "cm")

p8 <- DimPlot(gex, reduction = "umap_2", 
                  pt.size = .01, 
                  group.by = "mu_load", 
                  raster=FALSE) + 
          ggtitle("Mutation load") +
          ylab("UMAP 2") +
          xlab("UMAP 1")

ggsave2(paste0("../results/VDJ_Vh_UMAP_mutLoad.png"), 
          p8,
          width = 25, height = 20, unit = "cm")

head(gex@meta.data)


p9 <- ggplot(gex@meta.data, aes(x = BcelltypeMain, y = mu_freq_total)) +
      theme_classic() +
      ggtitle("Vh mutations") +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("perc mutated bp VH") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      #geom_jitter(size = 0.2, alpha = 0.3) +
      geom_boxplot(aes(color = patient_group), size = 0.3, alpha = 0.8)

ggsave2(paste0("../results/VDJ_Vh_mutationstatus_byCelltypePerPatGroup.png"), 
          p9,
          width = 30, height = 20, unit = "cm")

#
p10 <- ggplot(gex@meta.data, aes(x = patient_group, y = mu_freq_total)) +
      theme_classic() +
      ggtitle("Vh mutations") +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("perc mutated bp VH") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      #geom_jitter(size = 0.2, alpha = 0.3) +
      facet_grid(cols = vars(BcelltypeMain)) +
      geom_boxplot(aes(color = orig.ident), size = 0.3, alpha = 0.8)

ggsave2(paste0("../results/VDJ_Vh_mutationstatus_byCelltypePerSample.png"), 
          p10,
          width = 40, height = 20, unit = "cm")

p11 <- ggplot(gex@meta.data[gex$BcelltypeMain == "Memory" | gex$BcelltypeMain == "Naive", ], aes(x = patient_group, y = mu_freq_total)) +
      theme_classic() +
      ggtitle("Vh mutations") +
      theme(axis.text.x=element_text(angle = 45, hjust = 1, vjust = 1)) +
      ylab("perc mutated bp VH") +
      xlab("") +
      scale_y_continuous(labels = percent)  +
      #geom_jitter(size = 0.2, alpha = 0.3) +
      facet_grid(cols = vars(BcelltypeMain)) +
      geom_boxplot(aes(color = orig.ident), size = 0.3, alpha = 0.8)

ggsave2(paste0("../results/VDJ_Vh_mutationstatus_byCelltypePerSample_MemNaive.png"), 
          p11,
          width = 30, height = 20, unit = "cm")


```

###  plot IgH class, mu_load and clonetype per cluster

```{r}

##---- IgH per cluster
# res <- table(list(gex$seurat_clusters, gex$c_gene_10x))
# res <- t(res)
# res <- res[,order(colSums(res),decreasing = T)]
# res2 <- t( t(res) / colSums(res))
# res3 <- res / rowSums(res)
# res4 <- t( t(res3) / colSums(res3) )
# mypal <- RColorBrewer::brewer.pal(nrow(res2),"Set3")
# 
# #plot proportion of cells per cluster
# png(file = paste0("../results/cluster_IgHclass_abundance.png"),
#     width = 20, height = 8, units = "cm", res = 200)
# mypar(1,1,mar = c(2,4,2,5), xpd=TRUE, cex.axis = 0.7)
# barplot(res2*100, las = 1, col = mypal, main = "IgH class", ylab = "Proportion of cells\n")
# legend("topright", legend = rownames(res2), bty = "n",fill = mypal, inset=c(-0.15,0))
# dev.off()


p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.01, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(9, "Set3"), "grey")) + 
              theme(legend.position = "none") 

p2 <- ggplot(gex@meta.data, aes(x = seurat_clusters, fill = c_gene_10x)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      facet_grid(rows = vars(patient_group)) + 
scale_fill_brewer(palette="Set3", na.value="grey")

ggsave2("../results/cluster_c_gene_10x_abundance_2.png", 
        plot_grid(p1, p2, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


p1 <- DimPlot(gex, 
              group.by = "c_gene_10x", 
              split.by = "patient_group", 
              reduction = "umap_2", ncol = 2, raster = FALSE)
ggsave2("../results/UMAP_B-cells_IgHclass.png", 
        p1, 
        width = 30, height = 20, unit = "cm")


##---- mutation load per cluster
# res <- table(list(gex$seurat_clusters, gex$mu_load))
# res <- t(res)
# res <- res[,order(colSums(res),decreasing = T)]
# res2 <- t(t(res)/colSums(res))
# res3 <- res / rowSums(res)
# res4 <- t( t(res3) / colSums(res3) )
# mypal <- RColorBrewer::brewer.pal(nrow(res4),"Set3")
# 
# #plot proportion of cells per cluster
# png(file = paste0("../results/cluster_muLoad_abundance.png"),
#     width = 20, height = 8, units = "cm", res = 200)
# mypar(1,1,mar = c(2,4,2,6), xpd=TRUE, cex.axis = 0.7)
# barplot(res2*100, las = 1, col = mypal, main = "Mutation load", ylab = "Proportion of cells\n")
# legend("topright", legend = rownames(res2), bty = "n",fill = mypal, inset=c(-0.18,0))
# dev.off()

p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.01, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(4, "Dark2"), "grey")) +
theme(legend.position = "none") 

p2 <- ggplot(gex@meta.data, aes(x = seurat_clusters, fill = mu_load)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      facet_grid(rows = vars(patient_group)) + 
scale_fill_brewer(palette = "Dark2", na.value = "grey")

ggsave2("../results/cluster_mu_load_abundance_2.png", 
        plot_grid(p1, p2, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


p1 <- DimPlot(gex, 
              group.by = "mu_load", 
              split.by = "patient_group", 
              reduction = "umap_2", ncol = 2, raster = FALSE)  
              
ggsave2("../results/UMAP_B-cells_muLoad.png", 
        p1, 
        width = 30, height = 20, unit = "cm")



##---- cloneType per cluster

ggplot(gex@meta.data, aes(x = seurat_clusters, fill = c_call)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      facet_grid(rows = vars(patient_group))


res <- table(gex$seurat_clusters, gex$cloneType, gex$patient_group)

res <- t(res)
res <- res[,order(colSums(res),decreasing = T)]
res2 <- t(t(res)/colSums(res))
res3 <- res / rowSums(res)
res4 <- t( t(res3) / colSums(res3) )
mypal <- RColorBrewer::brewer.pal(nrow(res4),"Set3")

#plot proportion of cells per cluster
png(file = paste0("../results/cluster_cloneType_abundance.png"),
    width = 20, height = 8, units = "cm", res = 200)
mypar(1,1,mar = c(2,4,2,12), xpd=TRUE, cex.axis = 0.5)
barplot(res2*100, las = 1, col = mypal, main = "clonotype", ylab = "Proportion of cells\n")
legend("topright", legend = rownames(res2), bty = "n",fill = mypal, inset=c(-0.5,0))
dev.off()




p1 <- DimPlot(gex, 
              group.by = "cloneType", 
              split.by = "patient_group", 
              reduction = "umap_2", pt.size = 0.01, ncol = 2, 
              raster = TRUE, cols = c(brewer.pal(4, "Dark2"), "grey")) +
theme(legend.position = "none") 

# ggsave2("../results/UMAP_B-cells_cloneType.png", 
#         p1, 
#         width = 30, height = 20, unit = "cm")

p2 <- ggplot(gex@meta.data, aes(x = seurat_clusters, fill = cloneType)) +
      theme_classic() +
      geom_bar(position = "fill") + 
      ylab("proportion of cells") + 
      facet_grid(rows = vars(patient_group)) + 
scale_fill_brewer(palette="Dark2", na.value="grey")


ggsave2("../results/cluster_cloneType_abundance_2.png", 
        plot_grid(p1, p2, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


```


## Save rds GEX_BCELLS

```{r}

saveRDS(gex, file = "../results/GEX7_BVDJ.rds")
#gex <- readRDS("../results/GEX7_BVDJ.rds")

```


## Print sessionInfo()

```{r}

sessionInfo()

```



