---
output: html_document
editor_options: 
  chunk_output_type: console
---

# DE tests and Visualization {#DE-tests}
## Load packages

```{r}

source("./00_dependencies.R")

```


## Load data

```{r}

gex <- readRDS("../results/GEX7_BVDJ.rds")

```


## Check patient groups in Seurat object

```{r}

table(gex$patient_group, useNA = "always")
#Four non-overlapping groups
#1. SSAB - SSA+/SSB+
#2. SSA  - SSA+/SSB-
#3. DNEG - SSA-/SSB-
#4. CTRL - CTRL


```


#DE tests celltypesFine
## pseudo bulk DE test between patient subgroups for celltypesFine

```{r }

#https://jef.works/blog/2020/04/06/quickly-creating-pseudobulks/

DGE_list <- list()
for (j in sort(unique(gex$cluster_cellType_manual_fine))){

  message(j)
  
  #subset gex to only contain subtype j
  gex.tmp <- gex %>% 
    DietSeurat(dimreducs = "umap_2") %>% 
    tidyseurat::filter(orig.ident != "P024", 
                       cluster_cellType_manual_fine == j)

  #create a model matrix for pseudo bulk generation
  mm <- model.matrix(~ 0 + orig.ident, 
                     data = gex.tmp@meta.data); head(mm); colnames(mm)
  
  #sample x number of cells per condition
  gex.tmp$subgroups <- factor(as.character(gex.tmp$orig.ident))
  sample_size <- table(gex.tmp$subgroups)
  sample_size[sample_size > 100] <- 100
  
  DGE_cells <- lapply(names(sample_size), function(x){ 
    set.seed(1)
    sample(colnames(gex.tmp) [gex.tmp$subgroups == x] , 
           size = sample_size[x], 
           replace = FALSE)
    })
  
  DGE_cells <- unlist(DGE_cells)
  mm <- mm[DGE_cells, ]
  
  #clean up colnames
  colnames(mm) <- gsub("orig.ident", "", colnames(mm))
  
  # create pseudo bulk count matrix
  gex.mm <- gex.tmp@assays$RNA@counts[, DGE_cells] %*% mm; dim(gex.mm)
  gex.mm <- gex.mm[rowSums(gex.mm) != 0, ]; dim(gex.mm)
  
  cellcounts <- colSums(mm); length(cellcounts)
  
  y <- DGEList(gex.mm, 
               sample = gsub("_.*", "", colnames(gex.mm)), 
               group = gsub(".*_", "", colnames(gex.mm)),
               remove.zeros = TRUE); y
  
  y$samples$patient_group <- as.factor(unique(gex.tmp@meta.data[, c("orig.ident", "patient_group")])$patient_group[match(y$samples$samples, unique(gex.tmp@meta.data[, c("orig.ident", "patient_group")])$orig.ident)])
  
  keep <- apply(y$counts, 1, function(x){sum(x >= 5) >= 3}); table(keep)
  y <- y[keep, ]
  y <- calcNormFactors(y)
  
  design <- model.matrix(~ patient_group, data = y$samples); colnames(design)
  colnames(design) <- gsub("patient_group", "", colnames(design))
  
  y_glm <- estimateDisp(y, design)
  fit <- glmQLFit(y_glm, design); colnames(fit$coefficients)
  
  #conserved differences across patient subgroups
  qlm <- glmQLFTest(fit, coef = 2:4)
  de_g <- topTags(qlm, n = nrow(y_glm$counts))$table; head(de_g, 25)
  
  de_g$celltypeFine <- j
  de_g$gene <- rownames(de_g)
  
  DGE_list[[j]] <- de_g
  
  saveRDS(fit, paste0("../results/DGEfit_patientGroups_celltypeFine_", j,".rds"))
  write.csv(de_g, paste0("../results/DGE_patientGroups_celltypeFine_", j,".csv"), row.names = TRUE)

}

saveRDS(DGE_list, "../results/DGE_B-cells_celltypeFine_patientGroups_listAll.rds")
#DGE_list <- readRDS("../results/DGE_B-cells_celltypeFine_patientGroups_listAll.rds")


```


### filter rbind and export csv for all celltypeFine DE tests

```{r}

FDR_cutoff <- 0.25
DGE_list_out <- lapply(DGE_list, function(x){

  
  cat(unique(x$celltypeFine), "\n")
  
  #filter on FDR
  x <- x[x$FDR <= FDR_cutoff, ]
  
  cat(nrow(x), "\n")
  
  gex_tmp <- gex %>% 
    DietSeurat(dimreducs = "umap_2") %>% 
    tidyseurat::filter(cluster_cellType_manual_fine == unique(x$celltypeFine))
  
  #get "pct.exp" and "avg.exp" for filtering
  p <- DotPlot(gex_tmp, features = rownames(x), group.by = "patient_group")
  p <- p$data %>%  pivot_wider(names_from = "id", 
                          values_from = c("pct.exp", "avg.exp"), 
                          id_cols = "features.plot")
  x <- cbind(x, p)
  
  return(x)
  
})


DGE_list_out <- as.data.frame(do.call(rbind, DGE_list_out))

write.csv(DGE_list_out, paste0("../results/DGE_B-cells_celltypeFine_all_FDR", FDR_cutoff, ".csv"), row.names = FALSE)

```


### filter and plot dotplots for top expressed genes celltypeFine 

```{r}

#filter DE genes
FDR_cutoff <- 0.2
FC_cutoff <- 2
PCTexpr_cutoff <- 20

####---- subset and make plots
DGE_list_p <- lapply(DGE_list, function(x) {
  
  a <- x
  
  print(unique(a$celltypeFine))
  
  #filter on FDR
  a <- a[a$FDR <= FDR_cutoff, ]
  
  cat("after FDR: ",nrow(a), "\n")
  
  gex_tmp <- gex %>% 
    DietSeurat(dimreducs = "umap_2") %>% 
    tidyseurat::filter(cluster_cellType_manual_fine == unique(a$celltypeFine))
  
  
  #get "pct.exp" and "avg.exp" for filtering
  p <- DotPlot(gex_tmp, features = rownames(a), group.by = "patient_group")
  p <- p$data %>%  pivot_wider(names_from = "id", 
                          values_from = c("pct.exp", "avg.exp"), 
                          id_cols = "features.plot")
  a <- cbind(a, p)
  
  #filter on fold change
  a <- a[apply(a[, 1:3], 1, max) > log2(FC_cutoff) |
         apply(a[, 1:3], 1, min) < log2(1/FC_cutoff), ]
  
  cat("after FC: ",nrow(a), "\n")
  
  #filter on percent expressed
  a <- a[apply(a[, 11:13], 1, max) > PCTexpr_cutoff, ]
  
  cat("after %e: ",nrow(a), "\n")

  return(a)
  
   }
  )


####---- plot dotplots to list
ngenes <- 15

p_list <- list()
for (i in sort(unique(gex$cluster_cellType_manual_fine))[-c(7, 8, 12, 16)]){

  cat(i,"\n")
  
  a <- as_tibble(DGE_list_p[[i]]) %>% arrange(DGE_list_p[[i]]) %>% pull(gene)
  
  cat(length(a),"\n")
  cat(a,"\n")
  
  if(length(a) < ngenes) {
    p_features <- a
  } else {
    p_features <- a[1:ngenes]
  }
  
  cat(length(p_features),"\n")
  
  #subset gex
  DGE_DATA <- gex %>% 
  DietSeurat(dimreducs = "umap_2") %>% 
  tidyseurat::filter(orig.ident != "P024", 
                       cluster_cellType_manual_fine == i)

  p_list[[i]] <- DotPlot(DGE_DATA, 
        features = p_features, 
        group.by = "patient_group",
        dot.scale = 4) +
  theme_classic() +
  xlab("") + 
  ylab("") + 
  ggtitle(i) +
  coord_flip() + 
  theme(axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5,
                                   size = 8),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_line(colour = "grey", size = rel(0.5)))

}

ggsave2("../results/DGE_B-cells_celltypeFine_dotplots.png", 
        plot_grid(plotlist = p_list, ncol = 4, labels = NULL, align = "h"), 
        width = 28, height = 20, unit = "cm")


```

### GSEA celltypesFine

```{r}

#cluster genes before enrichR


p_list <- list()
for (i in sort(names(DGE_list))[-c(7, 8, 12)]){
  
  go_cluster <- enrichr(genes = DGE_list[[i]]$gene[1:50], 
                        databases = c("GO_Biological_Process_2021", 
                                      "WikiPathway_2021_Human",
                                      "KEGG_2021_Human",
                                      "MSigDB_Hallmark_2020"))
  
  go_cluster <- go_cluster$MSigDB_Hallmark_2020
  go_cluster <- go_cluster[order(go_cluster$P.value), ]
  go_cluster$logP <- -log10(go_cluster$P.value)
  
  p_list[[i]] <- ggplot(go_cluster[1:5,], aes(x = reorder(Term, logP), 
                                              y = logP)) +
    geom_col(position = "dodge") +
    theme_classic() + 
    coord_flip() +
    theme(axis.text.x = element_text(angle = 0, 
                                     hjust = 0.5, 
                                     vjust = 0.5,
                                     size = 12,
                                     color = "black"),
          axis.text.y = element_text(angle = 0, 
                                     hjust = 1, 
                                     vjust = 0.5,
                                     size = 12,
                                     color = "black"),
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0.2,0.2,0.2), "cm")) +
      ylab(expression('-log10('~italic(P)~')')) +
      xlab("") +
      ggtitle(paste0(i)) +
      geom_hline(aes(yintercept = 2), color = "black", linetype = "dashed")

  
}

ggsave2(paste0("../results/DGE_B-cells_GSE_celltypeFine.png"),
          plot_grid(plotlist = p_list, 
                    ncol = 4, 
                    labels = NULL, 
                    align = "vh" ) + bgcolor("white"),
          width = 45, height = 24, unit = "cm")


# DGE_list[[1]]$gene[1:50]
# enrichR::listEnrichrDbs()[, "libraryName"]
# go_cluster <- enrichr(genes = DGE_list[[1]]$gene[1:50], databases = c("GO_Biological_Process_2021",
#                                                                       "WikiPathway_2021_Human",
#                                                                       "KEGG_2021_Human",
#                                                                       "MSigDB_Hallmark_2020"))
# go_cluster <- go_cluster$MSigDB_Hallmark_2020
# go_cluster <- go_cluster[order(go_cluster$P.value),]
# 
# mypar(1,1,mar=c(2,10,2,2))
# barplot(-log10(go_cluster$P.value[5:1]),horiz=T,border=F,yaxs="i",
#            names.arg=go_cluster$Term[5:1],las=1,cex.names=0.8)
# abline(v = 2, lwd = 1, lty = 2, col ="gray20")



```


#DE tests celltypesMain
## pseudo bulk DE test between patient subgroups for celltypesMain

```{r }

#https://jef.works/blog/2020/04/06/quickly-creating-pseudobulks/

DGE_list <- list()
for (j in sort(unique(gex$cluster_cellType_manual_main))){

  message(j)
  
  #subset gex to only contain subtype j
  gex.tmp <- gex %>% 
    DietSeurat(dimreducs = "umap_2") %>% 
    tidyseurat::filter(orig.ident != "P024", 
                       cluster_cellType_manual_main == j)

  #create a model matrix for pseudo bulk generation
  mm <- model.matrix(~ 0 + orig.ident, 
                     data = gex.tmp@meta.data); head(mm); colnames(mm)
  
  #sample x number of cells per condition
  gex.tmp$subgroups <- factor(as.character(gex.tmp$orig.ident))
  sample_size <- table(gex.tmp$subgroups)
  sample_size[sample_size > 100] <- 100
  
  DGE_cells <- lapply(names(sample_size), function(x){ 
    set.seed(1)
    sample(colnames(gex.tmp) [gex.tmp$subgroups == x] , 
           size = sample_size[x], 
           replace = FALSE)
    })
  
  DGE_cells <- unlist(DGE_cells)
  mm <- mm[DGE_cells, ]
  
  #clean up colnames
  colnames(mm) <- gsub("orig.ident", "", colnames(mm))
  
  # create pseudo bulk count matrix
  gex.mm <- gex.tmp@assays$RNA@counts[, DGE_cells] %*% mm; dim(gex.mm)
  gex.mm <- gex.mm[rowSums(gex.mm) != 0, ]; dim(gex.mm)
  
  cellcounts <- colSums(mm); length(cellcounts)
  
  y <- DGEList(gex.mm, 
               sample = gsub("_.*", "", colnames(gex.mm)), 
               group = gsub(".*_", "", colnames(gex.mm)),
               remove.zeros = TRUE); y
  
  y$samples$patient_group <- as.factor(unique(gex.tmp@meta.data[, c("orig.ident", "patient_group")])$patient_group[match(y$samples$samples, unique(gex.tmp@meta.data[, c("orig.ident", "patient_group")])$orig.ident)])
  
  keep <- apply(y$counts, 1, function(x){sum(x >= 5) >= 3}); table(keep)
  y <- y[keep, ]
  y <- calcNormFactors(y)
  
  design <- model.matrix(~ patient_group, data = y$samples); colnames(design)
  colnames(design) <- gsub("patient_group", "", colnames(design))
  
  y_glm <- estimateDisp(y, design)
  fit <- glmQLFit(y_glm, design); colnames(fit$coefficients)
  
  #conserved differences across patient subgroups
  qlm <- glmQLFTest(fit, coef = 2:4)
  de_g <- topTags(qlm, n = nrow(y_glm$counts))$table; head(de_g, 25)
  
  de_g$celltypeMain <- j
  de_g$gene <- rownames(de_g)
  
  DGE_list[[j]] <- de_g
  
  saveRDS(fit, paste0("../results/DGEfit_patientGroups_celltypeMain_", j,".rds"))
  write.csv(de_g, paste0("../results/DGE_patientGroups_celltypeMain_", j,".csv"), row.names = TRUE)

}

saveRDS(DGE_list, "../results/DGE_B-cells_celltypeMain_patientGroups_listAll.rds")
#DGE_list <- readRDS("../results/DGE_B-cells_celltypeMain_patientGroups_listAll.rds")


```


### filter rbind and export csv for all celltypesMain DE tests

```{r}

FDR_cutoff <- 0.25
DGE_list_out <- lapply(DGE_list, function(x){

  
  cat(unique(x$celltypeMain), "\n")
  
  #filter on FDR
  x <- x[x$FDR <= FDR_cutoff, ]
  
  cat(nrow(x), "\n")
  
  #subset GEX to only include cell type
  gex_tmp <- gex %>% 
    DietSeurat(dimreducs = "umap_2") %>% 
    tidyseurat::filter(cluster_cellType_manual_main == unique(x$celltypeMain))
  
  
  #get "pct.exp" and "avg.exp" for filtering
  p <- DotPlot(gex_tmp, features = rownames(x), group.by = "patient_group")
  p <- p$data %>%  pivot_wider(names_from = "id", 
                          values_from = c("pct.exp", "avg.exp"), 
                          id_cols = "features.plot")
  x <- cbind(x, p)
  
  return(x)
  
})

DGE_list_out <- as.data.frame(do.call(rbind, DGE_list_out))

write.csv(DGE_list_out, paste0("../results/DGE_B-cells_celltypeMain_all_FDR", FDR_cutoff, ".csv"), row.names = FALSE)

```


### filter and plot dotplots for top expressed genes for celltypesMain

```{r}

#filter DE genes
FDR_cutoff <- 0.2
FC_cutoff <- 2
PCTexpr_cutoff <- 20

####---- subset and make plots
DGE_list_p <- lapply(DGE_list, function(x) {
  
  a <- x
  
  print(unique(a$celltypeMain))
  
  #filter on FDR
  a <- a[a$FDR <= FDR_cutoff, ]
  
  cat("after FDR: ",nrow(a), "\n")
  
  #subset GEX to only include cell type
  gex_tmp <- gex %>% 
    DietSeurat(dimreducs = "umap_2") %>% 
    tidyseurat::filter(cluster_cellType_manual_main == unique(a$celltypeMain))
  
  cat("nrow GEX: ", nrow(gex_tmp@meta.data), "\n")
  
  #get "pct.exp" and "avg.exp" for filtering
  p <- DotPlot(gex_tmp, features = rownames(a), group.by = "patient_group")
  p <- p$data %>%  pivot_wider(names_from = "id", 
                               values_from = c("pct.exp", "avg.exp"), 
                               id_cols = "features.plot")
  a <- cbind(a, p)
  
  #filter on fold change
  a <- a[apply(a[, 1:3], 1, max) > log2(FC_cutoff) |
         apply(a[, 1:3], 1, min) < log2(1/FC_cutoff), ]
  
  cat("after FC: ",nrow(a), "\n")
  
  #filter on percent expressed
  a <- a[apply(a[, 11:13], 1, max) > PCTexpr_cutoff, ]
  
  cat("after %e: ",nrow(a), "\n")

  return(a)
  
   }
  )


####---- plot dotplots to list
ngenes <- 25

p_list <- list()
for (i in sort(unique(gex$cluster_cellType_manual_main))[-c(3, 4, 6)]){

  cat(i,"\n")
  
  a <- as_tibble(DGE_list_p[[i]]) %>% arrange(PValue) %>% pull(gene)
  
  cat(length(a),"\n")
  cat(a,"\n")
  
  if(length(a) < ngenes) {
    p_features <- a
  } else {
    p_features <- a[1:ngenes]
  }
  
  cat(length(p_features),"\n")
  
  #subset gex
  DGE_DATA <- gex %>% 
  DietSeurat(dimreducs = "umap_2") %>% 
  tidyseurat::filter(orig.ident != "P024", 
                       cluster_cellType_manual_main == i)

  p_list[[i]] <- DotPlot(DGE_DATA, 
        features = p_features, 
        group.by = "patient_group",
        dot.scale = 4) +
  theme_classic() +
  xlab("") + 
  ylab("") + 
  ggtitle(i) +
  coord_flip() + 
  theme(axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5,
                                   size = 8),
        legend.position = "none",
        plot.title = element_text(hjust = 0.5,
                                  size = 10),
        panel.grid.major = element_line(colour = "grey", size = rel(0.5)))

}

ggsave2("../results/DGE_B-cells_celltypeMain_dotplots_2.png", 
        plot_grid(plotlist = p_list, ncol = 4, labels = NULL, align = "h"), 
        width = 25, height = 10, unit = "cm")


```


### GSEA celltypesMain

```{r}

#cluster genes before enrichR


p_list <- list()
for (i in sort(names(DGE_list))[-c(3, 4, 6)]){
  
  go_cluster <- enrichr(genes = DGE_list[[i]]$gene[1:50], 
                        databases = c("GO_Biological_Process_2021", 
                                      "WikiPathway_2021_Human",
                                      "KEGG_2021_Human",
                                      "MSigDB_Hallmark_2020"))
  
  go_cluster <- go_cluster$MSigDB_Hallmark_2020
  go_cluster <- go_cluster[order(go_cluster$P.value), ]
  go_cluster$logP <- -log10(go_cluster$P.value)
  
  p_list[[i]] <- ggplot(go_cluster[1:6,], aes(x = reorder(Term, logP), 
                                              y = logP)) +
    geom_col(position = "dodge") +
    theme_classic() + 
    coord_flip() +
    theme(axis.text.x = element_text(angle = 0, 
                                     hjust = 0.5, 
                                     vjust = 0.5,
                                     size = 12,
                                     color = "black"),
          axis.text.y = element_text(angle = 0, 
                                     hjust = 1, 
                                     vjust = 0.5,
                                     size = 12,
                                     color = "black"),
            strip.background = element_blank(), 
            strip.text.y = element_blank(),
            legend.position = "none",
            plot.margin = unit(c(0.2,0.2,0.2,0.2), "cm")) +
      ylab(expression('-log10('~italic(P)~')')) +
      xlab("") +
      ggtitle(paste0(i)) +
      geom_hline(aes(yintercept = 2), color = "black", linetype = "dashed")

  
}

ggsave2(paste0("../results/DGE_B-cells_GSE_celltypeMain.png"),
          plot_grid(plotlist = p_list, 
                    ncol = 4, 
                    labels = NULL, 
                    align = "vh" ) + bgcolor("white"),
          width = 40, height = 8, unit = "cm")


# DGE_list[[1]]$gene[1:50]
# enrichR::listEnrichrDbs()[, "libraryName"]
# go_cluster <- enrichr(genes = DGE_list[[1]]$gene[1:50], databases = c("GO_Biological_Process_2021",
#                                                                       "WikiPathway_2021_Human",
#                                                                       "KEGG_2021_Human",
#                                                                       "MSigDB_Hallmark_2020"))
# go_cluster <- go_cluster$MSigDB_Hallmark_2020
# go_cluster <- go_cluster[order(go_cluster$P.value),]
# 
# mypar(1,1,mar=c(2,10,2,2))
# barplot(-log10(go_cluster$P.value[5:1]),horiz=T,border=F,yaxs="i",
#            names.arg=go_cluster$Term[5:1],las=1,cex.names=0.8)
# abline(v = 2, lwd = 1, lty = 2, col ="gray20")



```


### violin plot for select DE features

```{r}
tmp <- Seurat::Read10X_h5("/Users/gusarv/Downloads/TS_Trachea.h5ad")

remotes::install_github("theislab/zellkonverter")
install.packages("zellkonverter")
library(zellkonverter)

DE_features_ifn <- c("EIF2AK2", "IFI44L", "IFITM1", "STAT1")

DE_features_2 <- c("LTB", "MX1", "CXCR4", "IL4R")

gex_p <- gex %>% 
  DietSeurat(dimreducs = "umap_2") %>% 
  tidyseurat::filter(cluster_cellType_manual_main != "Plasma_cell",
                     cluster_cellType_manual_main != "Memory_Platlet",
                     cluster_cellType_manual_main != "Memory_stressed")

p2 <- VlnPlot(gex %>% 
  DietSeurat(dimreducs = "umap_2") %>% 
  tidyseurat::filter(cluster_cellType_manual_main != "Plasma_cell",
                     cluster_cellType_manual_main != "Memory_Platlet",
                     cluster_cellType_manual_main != "Memory_stressed"), 
              features = DE_features_ifn, 
              group.by = "cluster_cellType_manual_main",
              split.by = "patient_group",
              #split.plot = TRUE,
              slot = "data", 
              log = FALSE, 
              ncol = 4)
ggsave2("../results/VlnPlot_markers_b_features_3.png", 
        p2, 
        width = 20, height = 10, unit = "cm")



```


### dotplots for top 50 differentially expressed genes celltypeFine 

```{r}

# for (i in unique(gex$cluster_cellType_manual_fine)){
# 
# message(i)
#   
# #get top expressed genes
# tmp <- DGE_list[[i]]
# tmp <- head(tmp, n = 50)
# 
# #subset seurat object
# DGE_DATA <- gex %>% 
#   DietSeurat(dimreducs = "umap_2") %>% 
#   tidyseurat::filter(orig.ident != "P024", 
#                        cluster_cellType_manual_fine == i)
# 
# DGE_DATA$patient_group <- factor(DGE_DATA$patient_group, levels = sort(unique(DGE_DATA$patient_group)))
# DGE_DATA$group <- paste0(DGE_DATA$cluster_cellType_manual_fine,
#                                 "__",
#                                 DGE_DATA$patient_group)
# DGE_DATA$group <- factor(DGE_DATA$group, levels = sort(unique(DGE_DATA$group)))
# 
# # detable$groups <- paste0(detable$cell_cluster, detable$cluster)
# # detable %>% group_by(groups)  %>% top_n(-60, p_val) %>% top_n(40, pct.diff) %>% top_n(20, log.pct.diff) -> top
# 
# ord <- factor(sapply(unique(as.character(tmp$gene)), 
#                      function(x) {getcluster(DGE_DATA, x, "group")}))
# ord
# 
# pdf(paste0("../results/DGE_B-cells_perPatientGroup_",i,".pdf"), 
#     width = 8, 
#     height = length(ord) / 5 + 2)
# mypar(1, 1, mar = c(10, 6, 1, 5))
# plot_dots(DGE_DATA, 
#           unique(as.character(tmp$gene))[order(as.numeric(as.character(ord)))], 
#           clustering = "group", 
#           show_grid = T,
#           main = paste0("top cluster markers; ", i),
#           cex.main = 1, 
#           font.main = 1, 
#           cex.col = 1, 
#           srt = 90, 
#           cex.row = 1.1)
# #abline(v = cumsum(c(table(sub("__.*", "", names(table(DGE_DATA$group)))))) + 0.5)
# dev.off()
# }
# 
# gex$celltype_fine_patient_group <- paste0(gex$cluster_cellType_manual_fine, gex$patient_group)
# tmp <- lapply(DGE_list, function(x){head(x, n = 10)})
# tmp <- do.call("rbind", tmp)
  
# DotPlot(gex, features = tmp$gene, group.by = "celltype_fine_patient_group", split.by = "patient_group") + RotatedAxis()

```



## plot expression for GWAS genes PMID: 34112769 https://pubmed-ncbi-nlm-nih-gov.ezproxy.its.uu.se/34112769/

```{r }

gwas_genes <- read.csv("/Users/gusarv/Desktop/pSS_GWASgenes_PMID34112769.txt", header = FALSE)
gwas_genes <- unique(gwas_genes$V1)
gwas_genes <- gwas_genes[!(gwas_genes %in% c("C6orf10", "HCG27", "LINC02571", "MSH5-SAPCD1", "LINC02571"))]

p <- VlnPlot(gex, gwas_genes, group.by = "cluster_cellType_manual_main", split.by = "patient_group")

ggsave2("../results/VlnPlot_gwas_2.png", 
        p, 
        width = 40, height = 60, unit = "cm")

p <- VlnPlot(gex, "FCRL4", group.by = "cluster_cellType_manual_fine", split.by = "patient_group")
dev.off()

pheatmap(prop.table(as.matrix(as.data.frame.matrix(table(gex$patient_group, gex$cluster_size_airr))), 1 ))
prop.table(as.matrix(as.data.frame.matrix(table(gex$patient_group, gex$cluster_size_airr))), 1 )

```


## DE between VH mutation classes for memory B-cells

```{r }



```


## Save rds 

```{r}

#saveRDS(gex, file = "../results/xxx.rds")

```


## Print sessionInfo()

```{r}

sessionInfo()

```



