---
title: "Analysis_master_bianca"
output: html_document
editor_options: 
  chunk_output_type: console
chunk_output_type: console
---

```{r}

knitr::opts_chunk$set(echo = TRUE)
paste0("R script started at:", Sys.time())


```


# Load libraries and setup workspace

```{r}

#source("./00_dependencies.R")

####----- system
suppressMessages(library(future)) #for planning multiprocessess
suppressMessages(library(BiocParallel))
suppressMessages(library(pryr)) #check memory usage

####----- general
suppressMessages(library(stringr)) #for string modulations, find and replace etc.
suppressMessages(library(biomaRt))
suppressMessages(library(rtracklayer))
suppressMessages(library(dplyr))
suppressMessages(library(tidyr))
suppressMessages(library(future.apply))

####----- plot
suppressMessages(library(plotly))
suppressMessages(library(ggplot2))
suppressMessages(library(cowplot))
suppressMessages(library(rafalib))

# remotes::install_github("czarnewski/niceRplots")
#library(niceRplots)

source("../../niceRplots/R/add_fig_label.R")
source("../../niceRplots/R/helper_functions.R")
source("../../niceRplots/R/plotting_functions.R")
source("../../niceRplots/R/trajectory_plots.R")

####----- scRNAseq
suppressMessages(library(Seurat)) #read and process scRNAseq data
suppressMessages(library(scater))
suppressMessages(library(scDblFinder))
suppressMessages(library(harmony))
suppressMessages(library(Nebulosa))


plan("multiprocess", workers = 16) #parallellization for seurat using future package:
options(future.globals.maxSize = 100000 * 1024 ^ 2) #resolve memory related issues related to parallellization 100000 here refers to approx 100GB memory

setwd("/castor/project/proj_nobackup/sjs1/R/pss_bcells_scRNAseq/code")
getwd()

paste0("Libraries loaded at: ", Sys.time())


```


# Load data

```{r}

####-----  List samples
samples <- str_replace(str_sort(list.files("../data/counts_mx"),
                                numeric = TRUE),
                       "_filtered_feature_bc_matrix",
                       "")
print("Samples:")
samples

####----- Create a named vector with paths to samples
samples.dirs <- sapply(samples, function(i) {
  d10x <- file.path('../data/counts_mx',
                    paste0(i, '_filtered_feature_bc_matrix'))
  d10x
})
print("Sample directories:")
samples.dirs

####----- Read count matrices
gex_data <- Read10X(data.dir = samples.dirs,
                    unique.features = TRUE,
                    strip.suffix = TRUE)
str(gex_data)

####----- Create seurat object
gex <- CreateSeuratObject(gex_data, 
                          project = "pSS",
                          min.cells = 0,
                          min.features = 0,
                          names.field = 1)
table(gex@meta.data$orig.ident)
nrow(gex@meta.data)

####-----  rm tmp objects
rm(samples, samples.dirs, gex_data)

####-----  change orig.ident for P007 and P008, for which libraries were generated and sequenced twice
gex@meta.data$orig.ident <- gsub("P007[ab]", "P007", gex@meta.data$orig.ident)
gex@meta.data$orig.ident <- gsub("P008[ab]", "P008", gex@meta.data$orig.ident); table(gex@meta.data$orig.ident)

####-----  add pheno info
pheno <- read.table("../../pSS_pheno_210623.csv", sep = ";", header = TRUE); head(pheno)
colnames(pheno)[c(1, 3)] <- c("Sample_id", "orig.ident"); head(pheno); dim(pheno)
pheno$orig.ident <- gsub("P007[ab]", "P007", pheno$orig.ident)
pheno$orig.ident <- gsub("P008[ab]", "P008", pheno$orig.ident)
pheno <- unique(pheno[, c(1, 3, 5, 7:8, 11:12)]); head(pheno); dim(pheno)

meta.data.tmp <- gex@meta.data
meta.data.tmp$cell <- row.names(meta.data.tmp)
meta.data.tmp <- merge(meta.data.tmp, pheno, by = "orig.ident", all.x = TRUE)
row.names(meta.data.tmp) <- meta.data.tmp$cell
meta.data.tmp$cell <- NULL

gex <- AddMetaData(gex, meta.data.tmp); head(gex@meta.data); table(gex@meta.data$orig.ident)

paste0("Data loaded at:", Sys.time())


```


# save rds GEX1

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX1_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX1_xxxxxx.rds"); gex; mem_used()

Idents(object = gex) <- "orig.ident"
gex1.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex1.dnsample.500, file = paste0("../results/GEX1_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX1 size:", print(object.size(gex), units = "GB"))


```


# Gather gene annotations for GRCh38.98

```{r}

#read GRCh38.98 .gtf file used during alignment
gtf.GRCh38.98 <- readGFF("../../refdata-gex-GRCh38-2020-A.genes.gtf.gz")
#saveRDS(gtf.GRCh38.98, "../../refdata-gex-GRCh38-2020-A.genes.gtf.rds")
#annot <- readRDS("../../refdata-gex-GRCh38-2020-A.genes.gtf.rds")

annot <- unique(gtf.GRCh38.98[gtf.GRCh38.98$type == "gene", c(1, 9, 11, 12)])
colnames(annot) <- c("chromosome_name", "gene_id", "gene_biotype", "external_gene_name")
annot$chromosome_name <- as.character(annot$chromosome_name)
annot[!grepl("^chr[123456789XYMT]", annot[, "chromosome_name"]), "chromosome_name"] <- "other"
head(annot)

saveRDS(annot, "../../refdata-gex-GRCh38-2020-A.genes.gtf.annot.rds")
#annot <- readRDS("../../refdata-gex-GRCh38-2020-A.genes.gtf.annot.rds")


```


# Mark doublets

```{r}

#scDblFinder requires around 3 min for this data set
gex <- as.Seurat(scDblFinder(as.SingleCellExperiment(gex), samples="orig.ident", BPPARAM=MulticoreParam(16)))
gex@meta.data$ident <- NULL


```


# Qualculate QC and mark problematic cells per sample

```{r}

gex <- SplitObject(gex, split.by = "orig.ident")

gex <- lapply(gex, annot = annot, function(x, annot){
  
  total <- colSums(x) #Total number of cells
  
#Might need to rerun this as "INMT-FAM188B", "INHBB", "INHBA", "INHBA-AS1", "PRPS1L1" "TRPS1"   "PRPSAP2", "PRPSAP1", "PRPS2", "PRPS1"  were excluded
  
  x$percent_mito <- colSums(x[grepl("^MT-", rownames(x)), ]) / total * 100
  x$percent_ribo <- colSums(x[grepl("^RP[LS]", rownames(x)) |
                              grepl("^MRP[LS]", rownames(x)), ]) / total * 100
  x$percent_hb <- colSums(x[grepl("^HB[AB]", rownames(x)), ]) / total * 100
  
  gene_biotype <- annot[ match(rownames(x), annot$external_gene_name), "gene_biotype"]
  x$percent_protein_coding <- colSums(x[(gene_biotype == "protein_coding") & 
                                          (!is.na(gene_biotype)), ]) / total * 100

  
  x$count_th <- x$nCount_RNA > quantile(x$nCount_RNA, 0.98) 
  x$feature_th <- (x$nFeature_RNA > quantile(x$nFeature_RNA, 0.98)) | (x$nFeature_RNA < 200)
  x$mito_th <- x$percent_mito > 25
  x$ribo_th <- x$percent_ribo < 5
  x$hb_th <- x$percent_hb > 1
  
  return(x)
  
})

gex <- merge(gex[[1]] , gex[-1])

gex$cells_discard <- rowSums(gex@meta.data[, grepl("_th", colnames(gex@meta.data)) ]) > 0

paste0("QC calculated at:", Sys.time())


```


# save rds GEX2

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX2_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX2_xxxxxx.rds"); gex; mem_used()

Idents(object = gex) <- "orig.ident"
gex2.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex2.dnsample.500, file = paste0("../results/GEX2_dnsample.500_",format(Sys.time(), "%y%m%d"),".rds"))

paste0("GEX2 size:", print(object.size(gex), units = "GB"))


```


# Summarize and Plot QC

```{r}

discarded_stats <- as_tibble(gex@meta.data) %>%
  group_by(orig.ident) %>% 
  summarise(cells = length(orig.ident),
            count_discarded = sum(count_th),
            feature_discarded = sum(feature_th),
            mito_discarded = sum(mito_th),
            ribo_discarded = sum(ribo_th),
            hb_discarded = sum(hb_th),
            pred_doublets = sum(scDblFinder.class=="doublet"),
            total_discarded = sum(cells_discard))
write.csv(discarded_stats, paste0("../results/stats_discarded_",format(Sys.time(), "%y%m%d"),".csv"), 
          row.names = FALSE, 
          quote = FALSE)

p_discarded_stats <- discarded_stats %>% 
  gather(discarded, cellcount, cells:total_discarded) %>% 
  ggplot(aes(x = orig.ident, y = cellcount, fill = discarded)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) + 
  theme(legend.title = element_blank(),
        axis.text.x = element_text(face="bold", angle = 90, vjust = 0.5, hjust = 1),
        axis.text.y = element_text(face="bold")) +
  ggtitle(paste0("Number of cells total/number of discarded cells (", 
                 sum(discarded_stats$cells), "/", 
                 sum(discarded_stats$total_discarded), ")"))

ggsave2(paste0("../results/Discarded_cellcounts_", format(Sys.time(), "%y%m%d"), ".png"), 
          p_discarded_stats,
          width = 25, height = 20, unit = "cm")

my_pars <- c("nCount_RNA", "nFeature_RNA", "percent_mito", "percent_ribo", "percent_hb")

#violin and density plot
for (i in my_pars){
  
  print(i)
  
  p1 <- plotColData(as.SingleCellExperiment(gex), 
                    x = "orig.ident", 
                    y = i, 
                    colour_by = "cells_discard") + 
    theme(axis.text.x = element_text(angle = 90)) +
    ggtitle(i) +
    xlab("") + 
    ylab(i)
  
  p2 <- gex@meta.data %>% as_tibble() %>% 
    ggplot(aes_string(x = i)) + 
    geom_density(aes(color = orig.ident, fill= orig.ident), alpha = 0.1) + 
    theme_classic() +
    scale_x_log10() +
    NoLegend() +
    ggtitle(paste0(i, " per sample"))
  
  p3 <- gex@meta.data[!gex@meta.data$cells_discard,  ] %>% as_tibble() %>% 
    ggplot(aes_string(x = i)) + 
    geom_density(aes(color = orig.ident, fill= orig.ident), alpha = 0.1) + 
    theme_classic() +
    scale_x_log10() +
    NoLegend() +
    ggtitle(paste0(i, " per sample, filtered"))
  
  
  ggsave2(paste0("../results/QCplot_", format(Sys.time(), "%y%m%d"), "_", i, ".png"), 
          plot_grid(p1, p2, p3, ncol = 3, labels = "AUTO", align = "h"),
          width = 40, height = 12, unit = "cm")
  
  
}

#QC overview plot
# png(paste0("./plots/QCplot_", format(Sys.time(), "%y%m%d"), "_mark_discarded.png"),
#     width = 2000, height = 2000, units = "px")
# rafalib::mypar(mar=c(5,6,1,1))
# barlist(data = gex, genes = my_pars,
#                     cex.axis = .8,
#                     srt = 45,
#                     orderby = "nFeature_RNA",
#                     clustering = "orig.ident",
#                     col = ifelse(gex$cells_discard, "red", "grey") )
# dev.off()


```


# Filter cells

```{r}

#filter annotation file to genes that should be kept
annot_coding <- unique(annot[!(annot$chromosome_name %in% c("chrY", "other", "chrM")) & 
                       (annot$gene_biotype %in% c("protein_coding", 
                                    "IG_V_gene", "IG_D_gene", "IG_J_gene", "IG_C_gene",

                                    "TR_V_gene", "TR_D_gene", "TR_J_gene", "TR_C_gene")), c("external_gene_name", "gene_biotype", "chromosome_name")])
genes_keep <- rownames(gex) %in% annot_coding$external_gene_name; table(genes_keep)

gex <- gex[genes_keep, !gex$cells_discard]; gex


```


# save rds GEX3

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX3_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX3_xxxxxx.rds")

Idents(object = gex) <- "orig.ident"
gex3.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex3.dnsample.500, file = paste0("../results/GEX3_dnsample.500_",format(Sys.time(), "%y%m%d"),".rds"))

paste0("GEX3 size:", print(object.size(gex), units = "GB"))


```


# Normalize and get variable features
 
```{r}

gex <- NormalizeData(gex, scale.factor = 10000)
gex <- FindVariableFeatures(gex, selection.method = "vst", nfeatures = 4000)

p1 <- LabelPoints(plot = VariableFeaturePlot(gex), 
                  points = head(VariableFeatures(gex), 80), 
                  repel = TRUE,
                  xnudge = 0,
                  ynudge = 0,
                  max.overlaps = Inf) + 
        theme_classic() +
        ggtitle(paste0(length(VariableFeatures(gex)), " Variable features - allCells"))

ggsave2(paste0("../results/VariableFeatures_allCells_", length(VariableFeatures(gex)), "_", format(Sys.time(), "%y%m%d"), ".png"), 
        p1, width = 25, height = 20, unit = "cm")


```


# Subset VariableFeatures(gex) to only include biotype == "protein_coding"

```{r}

VariableFeatures(gex) <- VariableFeatures(gex)[VariableFeatures(gex) %in% 
                                                 annot$external_gene_name[annot$gene_biotype == "protein_coding"]]

p1 <- LabelPoints(plot = VariableFeaturePlot(gex), 
                  points = head(VariableFeatures(gex), 80), 
                  repel = TRUE,
                  xnudge = 0,
                  ynudge = 0,
                  max.overlaps = Inf) + 
        theme_classic() +
        ggtitle(paste0(length(VariableFeatures(gex)), " Variable features - allCells (protein_coding)"))

ggsave2(paste0("../results/VariableFeatures_allCells_", length(VariableFeatures(gex)), "_protein_coding_", format(Sys.time(), "%y%m%d"), ".png"), 
        p1, width = 25, height = 20, unit = "cm")

                       
```


# Scale data

```{r}

gex <- ScaleData(gex, 
                 features = VariableFeatures(gex),
                 vars.to.regress = c("nFeature_RNA", "percent_mito", "percent_ribo")); mem_used() #15.8 GB

paste0("NormalizeData, FindVariableFeatures and ScaleData done at:", Sys.time())


```


# save rds GEX4

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX4_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX4_xxxxxx.rds")

Idents(object = gex) <- "orig.ident"
gex4.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex4.dnsample.500, file = paste0("../results/GEX4_dnsample.500_",format(Sys.time(), "%y%m%d"),".rds"))

paste0("GEX4 size:", print(object.size(gex), units = "GB"))


```


# PCA

```{r}

#RunPCA() takes around 2h for this data set
gex <- RunPCA(gex,
              assay = "RNA",
              features = VariableFeatures(gex),
              npcs = 100,
              reduction.name = "pca_1",
              verbose = TRUE)

gex@assays$RNA@scale.data <- matrix(0); gc()

paste0("PCA1 done at:", Sys.time())


```


# save rds GEX5

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX5_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX5_xxxxxx.rds")

Idents(object = gex) <- "orig.ident"
gex5.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex5.dnsample.500, file = paste0("../results/GEX5_dnsample.500_",format(Sys.time(), "%y%m%d"),".rds"))

paste0("GEX5 size:", print(object.size(gex), units = "GB"))


```


# Run Harmony

```{r}

gex <- RunHarmony(gex, 
                  group.by.vars = "orig.ident",
                  reduction = "pca_1",
                  reduction.save = "harmony_1")

paste0("Harmony1 done at:", Sys.time())

```


# plot PCA and harmony

```{r}

p.p <- DimPlot(object = gex, reduction = "pca_1", pt.size = .1, group.by = "orig.ident", raster=FALSE) + 
            NoLegend() + 
            ggtitle("PCA, by Sample")

p.h <- DimPlot(object = gex, reduction = "harmony_1", pt.size = .1, group.by = "orig.ident", raster=FALSE) + 
                NoLegend() + 
                ggtitle("Harmony, by Sample")

ggsave2(paste0("../results/PCA_harmony_allCells_", format(Sys.time(), "%y%m%d"),".png"), 
        plot_grid(p.p, p.h, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


```


# save rds GEX6

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX6_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX6_210604.rds")

Idents(object = gex) <- "orig.ident"
gex6.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex6.dnsample.500, file = paste0("../results/GEX6_dnsample.500_",format(Sys.time(), "%y%m%d"),".rds"))

paste0("GEX6 size:", print(object.size(gex), units = "GB"))


```


# run UMAP

```{r}

gex <- RunUMAP(gex, dims = 1:50, #or 100?
                reduction = "harmony_1",
                metric = "correlation",
                reduction.name = "umap_1",
                min.dist = 0.4, #local cell separation
                spread = .5, #global cell separation
                n.neighbors = 30, #30 on bianca!
                repulsion.strength = 0.4, #~2x min.dist, repulsion to cells fr annat cluster, global separation
                negative.sample.rate = 50, #ant ggr n.neighbors celler, global distance
                n.epochs = 100,
                n.components = 2)

paste0("UMAP1done at:", Sys.time())


```


# save rds GEX7

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX7_",format(Sys.time(), "%y%m%d"),".rds"))
#gex <- readRDS("../results/GEX7_xxxxxx.rds")

Idents(object = gex) <- "orig.ident"
gex7.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex7.dnsample.500, file = paste0("../results/GEX7_dnsample.500_",format(Sys.time(), "%y%m%d"),".rds"))

paste0("GEX7 size:", print(object.size(gex), units = "GB"))


```


# clustering

```{r}

gex <- FindNeighbors(gex, 
                     reduction = "harmony_1", 
                     dims = 1:50,
                     k.param = 30, #~30 on bianca!
                     verbose = TRUE)
#res <- 0.15 #RNA_snn res local
#res <- 0.6 #RNA_nn res local (knn, similar graph generated during UMAP)
#res <- 0.05 #Bianca RNA_snn resolution
res <- 0.4 #Bianca RNA_nn resolution

gex <- FindClusters(gex, resolution = res, verbose = TRUE, graph.name = "RNA_nn")

table(gex@meta.data$seurat_clusters)

paste0("Clustering1 done at:", Sys.time())


```


# save rds GEX8

```{r include = FALSE}

Idents(object = gex) <- "orig.ident"
saveRDS(gex, file = paste0("../results/GEX8_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX8_210624.rds")

gex8.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex8.dnsample.500, file = paste0("../results/GEX8_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX8 size:", print(object.size(gex), units = "GB"))


```


# hclust

```{r}

cluster_means <- sapply(as.character(unique(gex$seurat_clusters)), function(x){
  rowMeans( gex@assays$RNA@data[, gex$seurat_clusters == x] )
} )

adj <- (1-cor(cluster_means))/2
h <- hclust( as.dist( adj ), method = "ward.D2")

# my_dist <- dist(cluster_means,method = "euclidean")
# h <- hclust(my_dist, method = "complete")
# pheatmap::pheatmap(adj)

png(paste0("../results/hclust_clusters_allCells_clustRes_", res, "_",format(Sys.time(), "%y%m%d"), ".png"), 
    width = 2000, 
    height = 2000, 
    units = "px")
plot(as.dendrogram(h))
dev.off()


```


# plot UMAP 2D

```{r}

p1.u.s <- DimPlot(gex, 
                  reduction = "umap_1", 
                  pt.size = .1, 
                  group.by = "orig.ident", 
                  raster=FALSE) + 
          NoLegend() + 
          ggtitle("UMAP, by Sample")

p1.u.c <- DimPlot(gex, 
                  reduction = "umap_1", 
                  pt.size = .1, 
                  group.by = "seurat_clusters", 
                  label = TRUE, 
                  repel = TRUE, 
                  raster=FALSE) + 
          NoLegend() + 
          ggtitle("UMAP, by Cluster")

ggsave2(paste0("../results/UMAP_allCells_clustRes", res, "_",format(Sys.time(), "%y%m%d"), ".png"), 
        plot_grid(p1.u.s, p1.u.c, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


```


# plot cell specific marker genes

```{r}

ig.features <- c("IGHA1", "IGHA2", "IGHG1", "IGHG2", "IGHG3", "IGHG4",  "IGHD", "IGHE", "IGHM" )
b.markers <- c("CD79A", "CD79B", "MS4A1", "CD19", "CD27", "IGHA1", "IGHD", "IGHM", "JCHAIN", "MME")
pbmc.markers <- c("MS4A1", "CD19", "CD27", "CD79A", "GNLY", "CD3E", "CD14", "LYZ", "CD8A")
my_pars <- c("nCount_RNA", "nFeature_RNA", "percent_mito", "percent_ribo", "percent_hb")

#PBMC markers
p.pbmc.f <- FeaturePlot(gex, features = pbmc.markers, reduction = "umap_1", raster=FALSE, order = TRUE)
ggsave2(paste0("../results/FeaturePlot_allCells_pbmcMarkers_", format(Sys.time(), "%y%m%d"), ".png"), 
        p.pbmc.f, 
        width=25, height=25, unit="cm")

#B markers
p.b.markers.f <- FeaturePlot(gex, features = b.markers, reduction = "umap_1", raster=FALSE, order = TRUE)
ggsave2(paste0("../results/FeaturePlot_allCells_BMarkers_", format(Sys.time(), "%y%m%d"), ".png"), 
        p.b.markers.f, 
        width=30, height=20, unit="cm")

#IG markers
p.ig.markers.f <- FeaturePlot(gex, features = ig.features, reduction = "umap_1", raster=FALSE, order = TRUE)
ggsave2(paste0("../results/FeaturePlot_allCells_IGMarkers_", format(Sys.time(), "%y%m%d"), ".png"), 
        p.ig.markers.f, 
        width=25, height=25, unit="cm")

#my_pars
p.myPars.f <- FeaturePlot(gex, features = my_pars , reduction = "umap_1", raster=FALSE, order = TRUE)
ggsave2(paste0("../results/FeaturePlot_allCells_myPars", format(Sys.time(), "%y%m%d"), ".png"), 
        p.myPars.f, 
        width=25, height=25, unit="cm")


```


# get cluster marker genes

```{r}

sample_size <- table(gex$seurat_clusters)
sample_size[sample_size > 100] <- 100
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  sample( colnames(gex)[ gex$seurat_clusters == x ], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)
DGE_temp<- gex[, DGE_cells]
DGE_temp <- SetIdent(DGE_temp, value = "seurat_clusters")
detable <- FindAllMarkers( DGE_temp, 
                           min.pct = 0.1, 
                           min.cells.feature = 0.1, 
                           max.cells.per.ident = 50 , 
                           only.pos = T , 
                           logfc.threshold = 0.1, 
                           assay = "RNA" )

detable$pct.diff <- detable$pct.1 - detable$pct.2
write.csv(detable, paste0("../results/DGE_allCells_seurat_clusters_",format(Sys.time(), "%y%m%d"),".csv"))

detable <- detable[ detable$p_val < 0.05,  ]
detable$log.pct.diff <- log2(detable$pct.1 / (detable$pct.2 + 0.01) )


#markers_allCells <- FindAllMarkers(gex, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
#markers_allCells_top <- markers_allCells %>% group_by(cluster) %>% top_n(n = 9, wt = avg_log2FC)
#markers_allCells_top

detable %>% group_by(cluster) %>% top_n(-30, p_val)  %>% top_n(20, pct.diff) -> top5
ord <- factor(sapply(unique(as.character(top5$gene)), function(x){getcluster(DGE_temp, 
                                                                             x, 
                                                                             "seurat_clusters")}))
genes_to_plot <- unique(as.character(top5$gene))[order(as.numeric( as.character(ord) ))]

pdf(paste0("../results/DGE_allCells_seurat_clusters_dotplot_", format(Sys.time(), "%y%m%d"), ".pdf"), width = 10, height = 50)
rafalib::mypar(1, 1, mar = c(6, 6, 1, 5))
plot_dots(gex, 
          genes_to_plot, 
          clustering = "seurat_clusters", 
          show_grid = T, 
          main = "top cluster markers", 
          cex.main = 1, font.main = 1, cex.col = 1.3, srt = 0, cex.row = 1.3)
dev.off()


```


# annotating clusters

```{r}
#manual cluster annotation
# annotation <- c(
#   B_cell = 0,
#   B_cell = 1,
#   B_cell = 2,
#   B_cell = 3,
#   B_cell = 6,
#   B_cell = 7,
#   T_cell = 4,
#   Mono = 5,
#   Plasma = 8
# )

#auto annotation
annotation <- c(
  T_cell = getcluster(gex,"IL7R", "seurat_clusters"),
  NK_cell = getcluster(gex,"GZMA", "seurat_clusters"),
  Mono = getcluster(gex,"CST3", "seurat_clusters"),
  B_cell = getcluster(gex,"PF4", "seurat_clusters"),
  B_cell = getcluster(gex,"BACH2", "seurat_clusters"),
  B_cell = getcluster(gex,"TCL1A", "seurat_clusters"),
  B_cell = getcluster(gex,"IGHA1", "seurat_clusters"),
  B_cell = getcluster(gex,"SMARCB1", "seurat_clusters"),
  Plasma = getcluster(gex,"JCHAIN", "seurat_clusters")

)

gex$celltype <- names(annotation)[match(gex$seurat_clusters, annotation)] 

p1.celltype <- DimPlot(gex, 
                  reduction = "umap_1", 
                  pt.size = .1, 
                  group.by = "celltype", 
                  label = TRUE, 
                  repel = TRUE, 
                  raster = FALSE) + 
              NoLegend() + 
              ggtitle("UMAP, by celltype")

ggsave2(paste0("../results/UMAP_celltypes_allCells_", format(Sys.time(), "%y%m%d"), ".png"), 
        plot_grid(p1.u.s, p1.u.c, p1.celltype, ncol = 3, labels = "AUTO", align = "h"), 
        width = 32, height = 12, unit = "cm")


```


# subset GEX to include only B-cell clusters

```{r}

Idents(object = gex) <- "celltype"
gex <- subset(gex, idents = c("B_cell", "Plasma"))

colnames(gex@meta.data)[colnames(gex@meta.data) == "seurat_clusters"] <- "seurat_clusters_1"
DefaultAssay(object = gex) <- "RNA"; gex
Idents(object = gex) <- "orig.ident"

# An object of class Seurat 
# 19707 features across 232563 samples within 1 assay 
# Active assay: RNA (19707 features, 3734 variable features)
#  3 dimensional reductions calculated: pca_1, harmony_1, umap_1

paste0("B-cell subset done at:", Sys.time())


```


# save rds GEX9

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX9_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX9_xxxxxx.rds")

gex9.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex9.dnsample.500, file = paste0("../results/GEX9_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX9 size:", print(object.size(gex), units = "GB"))


```


# get variable features B-cells
 
```{r}

gex <- FindVariableFeatures(gex, selection.method = "vst", nfeatures = 4000)


```


# plot variable features B-cells
 
```{r}

p1 <- LabelPoints(plot = VariableFeaturePlot(gex), 
                  points = head(VariableFeatures(gex), 80), 
                  repel = TRUE,
                  xnudge = 0,
                  ynudge = 0,
                  max.overlaps = Inf) + 
        theme_classic() +
        ggtitle(paste0(length(VariableFeatures(gex)), " Variable features - B-cells"))

ggsave2(paste0("../results/VariableFeatures_B-cells_", length(VariableFeatures(gex)), "_", format(Sys.time(), "%y%m%d"), ".png"),
        p1, width = 25, height = 20, unit = "cm")


```


# Subset VariableFeatures(gex) to only include biotype == "protein_coding"

```{r}

VariableFeatures(gex) <- VariableFeatures(gex)[VariableFeatures(gex) %in% 
                                                 annot$external_gene_name[annot$gene_biotype == "protein_coding"]]
length(VariableFeatures(gex)) #[1] 3779

p1 <- LabelPoints(plot = VariableFeaturePlot(gex), 
                  points = head(VariableFeatures(gex), 80), 
                  repel = TRUE,
                  xnudge = 0,
                  ynudge = 0,
                  max.overlaps = Inf) + 
        theme_classic() +
        ggtitle(paste0(length(VariableFeatures(gex)), " Variable features - B-cells (protein_coding)"))

ggsave2(paste0("../results/VariableFeatures_B-cells_", length(VariableFeatures(gex)), "_protein_coding_", format(Sys.time(), "%y%m%d"), ".png"), 
        p1, width = 25, height = 20, unit = "cm")

                       
```


# Scale data B-cells

```{r}

gex <- ScaleData(gex, 
                 features = VariableFeatures(object = gex),
                 vars.to.regress = c("nFeature_RNA", "percent_mito", "percent_ribo"))

paste0("Varfeat and Scale done at:", Sys.time())


```


# save rds GEX10

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX10_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX10_xxxxxx.rds")

gex10.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex10.dnsample.500, file = paste0("../results/GEX10_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX10 size:", print(object.size(gex), units = "GB"))


```


# PCA B-cells

```{r}

#RunPCA() takes around 2h for this data set

gex <- RunPCA(gex,
              assay = "RNA",
              features = VariableFeatures(object = gex),
              npcs = 100,
              reduction.name = "pca_2",
              verbose = TRUE)

gex@assays$RNA@scale.data <- matrix(0); gc()

paste0("PCA2 done at:", Sys.time())


```


# save rds GEX11

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX11_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX11_xxxxxx.rds")

gex11.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex11.dnsample.500, file = paste0("../results/GEX11_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX11 size:", print(object.size(gex), units = "GB"))


```


# Run Harmony B-cells

```{r}

gex <- RunHarmony(gex, 
                  group.by.vars = "orig.ident",
                  reduction = "pca_2",
                  reduction.save = "harmony_2")

paste0("harmony2 done at:", Sys.time())


```


# plot PCA and harmony B-cells

```{r}

p.p <- DimPlot(object = gex, reduction = "pca_2", pt.size = .1, group.by = "orig.ident", raster=FALSE) + 
            NoLegend() + 
            ggtitle("PCA, by Sample")

p.h <- DimPlot(object = gex, reduction = "harmony_2", pt.size = .1, group.by = "orig.ident", raster=FALSE) + 
                NoLegend() + 
                ggtitle("Harmony, by Sample")

ggsave2(paste0("../results/PCA_harmony_B-cells_", format(Sys.time(), "%y%m%d"), ".png"), 
        plot_grid(p.p, p.h, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


```


# save rds GEX12

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX12_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX12_210621.rds")

gex12.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex12.dnsample.500, file = paste0("../results/GEX12_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX12 size:", print(object.size(gex), units = "GB"))


```


# run UMAP B-cells

```{r}

set.seed(42)

# gex <- RunUMAP(gex, dims = 1:50, #or 100?
#                 reduction = "harmony_1",
#                 metric = "correlation",
#                 reduction.name = "umap_1",
#                 min.dist = 0.4, #local cell separation
#                 spread = .5, #global cell separation
#                 n.neighbors = 30, #30 on bianca!
#                 repulsion.strength = 0.4, #~2x min.dist, repulsion to cells fr annat cluster, global separation
#                 negative.sample.rate = 50, #ant ggr n.neighbors celler, global distance
#                 n.epochs = 100,
#                 n.components = 2)

gex <- RunUMAP(gex, dims = 1:50, 
                reduction = "harmony_2",
                metric = "correlation",
                reduction.name = "umap_2",
                min.dist = 0.01,  #local cell separation
                spread = .3, #global cell separation
                n.neighbors = 15,
                repulsion.strength = 0.02, #~2x min.dist, repulsion to cells fr annat cluster, global
                negative.sample.rate = 50, #ant ggr n.neighbors celler, global distance
                n.epochs = 100,
                n.components = 2)

set.seed(42)
gex <- RunUMAP(gex, dims = 1:50,
                reduction = "harmony_2",
                metric = "correlation",
                reduction.name = "umap_2_3d",
                min.dist = 0.01, #local cell separation
                spread = .3, #global cell separation
                n.neighbors = 15,
                repulsion.strength = 0.02, #~2x min.dist, repulsion to cells fr annat cluster, global
                negative.sample.rate = 50, #ant ggr n.neighbors celler, global distance
                n.epochs = 100,
                n.components = 3,
                ret_extra = c("model", "nn", "fgraph"))

paste0("UMAP2 done at:", Sys.time())


```


# save rds GEX13

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX13_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX13_210621.rds")

gex13.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex13.dnsample.500, file = paste0("../results/GEX13_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))

paste0("GEX13 size:", print(object.size(gex), units = "GB"))


```


# clustering B-cells

```{r}


gex <- FindNeighbors(gex, 
                     reduction = "harmony_2", 
                     dims = 1:50,
                     k.param = 15,
                     verbose = TRUE)

#res <- 0.15 #RNA_snn res local
#res <- 0.6 #RNA_nn res local (knn, similar graph generated during UMAP)
#res <- 0.05 #Bianca RNA_snn resolution
res <- 0.4 #Bianca RNA_nn resolution

gex <- FindClusters(gex, resolution = res, verbose = TRUE, graph.name = "RNA_nn")
table(gex@meta.data$seurat_clusters)

paste0("Clustering2 done at:", Sys.time())

```


# save rds GEX14

```{r include = FALSE}

saveRDS(gex, file = paste0("../results/GEX14_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX14_210628.rds")

gex14.dnsample.500 <-  subset(x = gex, downsample = 500)
saveRDS(gex14.dnsample.500, file = paste0("../results/GEX14_dnsample.500_", format(Sys.time(), "%y%m%d"), ".rds"))
#gex <- readRDS("../results/GEX14_dnsample.500_210628.rds")

paste0("GEX14 size:", print(object.size(gex), units = "GB"))

#sort(sapply(ls(), function(x){format(object.size(get(x)), units = "Gb")}))

```


# hclust B-cell clusters

```{r}

cluster_means <- sapply(as.character(unique(gex$seurat_clusters)), function(x){
  rowMeans( gex@assays$RNA@data[, gex$seurat_clusters == x] )
} )

adj <- (1-cor(cluster_means))/2
h <- hclust( as.dist( adj ), method = "ward.D2")

# my_dist <- dist(cluster_means,method = "euclidean")
# h <- hclust(my_dist, method = "complete")
# pheatmap::pheatmap(adj)

png(paste0("../results/hclust_clusters_B-cells_clustRes_", res, "_",format(Sys.time(), "%y%m%d"), ".png"), 
    width = 2000, 
    height = 2000, 
    units = "px")
plot(as.dendrogram(h))
dev.off()


```


# plot UMAP 2D B-cells

```{r}

p1.u.s <- DimPlot(gex, 
                  reduction = "umap_2", 
                  pt.size = .1, 
                  group.by = "orig.ident", 
                  raster=FALSE) + 
              NoLegend() + 
              ggtitle("UMAP, by Sample")

p1.u.c <- DimPlot(gex, 
                  reduction = "umap_2", 
                  pt.size = .1, 
                  group.by = "seurat_clusters", 
                  label = TRUE, 
                  repel = TRUE, 
                  raster=FALSE) + 
              NoLegend() + 
              ggtitle("UMAP, by Cluster")

ggsave2(paste0("../results/UMAP_B-cells_", res, "_", format(Sys.time(), "%y%m%d"), ".png"), 
        plot_grid(p1.u.s, p1.u.c, ncol = 2, labels = "AUTO", align = "h"), 
        width = 30, height = 15, unit = "cm")


```


# plot UMAP 3D B-cells
 
```{r}

df <- data.frame(gex@reductions$umap_2_3d@cell.embeddings)
df <- data.frame(df, seurat_clusters = gex$seurat_clusters, orig.ident = gex$orig.ident)
colnames(df)[1:3] <- c("UMAP_1","UMAP_2","UMAP_3"); head(df)
pal <- c(scales::hue_pal()(8), RColorBrewer::brewer.pal(9,"Set1"),  RColorBrewer::brewer.pal(8,"Set2") )


####------- by cluster
p_State <- plot_ly(df, 
                   x = ~UMAP_1, 
                   y = ~UMAP_2, 
                   z = ~UMAP_3, 
                   color = ~seurat_clusters,
                   colors = pal, size=.5) %>%  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'UMAP_1'), 
                      yaxis = list(title = 'UMAP_2'), 
                      zaxis = list(title = 'UMAP_3')))

htmlwidgets::saveWidget(p_State, 
                        paste0("../results/UMAP_B-cells_3d_clusters_", format(Sys.time(), "%y%m%d"), ".html"))


####------- by orig.ident
p_State <- plot_ly(df, 
                   x = ~UMAP_1, 
                   y = ~UMAP_2, 
                   z = ~UMAP_3, 
                   color = ~orig.ident,
                   colors = pal, size = .5) %>%  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'UMAP_1'), 
                      yaxis = list(title = 'UMAP_2'), 
                      zaxis = list(title = 'UMAP_3')))

htmlwidgets::saveWidget(p_State, 
                        paste0("../results/UMAP_B-cells_3d_orig.ident_", format(Sys.time(), "%y%m%d"), ".html"))


```


# plot marker genes

```{r}

#ig.features <- c("IGHA1", "IGHA2", "IGHG1", "IGHG2", "IGHG3", "IGHG4",  "IGHD", "IGHE", "IGHM" )
#b.markers <- c("CD79A", "CD79B", "MS4A1", "CD19", "CD27", "IGHA1", "IGHD", "IGHM", "JCHAIN", "MME")
#pbmc.markers <- c("MS4A1", "CD19", "CD27", "CD79A", "GNLY", "CD3E", "CD14", "LYZ", "CD8A")
#my_pars <- c("nCount_RNA", "nFeature_RNA", "percent_mito", "percent_ribo", "percent_hb")

#pbmc
p.pbmc.markers.f <- FeaturePlot(gex, features = pbmc.markers, reduction = "umap_2", raster=FALSE)
ggsave2(paste0("../results/FeaturePlot_B-cells_pbmcMarkers_", format(Sys.time(), "%y%m%d"), ".png"), 
        p.pbmc.markers.f, 
        width=25, height=25, unit="cm")

#b markers
p.b.markers.f <- FeaturePlot(gex, features = b.markers, reduction = "umap_2", raster=FALSE)
ggsave2(paste0("../results/FeaturePlot_B-cells_BMarkers_", format(Sys.time(), "%y%m%d"), ".png"), 
        p.b.markers.f, 
        width=25, height=25, unit="cm")

#ig markers
p.ig.markers.f <- FeaturePlot(gex, features = ig.features, reduction = "umap_2", raster=FALSE)
ggsave2(paste0("../results/FeaturePlot_B-cells_IGMarkers_", format(Sys.time(), "%y%m%d"), ".png"), 
        p.ig.markers.f, 
        width=25, height=25, unit="cm")

#my_pars
p.myPars.f <- FeaturePlot(gex, features = my_pars , reduction = "umap_2", raster=FALSE, order = TRUE)
ggsave2(paste0("../results/FeaturePlot_B-cells_myPars", format(Sys.time(), "%y%m%d"), ".png"), 
        p.myPars.f, 
        width=25, height=25, unit="cm")


```


# get B-cell cluster marker genes

```{r}

sample_size <- table(gex$seurat_clusters)
sample_size[sample_size > 100] <- 100
DGE_cells <- lapply(names(sample_size), function(x){
  set.seed(1)
  sample( colnames(gex)[ gex$seurat_clusters == x ], size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)
DGE_temp<- gex[, DGE_cells]
DGE_temp <- SetIdent(DGE_temp, value = "seurat_clusters")
detable <- FindAllMarkers( DGE_temp, 
                           min.pct = 0.1, 
                           min.cells.feature = 0.1, 
                           max.cells.per.ident = 50 , 
                           only.pos = T , 
                           logfc.threshold = 0.1, 
                           assay = "RNA" )

detable$pct.diff <- detable$pct.1 - detable$pct.2
write.csv(detable, paste0("../results/DGE_B-cells_seurat_clusters_", format(Sys.time(), "%y%m%d"), ".csv"))

detable <- detable[ detable$p_val < 0.05,  ]
detable$log.pct.diff <- log2(detable$pct.1 / (detable2$pct.2 + 0.01) )

detable %>% group_by(cluster) %>%  top_n(-30, p_val)  %>% top_n(20, log.pct.diff) -> top5
ord <- factor(sapply(unique(as.character(top5$gene)), 
                     function(x){getcluster(DGE_temp, x, "seurat_clusters")}))
genes_to_plot <- unique(as.character(top5$gene))[order(as.numeric( as.character(ord) ))]

pdf(paste0("../results/DGE_B-cells_seurat_clusters_dotplot_", format(Sys.time(), "%y%m%d"), ".pdf"), 
    width = 5, height = 20)
rafalib::mypar(1, 1, mar=c(6, 6, 1, 5))
plot_dots(gex, 
          genes_to_plot, 
          clustering = "seurat_clusters", 
          show_grid = T, 
          main = "top cluster markers", 
          cex.main = 1, font.main = 1, cex.col = 1.3, srt = 0, cex.row = 1.3)
dev.off()

detable[detable$gene=="FCER1G",]
rafalib::mypar(2, 1, mar=c(1,1,2,4))

plot_feat(gex,red="umap_2",feat="FCER1G",cex=.8)


```


#annotate B-cell clusters

```{r}




```


#get B-cell celltype fractions per group

```{r}
# gex$patient_group <- paste0(ifelse(gex$SSA=="YES","SSA",""),"_",ifelse(gex$SSB=="YES","SSB",""))
# pat_annot <- c(SSAB="SSA_SSB",SSA="SSA_",DNEG="NA_NA",CTRL="_")
# gex$patient_group <- names(pat_annot)[match(gex$patient_group,pat_annot)]
# table(gex$patient_group)
# 
# 
# temp <- table( list(patient_group=gex$patient_group , seurat_clusters=gex$seurat_clusters)  )
# temp2 <- t(temp / rowSums(temp))
# barplot(temp2)
# 
# temp3 <- t( t(temp) / colSums(temp))
# barplot(temp3,las=1)


#Four non-overlapping groups
#1. SSA+/SSB+
#2. SSA+/SSB-
#3. SSA-/SSB-
#4. CTRL


```


#DE tests

```{r}

# for(i in unique(gex$seurat_clusters)){
# subset_gex <- gex[,gex$seurat_clusters == "0"]
# 
# sample_size <- table(subset_gex$patient_group)
# sample_size[sample_size > 50] <- 50
# DGE_cells <- lapply(names(sample_size), function(x){
#   set.seed(1)
#   sample( colnames(subset_gex)[ subset_gex$patient_group == x ], size = sample_size[x])
#   })
# DGE_cells <- unlist(DGE_cells)
# DGE_temp<- subset_gex[, DGE_cells]
# DGE_temp <- SetIdent(DGE_temp, value = "patient_group")
# 
# 
# detable <- FindAllMarkers( DGE_temp, 
#                            min.pct = 0.1, 
#                            min.cells.feature = 0.1, 
#                            max.cells.per.ident = 200 , 
#                            only.pos = T , 
#                            logfc.threshold = 0.1, 
#                            assay = "RNA" )
# detable <- detable[ detable$p_val < 0.05,  ]
# detable$pct.diff <- detable$pct.1 - detable$pct.2
# detable <- detable[ detable$pct.diff > 0.1,  ]
# 
# detable$log.pct.diff <- log2(detable$pct.1 / (detable$pct.2 + 0.01) )
# 
# write.csv(detable, paste0("../results/DGE_all_cluster_",i,"_Cells_patient_group_",format(Sys.time(), "%y%m%d"),".csv"))
# 
# #markers_allCells <- FindAllMarkers(gex, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
# #markers_allCells_top <- markers_allCells %>% group_by(cluster) %>% top_n(n = 9, wt = avg_log2FC)
# #markers_allCells_top
# 
# detable %>% group_by(cluster) %>% top_n(50, avg_log2FC)  %>% top_n(10, pct.diff) -> top5
# ord <- factor(sapply(unique(as.character(top5$gene)), function(x){getcluster(DGE_temp, 
#                                                                              x, 
#                                                                              "patient_group")}))
# genes_to_plot <- unique(as.character(top5$gene))[order(as.numeric( as.character(ord) ))]
# 
# pdf(paste0("../results/DGE_all_cluster_",i,"_Cells_patient_group_dotplot_", format(Sys.time(), "%y%m%d"), ".pdf"), width = 10, height = 50)
# rafalib::mypar(1, 1, mar = c(2, 6, 1, 5))
# plot_dots(gex, 
#           genes_to_plot, 
#           clustering = "patient_group", 
#           show_grid = T, 
#           main = "top cluster markers", 
#           cex.main = 1, font.main = 1, cex.col = 1, srt = 0, cex.row = .6)
# dev.off()

# #Four non-overlapping groups
# #1. SSA+/SSB+
# #2. SSA+/SSB-
# #3. SSA-/SSB-
# #4. CTRL
# 
# }


```


# Append 10x VDJ clonotypes

```{r}



```


# Append TRUST4 clonotypes

```{r}



```


# trajectory analysis
## Graph abstraction

```{r}
# library(igraph)
# library(Matrix)
# 
# # gex <- FindNeighbors(gex, reduction = "harmony_2",dims = 1:50,k.param = 15)
# g2 <- graph_abstraction(gex, 
#                         red = "umap_2", 
#                         clustering = "seurat_clusters", 
#                         graph = "RNA_snn", 
#                         cutoff = 0)
# 
# png(filename = paste0("../results/Graph_abstraction.png"), width = 1600*4, height = 1600*4, res = 400)
# 
# mypar(4,4)
# plot_meta(x = gex,red = "umap_2",feat = "orig.ident",frame=F,label = T)
# plot_meta(x = gex,red = "umap_2",feat = "seurat_clusters",frame=F,label = T)
# 
# for(i in 10^-(0:10) ){
#   centroids <- g2[g2$s==g2$p.Var, ]
#   g <- g2[as.numeric(g2$p.Freq) > i, ]
#   g <- g[as.numeric(g$p.Freq) > 0, ]
#   g <- g[g$s != g$p.Var, ]
#   g$norm.Freq <- g$p.Freq / max(g$p.Freq)
#   
#   plot_meta(x = gex, red = "umap_2", feat = "seurat_clusters", frame=T, label = F, main = paste0("cutoff=",i))
#   gpal <- colorRampPalette(c("grey20", "black"))(20)
#   
#   apply(g, 1, function(x){
#     lines(as.numeric(x[4:5]), as.numeric(x[6:7]),
#                                  lwd=5*as.numeric(x[8]),
#                                  col=gpal[round(as.numeric(x[8])*18)+1]) })
#   points(centroids[, 4],centroids[, 6],
#          bg='white',
#          pch=21, cex=1)
#   text(centroids[, 4], centroids[, 6], labels = centroids[, 1], cex = .3)
# }
# dev.off()
# 
# 
# ```
# 
# 
# # Plot Graph abstraction
# 
# ```{r}
# png(filename = paste0(opt$output_path,"/Graph_abstraction_matrix.png"),width = 1600,height = 1600,res = 200)
# 
# sizes <- matrix( g2$p.Freq , sqrt( nrow(g2) ) , sqrt( nrow(g2) ))
# colnames(sizes) <- unique(g2$p.Var)
# rownames(sizes) <- unique(g2$p.Var)
# 
# o <- hclust(dist(sizes,method = "binary"),method="complete")$order
# # sizes <- sizes[o,o]
# 
# # sizes[sizes > 7] <- 7
# sizes <- sizes * lower.tri(sizes)
# sizes <- sizes / max(sizes)
# 
# mypar(1,1,mar=c(4,2,4,8))
# plot(0,0, las=1,xlim=c(.5,ncol(sizes)+.5),ylim=c(.5,nrow(sizes)+.5),
#        axes=F,ylab="",xlab="",main="cluster connectivity",xaxs="i",yaxs="i", frame=T)
# for(i in 1:ncol(sizes)){
#   lines(c(i,i),c(nrow(sizes)+.5,0),col="grey95",lwd=.5,xpd=F)
# }
# for(i in 1:nrow(sizes) ){
#   lines(c(0,ncol(sizes)+0.5),c(i,i),col="grey95",lwd=.5,xpd=F)
# }
# points(sort(rep(1:ncol(sizes),nrow(sizes))),
#       rep(nrow(sizes):1,ncol(sizes)),
#         col=colorRampPalette(c("grey80","black"))(91)[cut( c(sizes) , 
#                                                            breaks = seq(0,1,length.out = 92) )],pch=16 ,
#       cex= sizes*2 + 0.3)
# srt=90
# text(1:ncol(sizes), par("usr")[3] - (par("usr")[4])/200, labels = colnames(sizes), srt = srt,
#      adj = c(ifelse(srt==0,.5,ifelse(srt==90,1,1)),ifelse(srt==0,1,ifelse(srt==90,.5,1))),
#      xpd = TRUE, cex=1)
# text(par("usr")[1] , nrow(sizes):1 , labels = rownames(sizes), srt = 0, adj = c(1,0.5), xpd = TRUE, cex=1)
# 
# add_size_legend(min.cex = 0.3,max.cex = 2,
#                 labels = format(seq(min(g$p.Freq),max(g$p.Freq),length.out = 5),scientific = T) )
# add_scale_legend(pal = colorRampPalette( c("grey","black") )(99),labels = c("min","max"))
# 
# dev.off()
# ```
# 
# 
# # Plot UMAP for QC parameters
# 
# ```{r}
# png(filename = paste0(opt$output_path,"/UMAP_QC.png"),width = 800*4,height = 800*4,res = 300)
# 
# feat_list <- c("nFeature_RNA","DoubletFinderScore","perc_mito","perc_rpl",
#                "perc_rps","perc_lncRNA","perc_Chr_Y","perc_protein_coding",
#                "G2M.Score","S.Score","CC.Diff","G1.Score")
# 
# rafalib::mypar(4,4,mar=c(0.5,0.5,2,0.5))
# plot_meta(x = DATA,red = "umap_harmony",feat = "Age",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "SampleID",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "seurat_clusters",frame=F,col = pal,label = T)
# for(i in feat_list){plot_feat(x = DATA,red = "umap_harmony",feat = i,frame=F)}
# dev.off()
# ```
# 
# 
# # Plot UMAP for a couple of genes
# 
# ```{r}
# png(filename = paste0(opt$output_path,"/UMAP_genes.png"),width = 800*5,height = 800*5,res = 300)
# 
# feat_list <- c("NRXN3","CACNA2D1","NEFL","SPARCL1",
#                "OLIG1","ATAD2","SPC25","MBP","SPP1","COL1A1","GATA3",
#                "KCNE5","RSPH1","MPZ","ITM2A","MSX1","EFNB3","MKI67","CDC20","CDK1")
# rafalib::mypar(5,5,mar=c(0.5,0.5,2,0.5))
# plot_meta(x = DATA,red = "umap_harmony",feat = "Age",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "SampleID",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "seurat_clusters",frame=F,col = pal,label = T)
# for(i in feat_list){plot_feat(x = DATA,red = "umap_harmony",feat = i,frame=F)}
# dev.off()


```




```{r}
# png(filename = paste0(opt$output_path,"/UMAP_genes_neurons.png"),width = 800*4,height = 800*4,res = 300)
# 
# feat_list <- c("KCNIP4","EBF1","EBF2",
#                "SPOCK3","NPY","ROBO3")
# 
# rafalib::mypar(4,4,mar=c(0.5,0.5,2,0.5))
# 
# niceRplots::plot_meta(x = DATA,
#                       red = "umap_harmony",
#                       feat = "Age",frame=F,col = pal)
# 
# for(i in feat_list){
#   niceRplots::plot_feat(x = DATA,
#                         red = "umap_harmony",
#                         feat = i,frame=F)}
# dev.off()


```



# Run UMAP in 3D

```{r}

# DATA <- RunUMAP(DATA, dims = 1:50, 
#                 reduction = "harmony",
#                 metric = "correlation",
#                 reduction.name = "umap_harmony_3D",
#                 min.dist = 0.01,
#                 spread = .3,
#                 n.neighbors = 10,
#                 n.epochs = 100,
#                 n.components = 3 )
# 
# 
# library(plotly)
# 
# df <- data.frame(DATA@reductions$umap_harmony_3D@cell.embeddings)
# df <- data.frame(df,seurat_clusters=DATA$seurat_clusters,sampleid = DATA$SampleID,age = DATA$Age)
# colnames(df)[1:3] <- c("UMAP_1","UMAP_2","UMAP_3")
# pal <- c(scales::hue_pal()(8),RColorBrewer::brewer.pal(9,"Set1"),RColorBrewer::brewer.pal(8,"Set2") )
# 
# p_State <- plot_ly(df,x = ~UMAP_1, y = ~UMAP_2, z=~UMAP_3,color = ~seurat_clusters,
#                    colors = pal, size=.5) %>%  add_markers() %>%
#   layout(scene = list(xaxis = list(title = 'UMAP_1'), yaxis = list(title = 'UMAP_2'),zaxis = list(title = 'UMAP_3')))
# htmlwidgets::saveWidget(p_State, "./umap_3d_clusters.html")
# 
# ggplotly(p_State,height=800,width=800)
# 
# p_State <- plot_ly(df,x = ~UMAP_1, y = ~UMAP_2, z=~UMAP_3,color = ~sampleid,
#                    colors = pal, size=.5) %>%  add_markers() %>%
#   layout(scene = list(xaxis = list(title = 'UMAP_1'), yaxis = list(title = 'UMAP_2'),zaxis = list(title = 'UMAP_3')))
# htmlwidgets::saveWidget(p_State, "./umap_3d_sampleID.html")
# 
# ggplotly(p_State,height=800,width=800)
# 
# 
# p_State <- plot_ly(df,x = ~UMAP_1, y = ~UMAP_2, z=~UMAP_3,color = ~age,
#                    colors = pal, size=.5) %>%  add_markers() %>%
#   layout(scene = list(xaxis = list(title = 'UMAP_1'), yaxis = list(title = 'UMAP_2'),zaxis = list(title = 'UMAP_3')))
# htmlwidgets::saveWidget(p_State, "./umap_3d_age.html")
# 
# ggplotly(p_State,height=800,width=800)
# gc()


```


```{r}

# probs <- matrix( g2$p.Freq , sqrt( nrow(g2) ) , sqrt( nrow(g2) ))
# colnames(probs) <- unique(g2$p.Var)
# rownames(probs) <- unique(g2$p.Var)
# 
# g3 <- graph_from_adjacency_matrix( probs , weighted = T , diag = T , mode="undirected")
# 
# mypath <- igraph::shortest_paths( g3, from = "0", to = "6" , weights = 1/E(g3)$weight )
# mypath$vpath[[1]]
# 
# mypath <- igraph::shortest_paths( g3, from = "0", to = "22" , weights = 1/E(g3)$weight )
# mypath$vpath[[1]]
# 
# mypath <- igraph::shortest_paths( g3, from = "0", to = "13" , weights = 1/E(g3)$weight )
# mypath$vpath[[1]]
# 
# mypath <- igraph::shortest_paths( g3, from = "0", to = "39" , weights = 1/E(g3)$weight )
# mypath$vpath[[1]]
# 
# mypath <- igraph::shortest_paths( g3, from = "0", to = "17" , weights = 1/E(g3)$weight )
# mypath$vpath[[1]]
# 
# 
# 
# 
# non_neuronal_clusters <- c(11, 29, 33, 27, 18, 35, 28, 36, 25, 41)
# DATA$non_neuronal <- DATA$seurat_clusters %in% non_neuronal_clusters
# 
# rafalib::mypar(2,2,mar=c(0.5,0.5,2,0.5))
# plot_meta(x = DATA,red = "umap_harmony",feat = "Age",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "SampleID",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "seurat_clusters",frame=F,col = pal,label = T)
# plot_meta(x = DATA,red = "umap_harmony",feat = "non_neuronal",frame=F,col = pal)
# 
# library(tradeSeq)
# library(slingshot)
# 
# 
# 
# mm <- model.matrix( ~ 0 + factor(DATA$seurat_clusters) )
# colnames(mm) <- levels(factor(DATA$seurat_clusters))
# res <- (DATA@graphs$RNA_snn>0)*1
# res <- res %*% as.matrix(mm)
# own <- rowSums(res * mm)
# all <- rowSums(res)
# DATA$specificity <- own/all
# DATA$inspecificity <- 1-own/all
# plot_meta(x = DATA,red = "umap_harmony",feat = "seurat_clusters",frame=F,col = pal,label = T)
# niceRplots::plot_feat(x = DATA, red = "umap_harmony", feat = "specificity",frame=F)
# niceRplots::plot_feat(x = DATA, red = "umap_harmony", feat = "inspecificity",frame=F)


```


# Investigating one branching 

```{r}

# sel <- DATA$seurat_clusters %in% c(16,9,15,5,0,34,8,23,21,20)
# temp2 <- DATA[,sel]
# temp2 <- RunUMAP(temp2, dims = 1:100, 
#                 reduction = "harmony",
#                 metric = "correlation",
#                 reduction.name = "umap_traj",
#                 min.dist = 0.01,
#                 spread = .3,
#                 n.neighbors = 10,
#                 repulsion.strength = 0.01,
#                 negative.sample.rate = 20,
#                 n.epochs = 200,
#                 n.components = 2 )
# png(filename = paste0(opt$output_path,"/progenitor_trajectory_umap.png"),width = 800*4,height = 800*4,res = 300)
# mypar(4,4)
# plot_meta(x = temp2,red = "umap_traj",feat = "seurat_clusters",frame=F,label = T)
# dev.off()
# 
# sel <- DATA$seurat_clusters %in% c(16,9,15,5,0)
# temp <- DATA[,sel]
# 
# temp <- RunUMAP(temp, dims = 1:100, 
#                 reduction = "harmony",
#                 metric = "correlation",
#                 reduction.name = "umap_traj",
#                 min.dist = 0.01,
#                 spread = .3,
#                 n.neighbors = 10,
#                 repulsion.strength = 1,
#                 negative.sample.rate = 1,
#                 n.epochs = 100,
#                 n.components = 2 )
# plot_meta(x = temp,red = "umap_traj",feat = "seurat_clusters",frame=F,col = pal,label = T)
# 
# 
# 
# # dm <- destiny::DiffusionMap( DATA@reductions$harmony@cell.embeddings[sel,], k = 4, n_eigs = 2 )
# # rownames(dm@eigenvectors) <- rownames(DATA@reductions$harmony@cell.embeddings[sel,])
# # temp@reductions[["dm"]] <- CreateDimReducObject(embeddings = dm@eigenvectors,key = "DC_",assay = "RNA")
# # plot_meta(x = temp,red = "dm",feat = "seurat_clusters",frame=F,col = pal,label = T,dims = c(1,2))
# 
# # temp <- FindNeighbors(temp,reduction = "harmony",k.param = 15,dims = 1:50)
# # temp <- FindClusters(temp,resolution = .3,graph.name = "RNA_snn")
# # plot_meta(x = temp,red = "umap_traj",feat = "seurat_clusters",frame=F,col = pal,label = T)
# 
# dimred <- temp@reductions$umap_traj@cell.embeddings
# clustering <- factor(DATA$seurat_clusters[sel])
# # dimvis <- DATA@reductions$umap_harmony@cell.embeddings[sel,c(1,2)]
# dimvis <- temp@reductions$umap_traj@cell.embeddings
# 
# library(slingshot)
# 
# set.seed(1)
# lineages <- getLineages(data = dimred,
#                         clusterLabels = clustering,
#                         start.clus = "0",
#                         end.clus=c("9","15"))
# lineages
# 
# lineages@reducedDim <- dimvis
# curves <- getCurves(lineages, thresh = 0.001, stretch = .000001, allow.breaks = T,approx_points = 30)
# pseudotime <- slingPseudotime(curves, na = FALSE)
# cellWeights <- slingCurveWeights(curves)
# 
# png(filename = paste0(opt$output_path,"/Neuron_trajectory_umap.png"),width = 800*4,height = 800*4,res = 300)
# mypar(4,4)
# plot_meta(x = temp,red = "umap_traj",feat = "seurat_clusters",frame=F,label = T)
# # lines(lineages, lwd = 2, col = c('black',"firebrick","navy","darkgreen") )
# lines(curves, lwd = 3, col = c('black',"firebrick","navy","darkgreen"))
# dev.off()
# 
# 
# 
# 
# set.seed(1)
# sel2 <- sample( colnames(DATA[,sel]) , size = 1000)
# counts <- DATA@assays$RNA@counts[Matrix::rowSums(DATA@assays$RNA@counts[,sel2] > 1) > 50 ,sel2]
# dim(counts)
# DATA <- NormalizeData(DATA,scale.factor = 1000)


```


# Plotting expression for up to 50 top differentially expressed genes

```{r}

# library(tradeSeq)
# sce <- fitGAM(counts = counts,
#               pseudotime = pseudotime[sel2,],
#               cellWeights = cellWeights[sel2,],
#                  nknots = 5, verbose = T,parallel=T)
# 
# 
# patternRes <- patternTest(sce)
# patternRes$pvalue[is.na(patternRes$pvalue)] <- 1
# patternRes <- patternRes[order(patternRes$waldStat,decreasing = T),]
# patternRes$FDR <- p.adjust(patternRes$pvalue,method = "BH")
# patternRes <- patternRes[ patternRes$FDR < 0.05 , ]
# head(patternRes,20)
# 
# gene_list <- rownames(patternRes)[1:min(50,nrow(patternRes))]
# lims <- quantile(pseudotime,c(0.02,.98) )
# res <- t(sapply(gene_list,
#               pseudotime = slingPseudotime(curves, na = FALSE),
#               cellWeights = slingCurveWeights(curves),
#               lims=lims,
#               data=DATA@assays$RNA@data[gene_list,],
#               function(gene,data,pseudotime,cellWeights,lims) {
#   ll <- lapply( 1:ncol(pseudotime),function(i){
#   l1 <- (cellWeights[,i] == 1 ) & (pt1 > lims[1]) & (pt1 < lims[2])
#   l1 <- colnames(temp)[l1]
#   sm <- spline(smooth.spline( pt1[ l1 ], data[gene,l1], nknots = 20,spar = .8),n = 20)
#   })
#   return( c( rev( ll[[1]]$y ) ,rep(NA,1), ll[[2]]$y ) )
# }))
# ```
# 
# 
# 
# # Plotting expression for one gene
# 
# 
# ```{r}
# gene <- "LAMP5"
# all <- c(sce@colData$tradeSeq$dm$t1,sce@colData$tradeSeq$dm$t2)
# lims <- quantile(pseudotime,c(0.02,.98) )
# plot( 0,cex=.5,pch=16,col="purple",type="n",las=1,xlim=lims,
#       ylim=range(DATA@assays$RNA@data[gene,rownames(pseudotime)]) )
# for(i in 1:ncol(pseudotime)){
#   pt1 <- pseudotime[,i]
#   l1 <- (cellWeights[,i] == 1 ) & (pt1 > lims[1]) & (pt1 < lims[2])
#   l1 <- colnames(temp)[l1]
#   points( pt1[ l1 ],cex=.3,pch=16,col=c("cyan4","orange3")[i],
#       DATA@assays$RNA@data[gene,l1]+runif(n = length(l1),0.0,0.02))
#   sm <- smooth.spline( pt1[ l1 ],
#                      DATA@assays$RNA@data[gene,l1], nknots = 20,spar = .8)
#   lines(sm$x,sm$y*1.3,lwd=3,col=c("navy","salmon2")[i] )
#   lines(sm$x,sm$y*1.3,lwd=.5,col="black")
# }


```


# Smoothed heatmap
```{r}

# png(filename = paste0(opt$output_path,"/Neuron_trajectory_branch.png"),width = 800*2,height = 800*3,res = 300)
# x <- t(apply(res,1,function(x){scale(x,T,T)}))
# to_order <- apply(x,1,function(x){ mean( (1:length(x))[order(x,decreasing = T)][1] )})
# to_order <- apply( (res[,1:20] - res[,22:41]),1,function(x){ mean( (1:length(x))[order(x,decreasing = F)][1] )})
# to_order <- apply(x,1,function(x){ which.max(x) })
# o <- order(to_order)
# o <- hclust(as.dist((1-cor(t(res),use = "pairwise.complete.obs"))/2),method = "complete" )$order
# par(mar=c(2,2,3,6))
# image( t(x[o,][nrow(res):1,]) ,axes=F,frame=F , col=c("grey95",colorRampPalette(c("grey95","grey70","firebrick","firebrick4"))(90)))
# text( par("usr")[2], seq(nrow(res),0,length.out = nrow(res)) / nrow(res) , labels = rownames(x)[o] , pos=4 ,xpd=T , cex=.8)
# text( par("usr")[1], par("usr")[4]-diff(par("usr")[4:3])/30, labels = "lineage 2" , adj= c(0,0),xpd=T , cex=.8 , pos=3)
# text( par("usr")[2], par("usr")[4]-diff(par("usr")[4:3])/30, labels = "lineage 1" , adj= c(1,0),xpd=T , cex=.8, pos=3)
# text( mean(par("usr")[1:2]), par("usr")[4]-diff(par("usr")[4:3])/30, labels = "origin" , adj= c(1,0),xpd=T , cex=.8, pos=3)
# 
# x <- curves@lineages$Lineage1
# text( seq(mean(par("usr")[1:2]),par("usr")[1],length.out = length(x)), par("usr")[4], labels = x , adj= c(0,0),xpd=T , cex=.8 , pos=3)
# x <- curves@lineages$Lineage2
# text( seq(mean(par("usr")[1:2]),par("usr")[2],length.out = length(x)), par("usr")[4], labels = x , adj= c(0,0),xpd=T , cex=.8 , pos=3)
# dev.off()
# 
# 
# # Lineage weights
# xs <- cellWeights
# plot(0,xlim=c(-1,1),ylim=range(pseudotime))
# sel <- xs[,2]>0.9 & (xs[,1]<=0.9)
# points(xs[sel,1]-xs[sel,2],pseudotime[sel,2],col="purple",cex=.5)
# sel <- xs[,1]>0.9 & (xs[,2]<=0.9)
# points(xs[sel,1]-xs[sel,2],pseudotime[sel,1],col="navy",cex=.5)
# sel <- (xs[,2]>0.9) & (xs[,1]>0.9)
# points(xs[sel,1]-xs[sel,2],pseudotime[sel,1],col="grey",cex=.5)
# 
# 
# 
# # Smoothed heatmap
# xs <- sapply(curves@curves,function(x){x$lambda})
# plot(c(-.3,.3),c(0,2.3),type="n")
# sel <- ( xs[,2] - xs[,1] ) > 0
# sum(sel)
# points(xs[sel,1] - xs[sel,2] ,pseudotime[sel,2],col="purple",cex=.5)
# sel <- ( xs[,2] - xs[,1] ) < 0
# sum(sel)
# points((xs[sel,1] - xs[sel,2]) ,pseudotime[sel,1],col="navy",cex=.5)


```



